{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Eventalign_collapse usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This program collapses the raw file generated by `nanopolish eventalign` by kmers rather than by event."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Options"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **input_fn**\n",
    "\n",
    "Path to a `nanopolish eventalign` tsv output file (read access required). In command line mode it is also possible to pipe the output of the `nanopolish eventalign` directly into `Eventalign_collapse`.\n",
    "\n",
    "* **outdir**\n",
    "\n",
    "Path to an existing directory where to write all the output files generated by `Eventalign_collapse`(write access required). If the directory does not exist an error is raised.\n",
    "\n",
    "* **outprefix**\n",
    "\n",
    "Prefix for all the files generated by the program\n",
    "\n",
    "* **write_samples**\n",
    "\n",
    "Write the concatenated sample data values corresponding to each kmer in the output data file. This options make only sense if `nanopolish eventalign` was ran with `--samples` option.\n",
    "  \n",
    "* **max_reads**\n",
    "\n",
    "Controls the maximum number of read to parse before stopping to read the input file. This could be usefull for testing or downsampling. \n",
    "\n",
    "* **stat_fields**\n",
    "\n",
    "Specify the list of statistical analyses to perform on a kmer basis and add in the output data file. This will only be performed if `nanopolish eventalign` was ran with `--samples` option.\n",
    "\n",
    "!!! note \"Valid statistics fields\"\n",
    "    * mean (mean of signal intensity)\n",
    "    * median (mean of signal intensity)\n",
    "    * std (standard deviation of signal intensity))\n",
    "    * mad (median absolute deviation of signal intensity))\n",
    "    * num_signals (Number of raw signal data point)\n",
    "\n",
    "* **threads**\n",
    "\n",
    "`Eventalign_collapse` is multi threaded to speed up the data processing and keep pace with Nanopolish if using the direct piping strategy. Take advantage of many threads if you have access to a large compute cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output files format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Contrary to `nanopolish eventalign` output text file, in `Eventalign_collapse` the reads are separated by a hashtag headers containing the read_id and ref_id. This reduces the redundancy and makes it easier to find the start and end of a read.\n",
    "\n",
    "Example : ```#7ef1d7b9-5824-4382-b23b-78d82c07ebbd\tYHR055C.```\n",
    "\n",
    "The main data file contains the following fields:\n",
    "\n",
    "* **ref_pos**: Reference sequence ID (contig).\n",
    "* **ref_kmer**: Sequence of the reference kmers.\n",
    "* **num_events**: Number of events for this kmer before collapsing.\n",
    "* **dwell_time**: dwell time for this kmer in seconds\n",
    "* **NNNNN_dwell_time**: dwell time of events for this kmers with a model sequence \"NNNNN\" (events ignored by nanopolish HMM).\n",
    "* **mismatch_dwell_time**:  dwell time of events for this kmers with a model sequence different from the reference kmer\n",
    "* **start_idx**: Only if nanopolish eventalign called with --signal_idx. Start coordinate on original raw signal in fast5 file\n",
    "* **end_idx**: Only if nanopolish eventalign called with --signal_idx. End coordinate on original raw signal in fast5 file\n",
    "* **mean**: Only if nanopolish eventalign called with --samples. Mean of the normalised signal values provided by Nanopolish eventalign\n",
    "* **median**: Only if nanopolish eventalign called with --samples. Median of the normalised signal values provided by Nanopolish eventalign\n",
    "* **std**: Only if nanopolish eventalign called with --samples. Standard deviation of the normalised signal values provided by Nanopolish eventalign\n",
    "* **mad**: Only if nanopolish eventalign called with --samples. Median absolute deviation of the normalised signal values provided by Nanopolish eventalign\n",
    "* **num_signals**: Only if nanopolish eventalign called with --samples. Number of raw signal points.\n",
    "* **samples**: Only if nanopolish eventalign called with --samples and Eventalign_collapse called with --write_samples. List of normalised signal intensity values for this kmer\n",
    "\n",
    "In addition `Eventalign_collapse` also generates an useful index file containing reads level information. It contains the following fields:\n",
    "\n",
    "* **read_id**: Name or index of the read\n",
    "* **ref_id**: Name of the reference sequence the read was aligned on (contig)\n",
    "* **ref_start**: Start coordinate of the alignment on the reference sequence\n",
    "* **ref_end**: End coordinate of the alignment on the reference sequence\n",
    "* **dwell_time**: Cumulative dwell time in seconds for the entire resquiggled sequence\n",
    "* **kmers**: Overall number of resquiggled kmers\n",
    "* **NNNNN_kmers**: Number of resquiggled kmers containing at least 1 event for which the model sequence was \"NNNNN\"\n",
    "* **mismatching_kmers**: Number of resquiggled kmers containing at least 1 event for which the model sequence diverged from the reference sequence\n",
    "* **missing_kmers**: Number of skipped/missing reference positions in nanopolish output\n",
    "* **byte_offset**: Number of characters before the start of the sequence in the main output file. **This can be used in conjunction with file.seek() to directly access the start of a read**. An example is provided in the Usage notebook.\n",
    "* **byte_len**: Length of characters after byte_offset to the end of the read, excluding the last newline. **This can be used in conjunction with read() to read all the text chunk corresponding to the read**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bash command line usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:44.181603Z",
     "start_time": "2019-05-02T09:32:43.073764Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: NanopolishComp Eventalign_collapse [-h] [-i INPUT_FN] [-o OUTDIR]\n",
      "                                          [-p OUTPREFIX] [-s] [-r MAX_READS]\n",
      "                                          [-f STAT_FIELDS [STAT_FIELDS ...]]\n",
      "                                          [-t THREADS] [-v | -q]\n",
      "\n",
      "Collapse the nanopolish eventalign output at kmers level and compute kmer\n",
      "level statistics\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -v, --verbose         Increase verbosity (default: False)\n",
      "  -q, --quiet           Reduce verbosity (default: False)\n",
      "\n",
      "Input/Output options:\n",
      "  -i INPUT_FN, --input_fn INPUT_FN\n",
      "                        Path to a nanopolish eventalign tsv output file. If\n",
      "                        '0' read from std input (default: 0)\n",
      "  -o OUTDIR, --outdir OUTDIR\n",
      "                        Path to the output folder (default: ./)\n",
      "  -p OUTPREFIX, --outprefix OUTPREFIX\n",
      "                        text outprefix for all the files generated (default:\n",
      "                        out)\n",
      "\n",
      "Run parameters options:\n",
      "  -s, --write_samples   If given, will write the raw sample if nanopolish\n",
      "                        eventalign was ran with --samples option (default:\n",
      "                        False)\n",
      "  -r MAX_READS, --max_reads MAX_READS\n",
      "                        Maximum number of read to parse. 0 to deactivate\n",
      "                        (default: 0)\n",
      "  -f STAT_FIELDS [STAT_FIELDS ...], --stat_fields STAT_FIELDS [STAT_FIELDS ...]\n",
      "                        List of statistical fields to compute if nanopolish\n",
      "                        eventalign was ran with --sample option. Valid values\n",
      "                        = mean, std, median, mad, num_signals (default:\n",
      "                        ['mean', 'median', 'num_signals'])\n",
      "\n",
      "Other options:\n",
      "  -t THREADS, --threads THREADS\n",
      "                        Total number of threads. 1 thread is used for the\n",
      "                        reader and 1 for the writer (default: 4)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "NanopolishComp Eventalign_collapse --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T17:02:13.860121Z",
     "start_time": "2018-08-10T17:02:13.847128Z"
    }
   },
   "source": [
    "#### From an existing nanopolish eventalign output to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:45.417414Z",
     "start_time": "2019-05-02T09:32:44.184921Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#0\tYGR240C\n",
      "ref_pos\tref_kmer\tnum_events\tdwell_time\tNNNNN_dwell_time\tmismatch_dwell_time\n",
      "1656\tGAAAA\t1\t0.00266\t0.0\t0.0\n",
      "1657\tAAAAC\t1\t0.00764\t0.0\t0.0\n",
      "1658\tAAACA\t1\t0.00398\t0.0\t0.0\n",
      "1659\tAACAA\t1\t0.00432\t0.0\t0.0\n",
      "1660\tACAAA\t1\t0.00498\t0.0\t0.0\n",
      "1661\tCAAAG\t1\t0.00564\t0.0\t0.0\n",
      "1662\tAAAGA\t1\t0.00963\t0.0\t0.0\n",
      "1663\tAAGAT\t1\t0.00299\t0.0\t0.0\n",
      "ref_id\tref_start\tref_end\tread_id\tkmers\tdwell_time\tNNNNN_kmers\tmismatch_kmers\tmissing_kmers\tbyte_offset\tbyte_len\n",
      "YGR240C\t1656\t2960\t0\t1250\t13.788570000000009\t35\t0\t54\t0\t38028\n",
      "YCR030C\t1578\t2576\t1\t971\t11.487010000000005\t23\t0\t27\t38029\t29704\n",
      "YHR174W\t0\t839\t2\t825\t9.659210000000002\t15\t0\t14\t67734\t24231\n",
      "YHR174W\t218\t1309\t3\t1028\t11.06325000000001\t36\t0\t63\t91966\t30801\n",
      "YHR174W\t462\t1309\t4\t818\t10.73776000000001\t18\t0\t29\t122768\t24589\n",
      "YLR441C\t173\t764\t5\t554\t5.556939999999999\t20\t0\t37\t147358\t16437\n",
      "YGR192C\t1\t989\t6\t927\t11.731470000000003\t37\t0\t61\t163796\t27388\n",
      "YDR500C\t9\t252\t8\t231\t2.9179100000000027\t7\t0\t14\t191185\t6751\n",
      "YGR192C\t3\t995\t7\t946\t12.30464000000001\t31\t0\t50\t197937\t28123\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments\n",
      "Testing output dir writability\n",
      "Starting to process files\n",
      "\r",
      "0 reads [00:00, ? reads/s]\r",
      "11 reads [00:00, 96.75 reads/s]\r",
      "21 reads [00:00, 118.91 reads/s]\n",
      "[Eventalign_collapse] total reads: 21 [116.81 reads/s]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "NanopolishComp Eventalign_collapse -i ./data/eventalign_collapse//nanopolish_reads.tsv -o ./output/eventalign_collapse/\n",
    "head ./output/eventalign_collapse/out_eventalign_collapse.tsv\n",
    "head ./output/eventalign_collapse/out_eventalign_collapse.tsv.idx"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From standard input to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:46.592916Z",
     "start_time": "2019-05-02T09:32:45.421223Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#0\tYGR240C\n",
      "ref_pos\tref_kmer\tnum_events\tdwell_time\tNNNNN_dwell_time\tmismatch_dwell_time\n",
      "1656\tGAAAA\t1\t0.00266\t0.0\t0.0\n",
      "1657\tAAAAC\t1\t0.00764\t0.0\t0.0\n",
      "1658\tAAACA\t1\t0.00398\t0.0\t0.0\n",
      "1659\tAACAA\t1\t0.00432\t0.0\t0.0\n",
      "1660\tACAAA\t1\t0.00498\t0.0\t0.0\n",
      "1661\tCAAAG\t1\t0.00564\t0.0\t0.0\n",
      "1662\tAAAGA\t1\t0.00963\t0.0\t0.0\n",
      "1663\tAAGAT\t1\t0.00299\t0.0\t0.0\n",
      "ref_id\tref_start\tref_end\tread_id\tkmers\tdwell_time\tNNNNN_kmers\tmismatch_kmers\tmissing_kmers\tbyte_offset\tbyte_len\n",
      "YGR240C\t1656\t2960\t0\t1250\t13.788570000000009\t35\t0\t54\t0\t38028\n",
      "YCR030C\t1578\t2576\t1\t971\t11.487010000000005\t23\t0\t27\t38029\t29704\n",
      "YHR174W\t0\t839\t2\t825\t9.659210000000002\t15\t0\t14\t67734\t24231\n",
      "YHR174W\t218\t1309\t3\t1028\t11.06325000000001\t36\t0\t63\t91966\t30801\n",
      "YHR174W\t462\t1309\t4\t818\t10.73776000000001\t18\t0\t29\t122768\t24589\n",
      "YLR441C\t173\t764\t5\t554\t5.556939999999999\t20\t0\t37\t147358\t16437\n",
      "YGR192C\t1\t989\t6\t927\t11.731470000000003\t37\t0\t61\t163796\t27388\n",
      "YDR500C\t9\t252\t8\t231\t2.9179100000000027\t7\t0\t14\t191185\t6751\n",
      "YGR192C\t3\t995\t7\t946\t12.30464000000001\t31\t0\t50\t197937\t28123\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Options summary\n",
      "\tpackage_name: NanopolishComp\n",
      "\tpackage_version: 0.6.1\n",
      "\ttimestamp: 2019-05-02 10:32:46.350455\n",
      "\tquiet: False\n",
      "\tverbose: True\n",
      "\tthreads: 4\n",
      "\tstat_fields: ['mean', 'median', 'num_signals']\n",
      "\twrite_samples: False\n",
      "\tmax_reads: None\n",
      "\toutprefix: out\n",
      "\toutdir: ./output\n",
      "\tinput_fn: 0\n",
      "\n",
      "Checking arguments\n",
      "\tTesting input file readability\n",
      "Testing output dir writability\n",
      "\tChecking number of threads\n",
      "\tChecking if stat_fields names are valid\n",
      "Starting to process files\n",
      "\t[split_reads] Start reading input file/stream\n",
      "\t[process_read 1] Starting processing reads\n",
      "\t[process_read 2] Starting processing reads\n",
      "\t[write_output] Start rwriting output\n",
      "\r",
      "0 reads [00:00, ? reads/s]\r",
      "12 reads [00:00, 119.55 reads/s]\t[split_reads] Done\n",
      "\t[process_read 1] Done\n",
      "\t[process_read 2] Done\n",
      "\r",
      "21 reads [00:00, 116.86 reads/s]\n",
      "\t[write_output] Done\n",
      "[Eventalign_collapse] total reads: 21 [115.22 reads/s]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "cat ./data/eventalign_collapse//nanopolish_reads_index.tsv | NanopolishComp Eventalign_collapse -o ./output --verbose\n",
    "head ./output/eventalign_collapse/out_eventalign_collapse.tsv\n",
    "head ./output/eventalign_collapse/out_eventalign_collapse.tsv.idx"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### On the fly, from nanopolish eventalign to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:47.658878Z",
     "start_time": "2019-05-02T09:32:46.598203Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#0\tYGR240C\n",
      "ref_pos\tref_kmer\tnum_events\tdwell_time\tNNNNN_dwell_time\tmismatch_dwell_time\n",
      "1656\tGAAAA\t1\t0.00266\t0.0\t0.0\n",
      "1657\tAAAAC\t1\t0.00764\t0.0\t0.0\n",
      "1658\tAAACA\t1\t0.00398\t0.0\t0.0\n",
      "1659\tAACAA\t1\t0.00432\t0.0\t0.0\n",
      "1660\tACAAA\t1\t0.00498\t0.0\t0.0\n",
      "1661\tCAAAG\t1\t0.00564\t0.0\t0.0\n",
      "1662\tAAAGA\t1\t0.00963\t0.0\t0.0\n",
      "1663\tAAGAT\t1\t0.00299\t0.0\t0.0\n",
      "ref_id\tref_start\tref_end\tread_id\tkmers\tdwell_time\tNNNNN_kmers\tmismatch_kmers\tmissing_kmers\tbyte_offset\tbyte_len\n",
      "YGR240C\t1656\t2960\t0\t1250\t13.788570000000009\t35\t0\t54\t0\t38028\n",
      "YCR030C\t1578\t2576\t1\t971\t11.487010000000005\t23\t0\t27\t38029\t29704\n",
      "YHR174W\t0\t839\t2\t825\t9.659210000000002\t15\t0\t14\t67734\t24231\n",
      "YHR174W\t218\t1309\t3\t1028\t11.06325000000001\t36\t0\t63\t91966\t30801\n",
      "YHR174W\t462\t1309\t4\t818\t10.73776000000001\t18\t0\t29\t122768\t24589\n",
      "YLR441C\t173\t764\t5\t554\t5.556939999999999\t20\t0\t37\t147358\t16437\n",
      "YGR192C\t1\t989\t6\t927\t11.731470000000003\t37\t0\t61\t163796\t27388\n",
      "YDR500C\t9\t252\t8\t231\t2.9179100000000027\t7\t0\t14\t191185\t6751\n",
      "YGR192C\t3\t995\t7\t946\t12.30464000000001\t31\t0\t50\t197937\t28123\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_319_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_243_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_243_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_243_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_36_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_36_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_36_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_343_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_343_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_343_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_237_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_237_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_237_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205861861120:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_348_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_348_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205861861120:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205861861120:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_348_strand.fast5\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_108_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_318_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_318_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_318_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_290_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_290_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_290_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_103_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_103_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_103_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_418_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_418_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_418_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_191_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_191_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_191_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_301_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_301_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_301_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_224_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_224_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_224_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_410_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_410_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_410_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_382_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_382_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_382_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_192_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_192_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_192_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_113_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_113_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205853468416:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_113_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_382_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_382_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205845075712:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_382_strand.fast5\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_38_strand.fast5\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205861861120:\n",
      "  #000: H5F.c line 509 in H5Fopen(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #001: H5Fint.c line 1400 in H5F__open(): unable to open file\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #002: H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu May  2 10:32:47 2019\n",
      ", name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_354_strand.fast5', tent_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "  #003: H5FD.c line 734 in H5FD_open(): open failed\n",
      "    major: Virtual File Layer\n",
      "    minor: Unable to initialize object\n",
      "  #004: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = './data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_354_strand.fast5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0\n",
      "    major: File accessibilty\n",
      "    minor: Unable to open file\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205861861120:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 140205861861120:\n",
      "  #000: H5L.c line 805 in H5Lexists(): not a location\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Inappropriate type\n",
      "  #001: H5Gloc.c line 246 in H5G_loc(): invalid object ID\n",
      "    major: Invalid arguments to routine\n",
      "    minor: Bad value\n",
      "[warning] fast5 file is unreadable and will be skipped: ./data/fast5//20180625_FAH77625_MN23126_sequencing_run_S1_57529_read_10_ch_354_strand.fast5\n",
      "[post-run summary] total reads: 21, unparseable: 0, qc fail: 0, could not calibrate: 0, no alignment: 0, bad fast5: 21\n",
      "Checking arguments\n",
      "Testing output dir writability\n",
      "Starting to process files\n",
      "\n",
      "An error occured. All processes were killed\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/aleg/.virtualenvs/Nanopolish_0.11.1/bin/NanopolishComp\", line 11, in <module>\n",
      "    load_entry_point('NanopolishComp', 'console_scripts', 'NanopolishComp')()\n",
      "  File \"/home/aleg/Programming/NanopolishComp/NanopolishComp/__main__.py\", line 65, in main\n",
      "    args.func(args)\n",
      "  File \"/home/aleg/Programming/NanopolishComp/NanopolishComp/__main__.py\", line 80, in Eventalign_collapse_main\n",
      "    quiet = args.quiet)\n",
      "  File \"/home/aleg/Programming/NanopolishComp/NanopolishComp/Eventalign_collapse.py\", line 148, in __init__\n",
      "    raise E\n",
      "  File \"/home/aleg/Programming/NanopolishComp/NanopolishComp/Eventalign_collapse.py\", line 138, in __init__\n",
      "    raise E\n",
      "NanopolishComp.common.NanopolishCompError: Traceback (most recent call last):\n",
      "  File \"/home/aleg/Programming/NanopolishComp/NanopolishComp/Eventalign_collapse.py\", line 176, in _split_reads\n",
      "    cur_read_id = event_l[idx[\"read_id\"]]\n",
      "IndexError: list index out of range\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "nanopolish eventalign -t 4 --samples --scale-events --print-read-name --reads ./data/eventalign_collapse//reads.fastq --bam ./data/eventalign_collapse//aligned_reads.bam --genome ./data/eventalign_collapse//reference.fa | NanopolishComp Eventalign_collapse -o ./output\n",
    "head ./output/eventalign_collapse/out_eventalign_collapse.tsv\n",
    "head ./output/eventalign_collapse/out_eventalign_collapse.tsv.idx"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import the package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:47.744368Z",
     "start_time": "2019-05-02T09:32:47.663524Z"
    }
   },
   "outputs": [],
   "source": [
    "# Import main program\n",
    "from NanopolishComp.Eventalign_collapse import Eventalign_collapse\n",
    "\n",
    "# Import helper functions\n",
    "from NanopolishComp.common import jhelp, head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### python API help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:47.769186Z",
     "start_time": "2019-05-02T09:32:47.746336Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "**NanopolishComp.Eventalign_collapse.__init__**\n",
       "\n",
       "Collapse the nanopolish eventalign output by kmers rather that by events. kmer level statistics (mean, median, std, mad) are only computed if nanopolish is run with --samples option\n",
       "\n",
       "---\n",
       "\n",
       "* **input_fn** *: str (required)*\n",
       "\n",
       "Path to a nanopolish eventalign tsv output file.\n",
       "\n",
       "* **outdir** *: str (default = ./)*\n",
       "\n",
       "Path to the output folder\n",
       "\n",
       "* **outprefix** *: str (default = out)*\n",
       "\n",
       "text outprefix for all the files generated\n",
       "\n",
       "* **max_reads** *: int (default = None)*\n",
       "\n",
       "Maximum number of read to parse. 0 to deactivate (default = 0)\n",
       "\n",
       "* **write_samples** *: bool (default = False)*\n",
       "\n",
       "If given, will write the raw sample if nanopolish eventalign was ran with --samples option\n",
       "\n",
       "* **stat_fields** *: list of str (default = ['mean', 'median', 'num_signals'])*\n",
       "\n",
       "List of statistical fields to compute if nanopolish eventalign was ran with --samples option. Valid values = \"mean\", \"std\", \"median\", \"mad\", \"num_signals\"\n",
       "\n",
       "* **threads** *: int (default = 4)*\n",
       "\n",
       "Total number of threads. 1 thread is used for the reader and 1 for the writer (default = 4)\n",
       "\n",
       "* **verbose** *: bool (default = False)*\n",
       "\n",
       "Increase verbosity\n",
       "\n",
       "* **quiet** *: bool (default = False)*\n",
       "\n",
       "Reduce verbosity\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(Eventalign_collapse)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example with minimal file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:48.110916Z",
     "start_time": "2019-05-02T09:32:47.771813Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments\n",
      "Testing output dir writability\n",
      "Starting to process files\n",
      "21 reads [00:00, 83.37 reads/s]\n",
      "[Eventalign_collapse] total reads: 21 [78.29 reads/s]\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#0\tYGR240C\n",
      "ref_pos\tref_kmer\tnum_events\tdwell_time\tNNNNN_dwell_time\tmismatch_dwell_time\n",
      "1656\tGAAAA\t1\t0.00266\t0.0\t0.0\n",
      "1657\tAAAAC\t1\t0.00764\t0.0\t0.0\n",
      "1658\tAAACA\t1\t0.00398\t0.0\t0.0\n",
      "1659\tAACAA\t1\t0.00432\t0.0\t0.0\n",
      "1660\tACAAA\t1\t0.00498\t0.0\t0.0\n",
      "1661\tCAAAG\t1\t0.00564\t0.0\t0.0\n",
      "1662\tAAAGA\t1\t0.00963\t0.0\t0.0\n",
      "1663\tAAGAT\t1\t0.00299\t0.0\t0.0\n",
      "\n",
      "ref_id  ref_start ref_end read_id kmers dwell_time         NNNNN_kmers mismatch_kmers missing_kmers byte_offset byte_len \n",
      "YGR240C 1656      2960    0       1250  13.788570000000009 35          0              54            0           38028    \n",
      "YCR030C 1578      2576    1       971   11.487010000000005 23          0              27            38029       29704    \n",
      "YHR174W 0         839     2       825   9.659210000000002  15          0              14            67734       24231    \n",
      "YHR174W 218       1309    3       1028  11.06325000000001  36          0              63            91966       30801    \n",
      "YLR441C 173       764     5       554   5.556939999999999  20          0              37            122768      16437    \n",
      "YHR174W 462       1309    4       818   10.73776000000001  18          0              29            139206      24589    \n",
      "YGR192C 1         989     6       927   11.731470000000003 37          0              61            163796      27388    \n",
      "YGR192C 3         995     7       946   12.30464000000001  31          0              50            191185      28123    \n",
      "YDR500C 9         252     8       231   2.9179100000000027 7           0              14            219309      6751     \n",
      "\n"
     ]
    }
   ],
   "source": [
    "input_fn = \"./data/eventalign_collapse//nanopolish_reads.tsv\"\n",
    "outdir = \"./output/eventalign_collapse\"\n",
    "outprefix = \"basic\"\n",
    "\n",
    "Eventalign_collapse (input_fn=input_fn, outdir=outdir, outprefix=outprefix, threads=6)\n",
    "\n",
    "head(\"./output/eventalign_collapse/basic_eventalign_collapse.tsv\")\n",
    "head(\"./output/eventalign_collapse/basic_eventalign_collapse.tsv.idx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example with indexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:48.423433Z",
     "start_time": "2019-05-02T09:32:48.115915Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking arguments\n",
      "Testing output dir writability\n",
      "Starting to process files\n",
      "21 reads [00:00, 92.53 reads/s]\n",
      "[Eventalign_collapse] total reads: 21 [87.47 reads/s]\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#0\tYGR240C\n",
      "ref_pos\tref_kmer\tnum_events\tdwell_time\tNNNNN_dwell_time\tmismatch_dwell_time\tstart_idx\tend_idx\n",
      "1656\tGAAAA\t1\t0.00266\t0.0\t0.0\t63446\t63454\n",
      "1657\tAAAAC\t1\t0.00764\t0.0\t0.0\t63423\t63446\n",
      "1658\tAAACA\t1\t0.00398\t0.0\t0.0\t63411\t63423\n",
      "1659\tAACAA\t1\t0.00432\t0.0\t0.0\t63398\t63411\n",
      "1660\tACAAA\t1\t0.00498\t0.0\t0.0\t63383\t63398\n",
      "1661\tCAAAG\t1\t0.00564\t0.0\t0.0\t63366\t63383\n",
      "1662\tAAAGA\t1\t0.00963\t0.0\t0.0\t63337\t63366\n",
      "1663\tAAGAT\t1\t0.00299\t0.0\t0.0\t63328\t63337\n",
      "\n",
      "ref_id  ref_start ref_end read_id kmers dwell_time         NNNNN_kmers mismatch_kmers missing_kmers byte_offset byte_len \n",
      "YGR240C 1656      2960    0       1250  13.788570000000009 35          0              54            0           53046    \n",
      "YCR030C 1578      2576    1       971   11.487010000000005 23          0              27            53047       41374    \n",
      "YHR174W 0         839     2       825   9.659210000000002  15          0              14            94422       33595    \n",
      "YHR174W 218       1309    3       1028  11.06325000000001  36          0              63            128018      43155    \n",
      "YHR174W 462       1309    4       818   10.73776000000001  18          0              29            171174      34423    \n",
      "YLR441C 173       764     5       554   5.556939999999999  20          0              37            205598      23103    \n",
      "YGR192C 1         989     6       927   11.731470000000003 37          0              61            228702      38529    \n",
      "YDR500C 9         252     8       231   2.9179100000000027 7           0              14            267232      9541     \n",
      "YDR224C 0         389     9       376   4.2840300000000004 8           0              13            276774      15465    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "input_fn = \"./data/eventalign_collapse//nanopolish_reads_index.tsv\"\n",
    "outdir = \"./output/eventalign_collapse\"\n",
    "outprefix = \"index\"\n",
    "\n",
    "Eventalign_collapse (input_fn=input_fn, outdir=outdir, outprefix=outprefix, threads=6)\n",
    "\n",
    "head(\"./output/eventalign_collapse/index_eventalign_collapse.tsv\")\n",
    "head(\"./output/eventalign_collapse/index_eventalign_collapse.tsv.idx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example including samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:48.663769Z",
     "start_time": "2019-05-02T09:32:48.425846Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Eventalign_collapse] total reads: 21 [106.77 reads/s]\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#0\tYGR240C\n",
      "ref_pos\tref_kmer\tnum_events\tdwell_time\tNNNNN_dwell_time\tmismatch_dwell_time\tstart_idx\tend_idx\n",
      "1656\tGAAAA\t1\t0.00266\t0.0\t0.0\t63446\t63454\n",
      "1657\tAAAAC\t1\t0.00764\t0.0\t0.0\t63423\t63446\n",
      "1658\tAAACA\t1\t0.00398\t0.0\t0.0\t63411\t63423\n",
      "1659\tAACAA\t1\t0.00432\t0.0\t0.0\t63398\t63411\n",
      "1660\tACAAA\t1\t0.00498\t0.0\t0.0\t63383\t63398\n",
      "1661\tCAAAG\t1\t0.00564\t0.0\t0.0\t63366\t63383\n",
      "1662\tAAAGA\t1\t0.00963\t0.0\t0.0\t63337\t63366\n",
      "1663\tAAGAT\t1\t0.00299\t0.0\t0.0\t63328\t63337\n",
      "\n",
      "ref_id  ref_start ref_end read_id kmers dwell_time         NNNNN_kmers mismatch_kmers missing_kmers byte_offset byte_len \n",
      "YGR240C 1656      2960    0       1250  13.788570000000009 35          0              54            0           53046    \n",
      "YCR030C 1578      2576    1       971   11.487010000000005 23          0              27            53047       41374    \n",
      "YHR174W 0         839     2       825   9.659210000000002  15          0              14            94422       33595    \n",
      "YHR174W 462       1309    4       818   10.73776000000001  18          0              29            128018      34423    \n",
      "YHR174W 218       1309    3       1028  11.06325000000001  36          0              63            162442      43155    \n",
      "YLR441C 173       764     5       554   5.556939999999999  20          0              37            205598      23103    \n",
      "YGR192C 1         989     6       927   11.731470000000003 37          0              61            228702      38529    \n",
      "YDR500C 9         252     8       231   2.9179100000000027 7           0              14            267232      9541     \n",
      "YDR224C 0         389     9       376   4.2840300000000004 8           0              13            276774      15465    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "input_fn = \"./data/eventalign_collapse//nanopolish_reads_index.tsv\"\n",
    "outdir = \"./output/eventalign_collapse\"\n",
    "outprefix = \"stats\"\n",
    "\n",
    "Eventalign_collapse (input_fn=input_fn, outdir=outdir, outprefix=outprefix, threads=6, quiet=True, stat_fields=[\"mean\", \"std\", \"median\", \"mad\", \"num_signals\"])\n",
    "\n",
    "head(\"./output/eventalign_collapse/stats_eventalign_collapse.tsv\")\n",
    "head(\"./output/eventalign_collapse/stats_eventalign_collapse.tsv.idx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example including samples and writing samples values in ouput file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:50.536323Z",
     "start_time": "2019-05-02T09:32:48.666817Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Eventalign_collapse] total reads: 21 [11.5 reads/s]\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#1\tYCR030C\n",
      "ref_pos\tref_kmer\tnum_events\tdwell_time\tNNNNN_dwell_time\tmismatch_dwell_time\tmean\tstd\tmedian\tmad\tnum_signals\tsamples\n",
      "1578\tCCACC\t1\t0.00365\t0.0\t0.0\t70.06341552734375\t2.593737840652466\t70.67729949951172\t1.4680023193359375\t11\t70.6773,72.1454,71.4114,63.6306,69.2093,69.2093,67.3008,72.5858,69.9433,72.8794,71.705\n",
      "1579\tCACCC\t15\t0.09761000000000002\t0.0\t0.0\t74.59468841552734\t3.2702603340148926\t75.08149719238281\t2.4958038330078125\t294\t75.0815,75.8156,79.6326,78.0177,75.3752,77.2837,78.0177,72.2922,76.99,74.7879,75.3752,75.522,75.6688,73.3199,74.9347,77.5773,73.7603,74.3475,72.7326,75.522,77.2837,75.2284,76.6964,78.4581,78.4581,79.7794,76.5496,76.8432,70.2369,70.2369,81.1006,77.8709,74.9347,72.439,75.8156,77.5773,75.9624,77.4305,76.1092,77.2837,64.5114,74.4943,78.4581,77.4305,76.8432,75.0815,76.1092,78.3113,76.5496,75.6688,76.5496,77.4305,78.6049,76.4028,75.8156,77.1368,71.8518,76.4028,75.8156,75.9624,74.9347,78.3113,72.2922,75.6688,75.0815,75.522,75.9624,70.6773,69.9433,69.2093,68.3284,69.3561,69.2093,69.7965,67.888,68.9157,68.622,69.0625,69.7965,73.173,76.1092,74.7879,75.2284,78.3113,76.256,78.3113,77.7241,72.7326,73.6135,78.6049,75.522,75.8156,74.0539,67.3008,72.439,72.2922,67.0072,74.6411,76.1092,75.8156,73.3199,77.2837,70.3837,73.4667,74.7879,76.6964,79.1921,73.3199,76.6964,76.4028,71.1178,76.1092,79.7794,77.7241,77.1368,74.3475,78.6049,78.8985,77.2837,77.7241,70.0901,79.0453,76.1092,73.6135,78.1645,79.0453,75.8156,79.3389,79.0453,77.1368,78.1645,78.3113,77.8709,79.4858,75.8156,76.4028,77.2837,77.4305,79.6326,77.4305,81.8347,76.5496,76.5496,78.8985,74.7879,77.2837,80.2198,75.6688,78.1645,76.99,79.3389,79.0453,71.5582,78.6049,79.7794,78.3113,78.1645,75.3752,76.8432,80.073,78.3113,80.2198,80.073,80.9538,77.5773,77.8709,75.6688,80.807,77.8709,78.0177,79.3389,80.2198,80.2198,76.1092,77.8709,76.5496,76.6964,77.8709,77.1368,75.9624,76.99,72.2922,71.8518,73.9071,68.9157,70.0901,71.8518,75.3752,73.9071,72.1454,73.6135,74.4943,73.173,75.9624,75.0815,74.6411,75.0815,75.2284,74.4943,75.522,76.6964,72.7326,73.7603,74.9347,75.522,74.0539,71.2646,70.3837,72.439,75.522,75.9624,71.705,72.2922,70.8241,71.1178,70.971,74.7879,75.6688,74.6411,72.1454,73.0262,71.4114,73.9071,74.2007,75.0815,74.4943,72.1454,72.8794,72.1454,69.6497,71.5582,72.439,70.6773,69.7965,68.4752,68.7689,73.6135,73.7603,72.2922,71.4114,69.2093,70.2369,73.3199,72.439,71.5582,67.4476,72.8794,72.2922,72.5858,72.439,73.9071,69.6497,72.2922,73.6135,74.4943,73.0262,70.0901,72.439,71.8518,72.1454,71.9986,70.2369,74.0539,70.971,72.1454,71.2646,71.8518,71.1178,71.5582,70.971,69.2093,70.8241,71.4114,71.2646,72.1454,70.5305,69.6497,70.6773,69.9433,72.8794,67.5944,66.7135,70.5305,69.7965,68.622,73.4667,73.7603,76.1092,77.1368,74.3475,75.8156,74.7879,76.8432,75.8156\n",
      "1580\tACCCT\t1\t0.00232\t0.0\t0.0\t67.23785400390625\t1.7063778638839722\t67.88800048828125\t2.2021026611328125\t7\t64.9519,65.6859,67.888,68.1816,68.1816,65.6859,70.0901\n",
      "1581\tCCCTC\t1\t0.00232\t0.0\t0.0\t72.0615005493164\t1.1475614309310913\t72.58580017089844\t1.0277023315429688\t7\t70.3837,72.5858,70.6773,71.4114,73.0262,73.6135,72.7326\n",
      "1582\tCCTCA\t5\t0.0156\t0.0\t0.0\t73.5634994506836\t2.3691813945770264\t73.76029968261719\t1.614898681640625\t47\t75.522,77.8709,75.2284,74.3475,75.9624,76.6964,72.1454,71.1178,71.1178,72.439,76.256,74.0539,73.0262,73.4667,70.6773,71.5582,69.7965,71.1178,68.3284,70.6773,73.6135,69.6497,67.4476,73.7603,74.3475,72.8794,74.0539,73.3199,74.4943,75.522,72.8794,72.5858,71.2646,73.4667,74.3475,74.4943,74.9347,71.9986,72.2922,76.8432,74.9347,76.5496,75.6688,75.2284,75.0815,76.6964,77.7241\n",
      "1583\tCTCAA\t1\t0.00398\t0.0\t0.0\t77.58948516845703\t1.5072672367095947\t78.16450500488281\t0.7339935302734375\t12\t74.3475,75.0815,78.3113,78.8985,77.8709,78.3113,76.4028,78.0177,77.1368,78.3113,78.8985,79.4858\n",
      "1584\tTCAAG\t1\t0.01096\t0.0\t0.0\t83.52072143554688\t3.1505558490753174\t84.03679656982422\t2.3488998413085938\t33\t88.441,79.9262,81.6879,81.9815,85.0644,86.9729,84.624,81.1006,76.99,79.0453,79.9262,84.0368,92.1112,83.0091,84.1836,80.807,84.0368,85.7985,87.2665,85.6516,83.0091,87.7069,81.2474,84.0368,79.1921,79.6326,87.2665,83.4495,84.0368,83.7432,80.807,84.624,84.7708\n",
      "1585\tCAAGG\t1\t0.00232\t0.0\t0.0\t105.23970794677734\t3.632479667663574\t104.2959976196289\t1.7620010375976562\t7\t113.251,102.534,104.296,103.415,106.498,101.36,105.324\n",
      "\n",
      "ref_id  ref_start ref_end read_id kmers dwell_time         NNNNN_kmers mismatch_kmers missing_kmers byte_offset byte_len \n",
      "YCR030C 1578      2576    1       971   11.487010000000005 23          0              27            0           376326   \n",
      "YHR174W 0         839     2       825   9.659210000000002  15          0              14            376327      317027   \n",
      "YHR174W 218       1309    3       1028  11.06325000000001  36          0              63            693355      372126   \n",
      "YGR240C 1656      2960    0       1250  13.788570000000009 35          0              54            1065482     461011   \n",
      "YLR441C 173       764     5       554   5.556939999999999  20          0              37            1526494     190670   \n",
      "YHR174W 462       1309    4       818   10.73776000000001  18          0              29            1717165     342528   \n",
      "YDR500C 9         252     8       231   2.9179100000000027 7           0              14            2059694     93821    \n",
      "YGR192C 1         989     6       927   11.731470000000003 37          0              61            2153516     376317   \n",
      "YDR224C 0         389     9       376   4.2840300000000004 8           0              13            2529834     141262   \n",
      "\n"
     ]
    }
   ],
   "source": [
    "input_fn = \"./data/eventalign_collapse//nanopolish_reads_samples.tsv\"\n",
    "outdir = \"./output/eventalign_collapse\"\n",
    "outprefix = \"samples\"\n",
    "\n",
    "Eventalign_collapse (input_fn=input_fn, outdir=outdir, outprefix=outprefix, threads=6, quiet=True, stat_fields=[\"mean\", \"std\", \"median\", \"mad\", \"num_signals\"], write_samples=True)\n",
    "\n",
    "head(\"./output/eventalign_collapse/samples_eventalign_collapse.tsv\")\n",
    "head(\"./output/eventalign_collapse/samples_eventalign_collapse.tsv.idx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using the index to random access a specific entry in the file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:32:51.031323Z",
     "start_time": "2019-05-02T09:32:50.543356Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Random index lines\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_id</th>\n",
       "      <th>ref_start</th>\n",
       "      <th>ref_end</th>\n",
       "      <th>read_id</th>\n",
       "      <th>kmers</th>\n",
       "      <th>dwell_time</th>\n",
       "      <th>NNNNN_kmers</th>\n",
       "      <th>mismatch_kmers</th>\n",
       "      <th>missing_kmers</th>\n",
       "      <th>byte_offset</th>\n",
       "      <th>byte_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>YIL117C</td>\n",
       "      <td>27</td>\n",
       "      <td>953</td>\n",
       "      <td>10</td>\n",
       "      <td>889</td>\n",
       "      <td>13.64007</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>331734</td>\n",
       "      <td>37225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>YPL198W</td>\n",
       "      <td>10</td>\n",
       "      <td>719</td>\n",
       "      <td>13</td>\n",
       "      <td>667</td>\n",
       "      <td>8.79275</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>44</td>\n",
       "      <td>409081</td>\n",
       "      <td>27943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>YDR500C</td>\n",
       "      <td>9</td>\n",
       "      <td>252</td>\n",
       "      <td>8</td>\n",
       "      <td>231</td>\n",
       "      <td>2.91791</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>267232</td>\n",
       "      <td>9541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>YGR192C</td>\n",
       "      <td>1</td>\n",
       "      <td>989</td>\n",
       "      <td>6</td>\n",
       "      <td>927</td>\n",
       "      <td>11.73147</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>228702</td>\n",
       "      <td>38529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>YLR110C</td>\n",
       "      <td>14</td>\n",
       "      <td>395</td>\n",
       "      <td>15</td>\n",
       "      <td>364</td>\n",
       "      <td>4.61444</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>450178</td>\n",
       "      <td>15023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ref_id  ref_start  ref_end  read_id  kmers  dwell_time  NNNNN_kmers  \\\n",
       "10  YIL117C         27      953       10    889    13.64007           23   \n",
       "13  YPL198W         10      719       13    667     8.79275           30   \n",
       "7   YDR500C          9      252        8    231     2.91791            7   \n",
       "6   YGR192C          1      989        6    927    11.73147           37   \n",
       "15  YLR110C         14      395       15    364     4.61444            9   \n",
       "\n",
       "    mismatch_kmers  missing_kmers  byte_offset  byte_len  \n",
       "10               0             37       331734     37225  \n",
       "13               0             44       409081     27943  \n",
       "7                0             14       267232      9541  \n",
       "6                0             61       228702     38529  \n",
       "15               0             17       450178     15023  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#10\tYIL117C\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ref_kmer</th>\n",
       "      <th>num_events</th>\n",
       "      <th>dwell_time</th>\n",
       "      <th>NNNNN_dwell_time</th>\n",
       "      <th>mismatch_dwell_time</th>\n",
       "      <th>start_idx</th>\n",
       "      <th>end_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27</td>\n",
       "      <td>GGCCT</td>\n",
       "      <td>4</td>\n",
       "      <td>0.03287</td>\n",
       "      <td>0.00498</td>\n",
       "      <td>0.0</td>\n",
       "      <td>57497</td>\n",
       "      <td>57596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31</td>\n",
       "      <td>TTCCA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00398</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>57485</td>\n",
       "      <td>57497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>887</th>\n",
       "      <td>951</td>\n",
       "      <td>GAGTA</td>\n",
       "      <td>3</td>\n",
       "      <td>0.01163</td>\n",
       "      <td>0.00432</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16588</td>\n",
       "      <td>16623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888</th>\n",
       "      <td>952</td>\n",
       "      <td>AGTAA</td>\n",
       "      <td>5</td>\n",
       "      <td>0.02590</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16510</td>\n",
       "      <td>16588</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>889 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ref_pos ref_kmer  num_events  dwell_time  NNNNN_dwell_time  \\\n",
       "0         27    GGCCT           4     0.03287           0.00498   \n",
       "1         31    TTCCA           1     0.00398           0.00000   \n",
       "..       ...      ...         ...         ...               ...   \n",
       "887      951    GAGTA           3     0.01163           0.00432   \n",
       "888      952    AGTAA           5     0.02590           0.00000   \n",
       "\n",
       "     mismatch_dwell_time  start_idx  end_idx  \n",
       "0                    0.0      57497    57596  \n",
       "1                    0.0      57485    57497  \n",
       "..                   ...        ...      ...  \n",
       "887                  0.0      16588    16623  \n",
       "888                  0.0      16510    16588  \n",
       "\n",
       "[889 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#13\tYPL198W\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ref_kmer</th>\n",
       "      <th>num_events</th>\n",
       "      <th>dwell_time</th>\n",
       "      <th>NNNNN_dwell_time</th>\n",
       "      <th>mismatch_dwell_time</th>\n",
       "      <th>start_idx</th>\n",
       "      <th>end_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>AAAAA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00697</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>44620</td>\n",
       "      <td>44641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>AAAAA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01959</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>44561</td>\n",
       "      <td>44620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>715</td>\n",
       "      <td>TGGTT</td>\n",
       "      <td>4</td>\n",
       "      <td>0.02557</td>\n",
       "      <td>0.00797</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18191</td>\n",
       "      <td>18268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666</th>\n",
       "      <td>718</td>\n",
       "      <td>TTAAG</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01162</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18156</td>\n",
       "      <td>18191</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>667 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ref_pos ref_kmer  num_events  dwell_time  NNNNN_dwell_time  \\\n",
       "0         10    AAAAA           1     0.00697           0.00000   \n",
       "1         11    AAAAA           1     0.01959           0.00000   \n",
       "..       ...      ...         ...         ...               ...   \n",
       "665      715    TGGTT           4     0.02557           0.00797   \n",
       "666      718    TTAAG           1     0.01162           0.00000   \n",
       "\n",
       "     mismatch_dwell_time  start_idx  end_idx  \n",
       "0                    0.0      44620    44641  \n",
       "1                    0.0      44561    44620  \n",
       "..                   ...        ...      ...  \n",
       "665                  0.0      18191    18268  \n",
       "666                  0.0      18156    18191  \n",
       "\n",
       "[667 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#8\tYDR500C\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ref_kmer</th>\n",
       "      <th>num_events</th>\n",
       "      <th>dwell_time</th>\n",
       "      <th>NNNNN_dwell_time</th>\n",
       "      <th>mismatch_dwell_time</th>\n",
       "      <th>start_idx</th>\n",
       "      <th>end_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>GGTAC</td>\n",
       "      <td>2</td>\n",
       "      <td>0.01095</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22637</td>\n",
       "      <td>22670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>GTACT</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00631</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22618</td>\n",
       "      <td>22637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>248</td>\n",
       "      <td>TAAGG</td>\n",
       "      <td>2</td>\n",
       "      <td>0.01560</td>\n",
       "      <td>0.01228</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13899</td>\n",
       "      <td>13946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>251</td>\n",
       "      <td>GGCTA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00598</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13881</td>\n",
       "      <td>13899</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>231 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ref_pos ref_kmer  num_events  dwell_time  NNNNN_dwell_time  \\\n",
       "0          9    GGTAC           2     0.01095           0.00000   \n",
       "1         10    GTACT           1     0.00631           0.00000   \n",
       "..       ...      ...         ...         ...               ...   \n",
       "229      248    TAAGG           2     0.01560           0.01228   \n",
       "230      251    GGCTA           1     0.00598           0.00000   \n",
       "\n",
       "     mismatch_dwell_time  start_idx  end_idx  \n",
       "0                    0.0      22637    22670  \n",
       "1                    0.0      22618    22637  \n",
       "..                   ...        ...      ...  \n",
       "229                  0.0      13899    13946  \n",
       "230                  0.0      13881    13899  \n",
       "\n",
       "[231 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#6\tYGR192C\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ref_kmer</th>\n",
       "      <th>num_events</th>\n",
       "      <th>dwell_time</th>\n",
       "      <th>NNNNN_dwell_time</th>\n",
       "      <th>mismatch_dwell_time</th>\n",
       "      <th>start_idx</th>\n",
       "      <th>end_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TGGTT</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01461</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>45267</td>\n",
       "      <td>45311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>GGTTA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00631</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>45248</td>\n",
       "      <td>45267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>925</th>\n",
       "      <td>987</td>\n",
       "      <td>GCCAA</td>\n",
       "      <td>2</td>\n",
       "      <td>0.01992</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10052</td>\n",
       "      <td>10112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926</th>\n",
       "      <td>988</td>\n",
       "      <td>CCAAG</td>\n",
       "      <td>5</td>\n",
       "      <td>0.02589</td>\n",
       "      <td>0.01228</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9974</td>\n",
       "      <td>10052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>927 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ref_pos ref_kmer  num_events  dwell_time  NNNNN_dwell_time  \\\n",
       "0          1    TGGTT           1     0.01461           0.00000   \n",
       "1          2    GGTTA           1     0.00631           0.00000   \n",
       "..       ...      ...         ...         ...               ...   \n",
       "925      987    GCCAA           2     0.01992           0.00000   \n",
       "926      988    CCAAG           5     0.02589           0.01228   \n",
       "\n",
       "     mismatch_dwell_time  start_idx  end_idx  \n",
       "0                    0.0      45267    45311  \n",
       "1                    0.0      45248    45267  \n",
       "..                   ...        ...      ...  \n",
       "925                  0.0      10052    10112  \n",
       "926                  0.0       9974    10052  \n",
       "\n",
       "[927 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#15\tYLR110C\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ref_kmer</th>\n",
       "      <th>num_events</th>\n",
       "      <th>dwell_time</th>\n",
       "      <th>NNNNN_dwell_time</th>\n",
       "      <th>mismatch_dwell_time</th>\n",
       "      <th>start_idx</th>\n",
       "      <th>end_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14</td>\n",
       "      <td>TGTCG</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00598</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23697</td>\n",
       "      <td>23715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15</td>\n",
       "      <td>GTCGC</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00730</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23675</td>\n",
       "      <td>23697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>393</td>\n",
       "      <td>TTGTT</td>\n",
       "      <td>7</td>\n",
       "      <td>0.04217</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9871</td>\n",
       "      <td>9998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>394</td>\n",
       "      <td>TGTTG</td>\n",
       "      <td>5</td>\n",
       "      <td>0.01859</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9815</td>\n",
       "      <td>9871</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>364 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ref_pos ref_kmer  num_events  dwell_time  NNNNN_dwell_time  \\\n",
       "0         14    TGTCG           2     0.00598               0.0   \n",
       "1         15    GTCGC           1     0.00730               0.0   \n",
       "..       ...      ...         ...         ...               ...   \n",
       "362      393    TTGTT           7     0.04217               0.0   \n",
       "363      394    TGTTG           5     0.01859               0.0   \n",
       "\n",
       "     mismatch_dwell_time  start_idx  end_idx  \n",
       "0                    0.0      23697    23715  \n",
       "1                    0.0      23675    23697  \n",
       "..                   ...        ...      ...  \n",
       "362                  0.0       9871     9998  \n",
       "363                  0.0       9815     9871  \n",
       "\n",
       "[364 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "output_fn = \"./output/eventalign_collapse/stats_eventalign_collapse.tsv\"\n",
    "index_fn = \"./output/eventalign_collapse/stats_eventalign_collapse.tsv.idx\"\n",
    "\n",
    "# Import the index in a pandas dataframe (because it is simple)\n",
    "import pandas as pd \n",
    "index_df = pd.read_csv (index_fn, sep=\"\\t\")\n",
    "\n",
    "# Select random lines\n",
    "random_lines = index_df.sample(5)\n",
    "print (\"Random index lines\")\n",
    "display (random_lines)\n",
    "\n",
    "# Open the collapsed event align file\n",
    "with open (output_fn) as fp:\n",
    "    for id, read in random_lines.iterrows():\n",
    "        \n",
    "        # Access the header corresponding to the randomly selected index line using seek \n",
    "        fp.seek(0) # Return to file start\n",
    "        fp.seek(read.byte_offset) # Move to the offset indicated in the index file\n",
    "        print (fp.readline().rstrip()) # Print read header\n",
    "        df = pd.read_csv (fp, nrows=read.kmers, sep=\"\\t\") # Read lines corresponding to the read\n",
    "        with pd.option_context(\"display.max_rows\",4): # display first and last lines\n",
    "            display(df)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python3.6",
   "language": "python",
   "name": "python3.6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
