{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Freq_meth_calculate usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate methylation frequency at genomic CpG sites from the output of `nanopolish call-methylation`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Options"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **input_fn**\n",
    "\n",
    "Path to a `nanopolish call-methylation` tsv output file (read access required). In command line mode it is also possible to pipe the output of the `nanopolish call-methylation` directly into `Freq_meth_calculate`.\n",
    "\n",
    "* **outdir**\n",
    "\n",
    "Path to an existing directory where to write all the output files generated by `Eventalign_collapse`(write access required). If the directory does not exist an error is raised.\n",
    "\n",
    "* **outprefix**\n",
    "\n",
    "Prefix for all the files generated by the program\n",
    "\n",
    "Filtering options:\n",
    "\n",
    "* **min_llr**\n",
    "\n",
    "Set the minimal required log likelihood ratio between methylated and unmethylated state. Otherwise the value is considered ambiguous. \n",
    "\n",
    "* **min_depth**\n",
    "\n",
    "Set the minimal number of reads with non -ambiguous call per genomic position to be written in the output files.\n",
    "\n",
    "* **min_meth_freq**\n",
    "\n",
    "Set the minimal methylation frequency per genomic position to be written in the output files.\n",
    "\n",
    "* **split_group**\n",
    "\n",
    "In CpG rich regions it is not sometimes hard to distinguish which kmers is methylated in this case `nanopolish` outputs a group containing multiple CpG (or other motif). This option splits each CpG in individual genomic site. This can be useful for visualisation in a genome browser.\n",
    "\n",
    "* **motif**\n",
    "\n",
    "Specify the motif type , akin the one used for `Nanopolish call-methylation`. This is only useful if the split_group option is also selected. \n",
    "\n",
    "!!! note \"Methylation motif to sequence correspondence\"\n",
    "    * cpg → CG\n",
    "    * gpc → GC\n",
    "    * dam → GATC\n",
    "    * dcm → CCAGG\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output files format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Freq_meth_calculate` generates 2 files, a standard BED file and a tabulated file containing extra information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### BED file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Standard genomic BED6 (https://genome.ucsc.edu/FAQ/FAQformat.html#format1). The score correspond to the methylation frequency multiplied by 1000. The file is sorted by coordinates and can be rendered with a genome browser such as [IGV](https://software.broadinstitute.org/software/igv/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Tabulated TSV file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Contrary to the bed file, in the tabulated report, positions are ordered by decreasing methylation frequency.\n",
    "\n",
    "The file contains the following fields:\n",
    "\n",
    "* **chrom / start / end / strand**: Genomic coordinates of the motif or group of motifs in case split_group was not selected.\n",
    "* **site_id**: Unique integer identifier of the genomic position.\n",
    "* **methylated_reads / unmethylated_reads / ambiguous_reads**: Number of reads at a given genomic location with a higher likelyhood of being methylated or unmethylated or with an ambiguous methylation call.\n",
    "* **sequence**: -5 to +5 sequence of the motif or group of motifs in case split_group was not selected.\n",
    "* **num_motifs**: Number of motif in the group.\n",
    "* **meth_freq**: Methylation frequency (out of non anbiguous calls)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bash command line usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T09:27:30.923990Z",
     "start_time": "2019-05-01T09:27:29.879426Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: NanopolishComp Freq_meth_calculate [-h] [-i INPUT_FN] [-o OUTDIR]\n",
      "                                          [-p OUTPREFIX] [-l MIN_LLR]\n",
      "                                          [-d MIN_DEPTH] [-f MIN_METH_FREQ]\n",
      "                                          [-s] [-m {cpg,gpc,dam,dcm}]\n",
      "                                          [-v | -q]\n",
      "\n",
      "Calculate methylation frequency at genomic CpG sites from the output of\n",
      "nanopolish call-methylation\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -v, --verbose         Increase verbosity (default: False)\n",
      "  -q, --quiet           Reduce verbosity (default: False)\n",
      "\n",
      "Input/Output options:\n",
      "  -i INPUT_FN, --input_fn INPUT_FN\n",
      "                        Path to a nanopolish call_methylation tsv output file.\n",
      "                        If not specified read from std input\n",
      "  -o OUTDIR, --outdir OUTDIR\n",
      "                        Path to the output folder (default: ./)\n",
      "  -p OUTPREFIX, --outprefix OUTPREFIX\n",
      "                        text outprefix for all the files generated (default:\n",
      "                        out)\n",
      "\n",
      "Filtering options:\n",
      "  -l MIN_LLR, --min_llr MIN_LLR\n",
      "                        Log likelihood ratio threshold (default: 2.5)\n",
      "  -d MIN_DEPTH, --min_depth MIN_DEPTH\n",
      "                        Minimal number of reads covering a site to be reported\n",
      "                        (default: 10)\n",
      "  -f MIN_METH_FREQ, --min_meth_freq MIN_METH_FREQ\n",
      "                        Minimal methylation frequency of a site to be reported\n",
      "                        (default: 0.05)\n",
      "\n",
      "Motif options:\n",
      "  -s, --split_group     If True, multi motif groups (sequence with close\n",
      "                        motifs) are split in individual site (default: False)\n",
      "  -m {cpg,gpc,dam,dcm}, --motif {cpg,gpc,dam,dcm}\n",
      "                        Methylation motif type (default: cpg)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From an existing nanopolish call_methylation file output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T09:27:53.132056Z",
     "start_time": "2019-05-01T09:27:30.928030Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t216740\t000090\t-\n",
      "chr-XII\t451976\t451980\t216746\t000071\t-\n",
      "chr-XII\t452137\t452138\t216753\t000050\t-\n",
      "chr-XII\t452217\t452221\t216756\t000272\t-\n",
      "chr-XII\t452331\t452341\t216384\t000062\t+\n",
      "chr-XII\t452632\t452633\t216391\t000066\t+\n",
      "chr-XII\t452654\t452655\t216769\t000100\t-\n",
      "chr-XII\t453086\t453087\t216781\t000055\t-\n",
      "chr-XII\t453336\t453337\t216790\t000066\t-\n",
      "chrom\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tmeth_freq\n",
      "chr-XII\t465335\t465339\t-\t217103\t6\t5\t29\tCAGCACGACGGAGT\t2\t0.545455\n",
      "chr-XII\t456198\t456202\t-\t216869\t7\t7\t32\tCAGCACGACGGAGT\t2\t0.500000\n",
      "chr-XII\t460773\t460774\t+\t216597\t17\t20\t77\tCTCATCGTATA\t1\t0.459459\n",
      "chr-XII\t457104\t457105\t+\t216519\t6\t8\t43\tAACTACGAGCT\t1\t0.428571\n",
      "chr-XII\t461274\t461275\t+\t216614\t5\t7\t85\tTTATCCGAATG\t1\t0.416667\n",
      "chr-XII\t463406\t463407\t+\t216682\t6\t12\t29\tTCTTTCGCCCC\t1\t0.333333\n",
      "chr-XII\t457887\t457888\t-\t216914\t6\t12\t53\tGCTGCCGGAAA\t1\t0.333333\n",
      "chr-XII\t461077\t461078\t+\t216605\t8\t17\t74\tCTTTCCGCCAA\t1\t0.320000\n",
      "chr-XII\t462283\t462284\t-\t217018\t5\t11\t57\tAGGCACGCAAG\t1\t0.312500\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "\r",
      "0 lines [00:00, ? lines/s]\r",
      "2533 lines [00:00, 25325.86 lines/s]\r",
      "5298 lines [00:00, 25978.59 lines/s]\r",
      "8065 lines [00:00, 26462.17 lines/s]\r",
      "10793 lines [00:00, 26701.77 lines/s]\r",
      "13780 lines [00:00, 27578.39 lines/s]\r",
      "16900 lines [00:00, 28568.83 lines/s]\r",
      "19963 lines [00:00, 29156.55 lines/s]\r",
      "23322 lines [00:00, 30357.96 lines/s]\r",
      "26634 lines [00:00, 31135.90 lines/s]\r",
      "29681 lines [00:01, 30932.02 lines/s]\r",
      "32994 lines [00:01, 31558.33 lines/s]\r",
      "36114 lines [00:01, 29945.91 lines/s]\r",
      "39101 lines [00:01, 28730.37 lines/s]\r",
      "41979 lines [00:01, 28684.72 lines/s]\r",
      "45092 lines [00:01, 29376.60 lines/s]\r",
      "48037 lines [00:01, 29074.21 lines/s]\r",
      "50950 lines [00:01, 28265.11 lines/s]\r",
      "53786 lines [00:01, 27638.64 lines/s]\r",
      "56811 lines [00:01, 28372.09 lines/s]\r",
      "60062 lines [00:02, 29496.83 lines/s]\r",
      "63485 lines [00:02, 30771.80 lines/s]\r",
      "66896 lines [00:02, 31700.79 lines/s]\r",
      "70093 lines [00:02, 31759.23 lines/s]\r",
      "73288 lines [00:02, 30197.40 lines/s]\r",
      "76337 lines [00:02, 30020.83 lines/s]\r",
      "79360 lines [00:02, 29604.47 lines/s]\r",
      "82646 lines [00:02, 30509.73 lines/s]\r",
      "86010 lines [00:02, 31385.18 lines/s]\r",
      "89168 lines [00:02, 29989.32 lines/s]\r",
      "92194 lines [00:03, 29610.42 lines/s]\r",
      "95175 lines [00:03, 29260.45 lines/s]\r",
      "98116 lines [00:03, 29251.59 lines/s]\r",
      "101052 lines [00:03, 28430.63 lines/s]\r",
      "104365 lines [00:03, 29693.27 lines/s]\r",
      "107652 lines [00:03, 30575.38 lines/s]\r",
      "110854 lines [00:03, 30992.76 lines/s]\r",
      "113971 lines [00:03, 30175.00 lines/s]\r",
      "117114 lines [00:03, 30540.54 lines/s]\r",
      "120383 lines [00:04, 31153.39 lines/s]\r",
      "123852 lines [00:04, 32134.25 lines/s]\r",
      "127287 lines [00:04, 32765.92 lines/s]\r",
      "130598 lines [00:04, 32867.61 lines/s]\r",
      "133895 lines [00:04, 32475.57 lines/s]\r",
      "137197 lines [00:04, 32635.49 lines/s]\r",
      "140467 lines [00:04, 32106.55 lines/s]\r",
      "143767 lines [00:04, 32368.73 lines/s]\r",
      "147051 lines [00:04, 32506.55 lines/s]\r",
      "150306 lines [00:04, 31735.48 lines/s]\r",
      "153487 lines [00:05, 31745.84 lines/s]\r",
      "156774 lines [00:05, 32072.12 lines/s]\r",
      "159986 lines [00:05, 27987.96 lines/s]\r",
      "162940 lines [00:05, 28435.55 lines/s]\r",
      "165944 lines [00:05, 28897.89 lines/s]\r",
      "169259 lines [00:05, 30053.65 lines/s]\r",
      "172312 lines [00:05, 27608.37 lines/s]\r",
      "175253 lines [00:05, 28124.61 lines/s]\r",
      "178585 lines [00:05, 29504.52 lines/s]\r",
      "181911 lines [00:06, 30538.54 lines/s]\r",
      "185080 lines [00:06, 30873.12 lines/s]\r",
      "188202 lines [00:06, 28941.36 lines/s]\r",
      "191145 lines [00:06, 28428.21 lines/s]\r",
      "194024 lines [00:06, 27746.17 lines/s]\r",
      "197032 lines [00:06, 28406.07 lines/s]\r",
      "200175 lines [00:06, 29249.64 lines/s]\r",
      "203564 lines [00:06, 30495.94 lines/s]\r",
      "206857 lines [00:06, 31185.92 lines/s]\r",
      "210267 lines [00:06, 31995.01 lines/s]\r",
      "213743 lines [00:07, 32773.95 lines/s]\r",
      "217136 lines [00:07, 33107.08 lines/s]\r",
      "220462 lines [00:07, 32706.78 lines/s]\r",
      "223907 lines [00:07, 33209.48 lines/s]\r",
      "227239 lines [00:07, 33171.70 lines/s]\r",
      "230564 lines [00:07, 32994.26 lines/s]\r",
      "233869 lines [00:07, 32677.95 lines/s]\r",
      "237142 lines [00:07, 32675.72 lines/s]\r",
      "240413 lines [00:07, 30519.78 lines/s]\r",
      "243496 lines [00:08, 28493.23 lines/s]\r",
      "246395 lines [00:08, 27888.13 lines/s]\r",
      "249646 lines [00:08, 29128.62 lines/s]\r",
      "252867 lines [00:08, 29989.07 lines/s]\r",
      "255901 lines [00:08, 25671.72 lines/s]\r",
      "258601 lines [00:08, 24969.15 lines/s]\r",
      "261196 lines [00:08, 24261.54 lines/s]\r",
      "264037 lines [00:08, 25372.94 lines/s]\r",
      "267030 lines [00:08, 26586.81 lines/s]\r",
      "270152 lines [00:09, 27824.25 lines/s]\r",
      "273379 lines [00:09, 29022.58 lines/s]\r",
      "276645 lines [00:09, 30025.63 lines/s]\r",
      "279694 lines [00:09, 28699.40 lines/s]\r",
      "282948 lines [00:09, 29748.12 lines/s]\r",
      "285964 lines [00:09, 28897.95 lines/s]\r",
      "289262 lines [00:09, 30011.40 lines/s]\r",
      "292517 lines [00:09, 30728.92 lines/s]\r",
      "295619 lines [00:09, 29287.57 lines/s]\r",
      "298583 lines [00:09, 27673.31 lines/s]\r",
      "301393 lines [00:10, 26535.78 lines/s]\r",
      "304087 lines [00:10, 25985.28 lines/s]\r",
      "306716 lines [00:10, 25946.97 lines/s]\r",
      "309332 lines [00:10, 24599.48 lines/s]\r",
      "311822 lines [00:10, 24526.38 lines/s]\r",
      "314325 lines [00:10, 24675.27 lines/s]\r",
      "316808 lines [00:10, 24274.63 lines/s]\r",
      "319347 lines [00:10, 24597.64 lines/s]\r",
      "322061 lines [00:10, 25307.31 lines/s]\r",
      "324885 lines [00:11, 26119.97 lines/s]\r",
      "327746 lines [00:11, 26819.41 lines/s]\r",
      "330443 lines [00:11, 24233.16 lines/s]\r",
      "333700 lines [00:11, 26248.10 lines/s]\r",
      "336893 lines [00:11, 27725.07 lines/s]\r",
      "340116 lines [00:11, 28937.55 lines/s]\r",
      "343173 lines [00:11, 29405.25 lines/s]\r",
      "346292 lines [00:11, 29917.45 lines/s]\r",
      "349645 lines [00:11, 30914.94 lines/s]\r",
      "352888 lines [00:11, 31353.18 lines/s]\r",
      "356050 lines [00:12, 31009.63 lines/s]\r",
      "359303 lines [00:12, 31448.76 lines/s]\r",
      "362464 lines [00:12, 31407.19 lines/s]\r",
      "365616 lines [00:12, 30345.78 lines/s]\r",
      "368666 lines [00:12, 29654.75 lines/s]\r",
      "371646 lines [00:12, 29640.67 lines/s]\r",
      "375153 lines [00:12, 31083.86 lines/s]\r",
      "378812 lines [00:12, 32553.15 lines/s]\r",
      "382514 lines [00:12, 33773.95 lines/s]\r",
      "386185 lines [00:13, 34602.24 lines/s]\r",
      "389907 lines [00:13, 35345.82 lines/s]\r",
      "393483 lines [00:13, 35466.74 lines/s]\r",
      "397213 lines [00:13, 35997.05 lines/s]\r",
      "400924 lines [00:13, 36321.94 lines/s]\r",
      "404568 lines [00:13, 34961.53 lines/s]\r",
      "408083 lines [00:13, 34136.37 lines/s]\r",
      "411514 lines [00:13, 34061.56 lines/s]\r",
      "414933 lines [00:13, 34079.99 lines/s]\r",
      "418350 lines [00:13, 33800.01 lines/s]\r",
      "421787 lines [00:14, 33968.49 lines/s]\r",
      "425189 lines [00:14, 33810.80 lines/s]\r",
      "428574 lines [00:14, 33649.91 lines/s]\r",
      "431942 lines [00:14, 31512.31 lines/s]\r",
      "435123 lines [00:14, 29187.32 lines/s]\r",
      "438097 lines [00:14, 28114.61 lines/s]\r",
      "440956 lines [00:14, 27119.62 lines/s]\r",
      "443709 lines [00:14, 26174.29 lines/s]\r",
      "446362 lines [00:14, 24951.57 lines/s]\r",
      "449290 lines [00:15, 26108.01 lines/s]\r",
      "452139 lines [00:15, 26775.61 lines/s]\r",
      "454849 lines [00:15, 25546.85 lines/s]\r",
      "457495 lines [00:15, 25813.97 lines/s]\r",
      "460102 lines [00:15, 25198.53 lines/s]\r",
      "462757 lines [00:15, 25584.05 lines/s]\r",
      "465348 lines [00:15, 25679.03 lines/s]\r",
      "468121 lines [00:15, 26261.02 lines/s]\r",
      "471013 lines [00:15, 27003.33 lines/s]\r",
      "473727 lines [00:16, 21839.12 lines/s]\r",
      "476195 lines [00:16, 22619.19 lines/s]\r",
      "478582 lines [00:16, 21881.69 lines/s]\r",
      "480863 lines [00:16, 21528.59 lines/s]\r",
      "483083 lines [00:16, 21645.61 lines/s]\r",
      "485977 lines [00:16, 23415.08 lines/s]\r",
      "489047 lines [00:16, 25207.32 lines/s]\r",
      "492284 lines [00:16, 26998.49 lines/s]\r",
      "495423 lines [00:16, 28180.65 lines/s]\r",
      "498348 lines [00:16, 28492.25 lines/s]\r",
      "501517 lines [00:17, 29381.08 lines/s]\r",
      "504844 lines [00:17, 30447.58 lines/s]\r",
      "508179 lines [00:17, 31262.06 lines/s]\r",
      "511342 lines [00:17, 31053.03 lines/s]\r",
      "514571 lines [00:17, 31413.82 lines/s]\r",
      "517814 lines [00:17, 31710.78 lines/s]\r",
      "520999 lines [00:17, 31512.38 lines/s]\r",
      "524161 lines [00:17, 31376.57 lines/s]\r",
      "527513 lines [00:17, 31988.88 lines/s]\r",
      "530720 lines [00:17, 31876.88 lines/s]\r",
      "534108 lines [00:18, 32451.26 lines/s]\r",
      "537437 lines [00:18, 32698.35 lines/s]\r",
      "540712 lines [00:18, 30900.64 lines/s]\r",
      "543974 lines [00:18, 31395.58 lines/s]\r",
      "547224 lines [00:18, 31716.03 lines/s]\r",
      "550506 lines [00:18, 32037.25 lines/s]\r",
      "553823 lines [00:18, 32368.15 lines/s]\r",
      "557298 lines [00:18, 33047.29 lines/s]\r",
      "560669 lines [00:18, 33241.32 lines/s]\r",
      "564025 lines [00:19, 33334.90 lines/s]\r",
      "567364 lines [00:19, 33157.31 lines/s]\r",
      "570727 lines [00:19, 33297.52 lines/s]\r",
      "574181 lines [00:19, 33659.14 lines/s]\r",
      "577550 lines [00:19, 33491.68 lines/s]\r",
      "580902 lines [00:19, 33047.39 lines/s]\r",
      "584210 lines [00:19, 31939.18 lines/s]\r",
      "587414 lines [00:19, 31939.37 lines/s]\r",
      "590767 lines [00:19, 32400.22 lines/s]\r",
      "594014 lines [00:20, 25522.69 lines/s]\r",
      "597094 lines [00:20, 26903.91 lines/s]\r",
      "600191 lines [00:20, 28002.33 lines/s]\r",
      "603478 lines [00:20, 29304.08 lines/s]\r",
      "605248 lines [00:20, 29707.87 lines/s]\n",
      "Read sites summary\n",
      "\ttotal: 605,248\n",
      "\tunmethylated: 596,963\n",
      "\tmethylated: 8,285\n",
      "\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "\r",
      "  0%|          | 0/340081 [00:00<?, ? positions/s]\r",
      " 18%|█▊        | 60237/340081 [00:00<00:00, 602367.70 positions/s]\r",
      " 35%|███▌      | 119858/340081 [00:00<00:00, 600506.39 positions/s]\r",
      " 52%|█████▏    | 176620/340081 [00:00<00:00, 590246.04 positions/s]\r",
      " 69%|██████▉   | 233871/340081 [00:00<00:00, 584809.37 positions/s]\r",
      " 85%|████████▌ | 289988/340081 [00:00<00:00, 577508.67 positions/s]\r",
      "100%|██████████| 340081/340081 [00:00<00:00, 578600.08 positions/s]\n",
      "Genomic positions summary\n",
      "\ttotal: 340,081\n",
      "\tlow_coverage: 339,292\n",
      "\tlow_meth_freq: 647\n",
      "\tvalid: 142\n",
      "\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --verbose -i data/freq_meth_calculate/methylation_calls.tsv -o ./output/freq_meth_calculate/\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From standard input to a file (the same can be done with Nanopolish call-methylation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example the split_group option was also selected leading to motif groups to be split in individual positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T09:28:14.667714Z",
     "start_time": "2019-05-01T09:27:53.135510Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t289557\t000090\t-\n",
      "chr-XII\t451976\t451977\t289563\t000071\t-\n",
      "chr-XII\t451979\t451980\t289564\t000071\t-\n",
      "chr-XII\t452137\t452138\t289574\t000050\t-\n",
      "chr-XII\t452217\t452218\t289577\t000272\t-\n",
      "chr-XII\t452220\t452221\t289578\t000272\t-\n",
      "chr-XII\t452331\t452332\t288999\t000062\t+\n",
      "chr-XII\t452340\t452341\t289000\t000062\t+\n",
      "chr-XII\t452632\t452633\t289008\t000066\t+\n",
      "chrom\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tmeth_freq\n",
      "chr-XII\t465335\t465336\t-\t290122\t6\t5\t29\tCAGCACGACGG\t2\t0.545455\n",
      "chr-XII\t465338\t465339\t-\t290123\t6\t5\t29\tCACGACGGAGT\t2\t0.545455\n",
      "chr-XII\t456198\t456199\t-\t289760\t7\t7\t32\tCAGCACGACGG\t2\t0.500000\n",
      "chr-XII\t456201\t456202\t-\t289761\t7\t7\t32\tCACGACGGAGT\t2\t0.500000\n",
      "chr-XII\t460773\t460774\t+\t289338\t17\t20\t77\tCTCATCGTATA\t1\t0.459459\n",
      "chr-XII\t457104\t457105\t+\t289216\t6\t8\t43\tAACTACGAGCT\t1\t0.428571\n",
      "chr-XII\t461274\t461275\t+\t289359\t5\t7\t85\tTTATCCGAATG\t1\t0.416667\n",
      "chr-XII\t463406\t463407\t+\t289460\t6\t12\t29\tTCTTTCGCCCC\t1\t0.333333\n",
      "chr-XII\t457887\t457888\t-\t289825\t6\t12\t53\tGCTGCCGGAAA\t1\t0.333333\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "\r",
      "0 lines [00:00, ? lines/s]\r",
      "3300 lines [00:00, 32998.69 lines/s]\r",
      "6449 lines [00:00, 32529.45 lines/s]\r",
      "9769 lines [00:00, 32726.01 lines/s]\r",
      "12995 lines [00:00, 32581.51 lines/s]\r",
      "16288 lines [00:00, 32683.71 lines/s]\r",
      "19436 lines [00:00, 32311.56 lines/s]\r",
      "22373 lines [00:00, 31365.78 lines/s]\r",
      "25676 lines [00:00, 31845.90 lines/s]\r",
      "28793 lines [00:00, 31637.99 lines/s]\r",
      "31952 lines [00:01, 31620.78 lines/s]\r",
      "35102 lines [00:01, 31584.10 lines/s]\r",
      "38245 lines [00:01, 31534.75 lines/s]\r",
      "41355 lines [00:01, 31147.20 lines/s]\r",
      "44441 lines [00:01, 31010.44 lines/s]\r",
      "47522 lines [00:01, 30043.16 lines/s]\r",
      "50565 lines [00:01, 30156.45 lines/s]\r",
      "53576 lines [00:01, 30137.03 lines/s]\r",
      "56801 lines [00:01, 30739.24 lines/s]\r",
      "59971 lines [00:01, 31020.09 lines/s]\r",
      "63074 lines [00:02, 30794.76 lines/s]\r",
      "66223 lines [00:02, 30997.89 lines/s]\r",
      "69324 lines [00:02, 30888.77 lines/s]\r",
      "72414 lines [00:02, 30290.65 lines/s]\r",
      "75447 lines [00:02, 28803.43 lines/s]\r",
      "78345 lines [00:02, 28786.58 lines/s]\r",
      "81579 lines [00:02, 29766.63 lines/s]\r",
      "84573 lines [00:02, 29798.48 lines/s]\r",
      "87856 lines [00:02, 30647.06 lines/s]\r",
      "91077 lines [00:02, 31098.80 lines/s]\r",
      "94327 lines [00:03, 31505.94 lines/s]\r",
      "97488 lines [00:03, 31507.13 lines/s]\r",
      "100646 lines [00:03, 31470.51 lines/s]\r",
      "103798 lines [00:03, 31366.78 lines/s]\r",
      "106939 lines [00:03, 30289.68 lines/s]\r",
      "110108 lines [00:03, 30695.09 lines/s]\r",
      "113311 lines [00:03, 31082.16 lines/s]\r",
      "116427 lines [00:03, 30629.60 lines/s]\r",
      "119619 lines [00:03, 31003.87 lines/s]\r",
      "122764 lines [00:03, 31135.28 lines/s]\r",
      "125882 lines [00:04, 30892.26 lines/s]\r",
      "129022 lines [00:04, 31040.80 lines/s]\r",
      "132129 lines [00:04, 27645.66 lines/s]\r",
      "135242 lines [00:04, 28604.72 lines/s]\r",
      "138376 lines [00:04, 29372.08 lines/s]\r",
      "141501 lines [00:04, 29903.06 lines/s]\r",
      "144617 lines [00:04, 30267.31 lines/s]\r",
      "147771 lines [00:04, 30635.40 lines/s]\r",
      "151027 lines [00:04, 31188.45 lines/s]\r",
      "154306 lines [00:05, 31651.77 lines/s]\r",
      "157484 lines [00:05, 31204.44 lines/s]\r",
      "160615 lines [00:05, 30769.87 lines/s]\r",
      "163701 lines [00:05, 30596.70 lines/s]\r",
      "166786 lines [00:05, 30671.19 lines/s]\r",
      "169971 lines [00:05, 31007.93 lines/s]\r",
      "173076 lines [00:05, 30903.97 lines/s]\r",
      "176170 lines [00:05, 30769.97 lines/s]\r",
      "179326 lines [00:05, 31001.13 lines/s]\r",
      "182429 lines [00:05, 30568.34 lines/s]\r",
      "185489 lines [00:06, 27546.89 lines/s]\r",
      "188556 lines [00:06, 28414.88 lines/s]\r",
      "191663 lines [00:06, 29161.41 lines/s]\r",
      "194813 lines [00:06, 29825.26 lines/s]\r",
      "197934 lines [00:06, 30227.44 lines/s]\r",
      "201084 lines [00:06, 30597.74 lines/s]\r",
      "204239 lines [00:06, 30875.74 lines/s]\r",
      "207400 lines [00:06, 31091.65 lines/s]\r",
      "210519 lines [00:06, 29200.45 lines/s]\r",
      "213755 lines [00:06, 30079.23 lines/s]\r",
      "216938 lines [00:07, 30580.71 lines/s]\r",
      "220075 lines [00:07, 30811.57 lines/s]\r",
      "223230 lines [00:07, 31029.41 lines/s]\r",
      "226448 lines [00:07, 31364.55 lines/s]\r",
      "229786 lines [00:07, 31940.68 lines/s]\r",
      "232989 lines [00:07, 31943.17 lines/s]\r",
      "236257 lines [00:07, 32158.59 lines/s]\r",
      "239493 lines [00:07, 32218.57 lines/s]\r",
      "242719 lines [00:07, 31885.63 lines/s]\r",
      "245920 lines [00:07, 31921.99 lines/s]\r",
      "249187 lines [00:08, 32136.08 lines/s]\r",
      "252403 lines [00:08, 27737.48 lines/s]\r",
      "255575 lines [00:08, 28820.25 lines/s]\r",
      "258706 lines [00:08, 29522.54 lines/s]\r",
      "261803 lines [00:08, 29941.20 lines/s]\r",
      "264942 lines [00:08, 30361.23 lines/s]\r",
      "268119 lines [00:08, 30768.08 lines/s]\r",
      "271244 lines [00:08, 30909.11 lines/s]\r",
      "274353 lines [00:08, 30426.54 lines/s]\r",
      "277567 lines [00:09, 30920.92 lines/s]\r",
      "280671 lines [00:09, 30764.71 lines/s]\r",
      "283843 lines [00:09, 31044.12 lines/s]\r",
      "287075 lines [00:09, 31414.11 lines/s]\r",
      "290223 lines [00:09, 31294.26 lines/s]\r",
      "293388 lines [00:09, 31397.63 lines/s]\r",
      "296632 lines [00:09, 31702.58 lines/s]\r",
      "299903 lines [00:09, 31996.76 lines/s]\r",
      "303106 lines [00:09, 31732.53 lines/s]\r",
      "306282 lines [00:09, 31647.77 lines/s]\r",
      "309450 lines [00:10, 31656.45 lines/s]\r",
      "312617 lines [00:10, 31636.95 lines/s]\r",
      "315794 lines [00:10, 31675.13 lines/s]\r",
      "318963 lines [00:10, 26839.36 lines/s]\r",
      "322002 lines [00:10, 27813.62 lines/s]\r",
      "325107 lines [00:10, 28711.34 lines/s]\r",
      "328252 lines [00:10, 29480.81 lines/s]\r",
      "331359 lines [00:10, 29938.44 lines/s]\r",
      "334403 lines [00:10, 30085.20 lines/s]\r",
      "337685 lines [00:11, 30855.44 lines/s]\r",
      "340797 lines [00:11, 30925.40 lines/s]\r",
      "343999 lines [00:11, 31244.41 lines/s]\r",
      "347173 lines [00:11, 31390.17 lines/s]\r",
      "350332 lines [00:11, 31445.68 lines/s]\r",
      "353540 lines [00:11, 31632.78 lines/s]\r",
      "356709 lines [00:11, 30457.73 lines/s]\r",
      "360048 lines [00:11, 31280.12 lines/s]\r",
      "363265 lines [00:11, 31540.46 lines/s]\r",
      "366797 lines [00:11, 32585.25 lines/s]\r",
      "370328 lines [00:12, 33355.45 lines/s]\r",
      "373851 lines [00:12, 33895.93 lines/s]\r",
      "377329 lines [00:12, 34154.80 lines/s]\r",
      "380805 lines [00:12, 34332.88 lines/s]\r",
      "384246 lines [00:12, 34288.61 lines/s]\r",
      "387680 lines [00:12, 34136.45 lines/s]\r",
      "391166 lines [00:12, 34349.53 lines/s]\r",
      "394624 lines [00:12, 34417.06 lines/s]\r",
      "398218 lines [00:12, 34858.56 lines/s]\r",
      "401707 lines [00:12, 34813.89 lines/s]\r",
      "405191 lines [00:13, 33719.58 lines/s]\r",
      "408572 lines [00:13, 33089.89 lines/s]\r",
      "411890 lines [00:13, 32477.76 lines/s]\r",
      "415147 lines [00:13, 32331.21 lines/s]\r",
      "418387 lines [00:13, 31940.46 lines/s]\r",
      "421611 lines [00:13, 32026.97 lines/s]\r",
      "424818 lines [00:13, 31720.43 lines/s]\r",
      "427994 lines [00:13, 31544.92 lines/s]\r",
      "431186 lines [00:13, 31655.22 lines/s]\r",
      "434354 lines [00:13, 31508.12 lines/s]\r",
      "437549 lines [00:14, 31639.03 lines/s]\r",
      "440715 lines [00:14, 31088.83 lines/s]\r",
      "443845 lines [00:14, 31149.38 lines/s]\r",
      "446963 lines [00:14, 31143.61 lines/s]\r",
      "450079 lines [00:14, 31098.55 lines/s]\r",
      "453234 lines [00:14, 31230.73 lines/s]\r",
      "456374 lines [00:14, 31279.43 lines/s]\r",
      "459503 lines [00:14, 25429.34 lines/s]\r",
      "462699 lines [00:14, 27088.33 lines/s]\r",
      "465943 lines [00:15, 28497.08 lines/s]\r",
      "469128 lines [00:15, 29424.77 lines/s]\r",
      "472168 lines [00:15, 29561.70 lines/s]\r",
      "475193 lines [00:15, 25954.31 lines/s]\r",
      "478215 lines [00:15, 27101.17 lines/s]\r",
      "481280 lines [00:15, 28076.01 lines/s]\r",
      "484464 lines [00:15, 29107.63 lines/s]\r",
      "487645 lines [00:15, 29865.91 lines/s]\r",
      "490859 lines [00:15, 30511.51 lines/s]\r",
      "493950 lines [00:16, 30437.74 lines/s]\r",
      "497056 lines [00:16, 30615.75 lines/s]\r",
      "500195 lines [00:16, 30842.28 lines/s]\r",
      "503308 lines [00:16, 30925.50 lines/s]\r",
      "506502 lines [00:16, 31223.00 lines/s]\r",
      "509632 lines [00:16, 30915.83 lines/s]\r",
      "512762 lines [00:16, 31027.99 lines/s]\r",
      "515870 lines [00:16, 30862.17 lines/s]\r",
      "519011 lines [00:16, 31023.17 lines/s]\r",
      "522150 lines [00:16, 31131.66 lines/s]\r",
      "525375 lines [00:17, 31451.99 lines/s]\r",
      "528523 lines [00:17, 31204.35 lines/s]\r",
      "531781 lines [00:17, 31601.48 lines/s]\r",
      "534944 lines [00:17, 30757.18 lines/s]\r",
      "538027 lines [00:17, 30402.60 lines/s]\r",
      "541127 lines [00:17, 30578.88 lines/s]\r",
      "544245 lines [00:17, 30754.82 lines/s]\r",
      "547416 lines [00:17, 31033.22 lines/s]\r",
      "550617 lines [00:17, 31319.55 lines/s]\r",
      "553779 lines [00:17, 31408.83 lines/s]\r",
      "557114 lines [00:18, 31964.78 lines/s]\r",
      "560315 lines [00:18, 31565.43 lines/s]\r",
      "563525 lines [00:18, 31723.37 lines/s]\r",
      "566707 lines [00:18, 31750.72 lines/s]\r",
      "569885 lines [00:18, 24325.86 lines/s]\r",
      "573099 lines [00:18, 26238.44 lines/s]\r",
      "576111 lines [00:18, 27292.00 lines/s]\r",
      "579181 lines [00:18, 28229.71 lines/s]\r",
      "582334 lines [00:18, 29143.14 lines/s]\r",
      "585360 lines [00:19, 29467.63 lines/s]\r",
      "588428 lines [00:19, 29820.94 lines/s]\r",
      "591513 lines [00:19, 30122.26 lines/s]\r",
      "594615 lines [00:19, 30378.89 lines/s]\r",
      "597678 lines [00:19, 30326.86 lines/s]\r",
      "600906 lines [00:19, 30886.12 lines/s]\r",
      "604175 lines [00:19, 31405.19 lines/s]\r",
      "605248 lines [00:19, 30767.93 lines/s]\n",
      "Read sites summary\n",
      "\ttotal: 605,248\n",
      "\tunmethylated: 596,963\n",
      "\tmethylated: 8,285\n",
      "\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "\r",
      "  0%|          | 0/455542 [00:00<?, ? positions/s]\r",
      " 13%|█▎        | 59241/455542 [00:00<00:00, 592407.74 positions/s]\r",
      " 26%|██▌       | 118712/455542 [00:00<00:00, 593095.02 positions/s]\r",
      " 40%|███▉      | 180963/455542 [00:00<00:00, 601623.18 positions/s]\r",
      " 53%|█████▎    | 241443/455542 [00:00<00:00, 602571.17 positions/s]\r",
      " 66%|██████▌   | 301702/455542 [00:00<00:00, 602573.98 positions/s]\r",
      " 79%|███████▉  | 360886/455542 [00:00<00:00, 599312.43 positions/s]\r",
      " 92%|█████████▏| 418433/455542 [00:00<00:00, 591954.54 positions/s]\r",
      "100%|██████████| 455542/455542 [00:00<00:00, 594013.62 positions/s]\n",
      "Genomic positions summary\n",
      "\ttotal: 455,542\n",
      "\tlow_coverage: 454,219\n",
      "\tlow_meth_freq: 1,148\n",
      "\tvalid: 175\n",
      "\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "cat data/freq_meth_calculate/methylation_calls.tsv | NanopolishComp Freq_meth_calculate --verbose -o ./output/freq_meth_calculate/ --split_group\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import the package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T09:28:14.680678Z",
     "start_time": "2019-05-01T09:28:14.670868Z"
    }
   },
   "outputs": [],
   "source": [
    "# Import main program\n",
    "from NanopolishComp.Freq_meth_calculate import Freq_meth_calculate\n",
    "\n",
    "# Import helper functions\n",
    "from NanopolishComp.common import jhelp, head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### python API help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T09:28:14.696867Z",
     "start_time": "2019-05-01T09:28:14.683304Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "**NanopolishComp.Freq_meth_calculate.__init__**\n",
       "\n",
       "Calculate methylation frequency at genomic CpG sites from the output of nanopolish call-methylation\n",
       "\n",
       "---\n",
       "\n",
       "* **input_fn** *: str (required)*\n",
       "\n",
       "Path to a nanopolish call_methylation tsv output file\n",
       "\n",
       "* **outdir** *: str (default = ./)*\n",
       "\n",
       "Path to the output folder\n",
       "\n",
       "* **outprefix** *: str (default = out)*\n",
       "\n",
       "text outprefix for all the files generated\n",
       "\n",
       "* **min_llr** *: float (default = 2.5)*\n",
       "\n",
       "Log likelihood ratio threshold\n",
       "\n",
       "* **min_depth** *: int (default = 10)*\n",
       "\n",
       "Minimal number of reads covering a site to be reported\n",
       "\n",
       "* **min_meth_freq** *: float (default = 0.05)*\n",
       "\n",
       "Minimal methylation frequency of a site to be reported\n",
       "\n",
       "* **split_group** *: bool (default = False)*\n",
       "\n",
       "If True, multi motif groups (sequence with close motifs) are split in individual site\n",
       "\n",
       "* **motif** *: {cpg,gpc,dam,dcm} (default = cpg)*\n",
       "\n",
       "Methylation motif type\n",
       "\n",
       "* **verbose** *: bool (default = False)*\n",
       "\n",
       "Increase verbosity\n",
       "\n",
       "* **quiet** *: bool (default = False)*\n",
       "\n",
       "Reduce verbosity\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(Freq_meth_calculate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-23T09:19:41.284517Z",
     "start_time": "2019-04-23T09:19:41.165804Z"
    },
    "scrolled": false
   },
   "source": [
    "#### basic setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T09:28:36.301067Z",
     "start_time": "2019-05-01T09:28:14.699914Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:20, 29019.64 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 340081/340081 [00:00<00:00, 536274.02 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read sites summary\n",
      "\ttotal: 605,248\n",
      "\tunmethylated: 596,963\n",
      "\tmethylated: 8,285\n",
      "Genomic positions summary\n",
      "\ttotal: 340,081\n",
      "\tlow_coverage: 339,292\n",
      "\tlow_meth_freq: 647\n",
      "\tvalid: 142\n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence       num_motifs meth_freq \n",
      "chr-XII 465335 465339 -      217103  6                5                  29              CAGCACGACGGAGT 2          0.545455  \n",
      "chr-XII 456198 456202 -      216869  7                7                  32              CAGCACGACGGAGT 2          0.500000  \n",
      "chr-XII 460773 460774 +      216597  17               20                 77              CTCATCGTATA    1          0.459459  \n",
      "chr-XII 457104 457105 +      216519  6                8                  43              AACTACGAGCT    1          0.428571  \n",
      "chr-XII 461274 461275 +      216614  5                7                  85              TTATCCGAATG    1          0.416667  \n",
      "chr-XII 463406 463407 +      216682  6                12                 29              TCTTTCGCCCC    1          0.333333  \n",
      "chr-XII 457887 457888 -      216914  6                12                 53              GCTGCCGGAAA    1          0.333333  \n",
      "chr-XII 461077 461078 +      216605  8                17                 74              CTTTCCGCCAA    1          0.320000  \n",
      "chr-XII 462283 462284 -      217018  5                11                 57              AGGCACGCAAG    1          0.312500  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t216740\t000090\t-\n",
      "chr-XII\t451976\t451980\t216746\t000071\t-\n",
      "chr-XII\t452137\t452138\t216753\t000050\t-\n",
      "chr-XII\t452217\t452221\t216756\t000272\t-\n",
      "chr-XII\t452331\t452341\t216384\t000062\t+\n",
      "chr-XII\t452632\t452633\t216391\t000066\t+\n",
      "chr-XII\t452654\t452655\t216769\t000100\t-\n",
      "chr-XII\t453086\t453087\t216781\t000055\t-\n",
      "chr-XII\t453336\t453337\t216790\t000066\t-\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\")\n",
    "\n",
    "print(f)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### split motif groups in individual genomic position "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T09:28:59.684600Z",
     "start_time": "2019-05-01T09:28:36.304167Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:22, 27011.98 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 455542/455542 [00:00<00:00, 534742.10 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read sites summary\n",
      "\ttotal: 605,248\n",
      "\tunmethylated: 596,963\n",
      "\tmethylated: 8,285\n",
      "Genomic positions summary\n",
      "\ttotal: 455,542\n",
      "\tlow_coverage: 454,219\n",
      "\tlow_meth_freq: 1,148\n",
      "\tvalid: 175\n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence    num_motifs meth_freq \n",
      "chr-XII 465335 465336 -      630203  6                5                  29              CAGCACGACGG 2          0.545455  \n",
      "chr-XII 465338 465339 -      630204  6                5                  29              CACGACGGAGT 2          0.545455  \n",
      "chr-XII 456198 456199 -      629841  7                7                  32              CAGCACGACGG 2          0.500000  \n",
      "chr-XII 456201 456202 -      629842  7                7                  32              CACGACGGAGT 2          0.500000  \n",
      "chr-XII 460773 460774 +      629419  17               20                 77              CTCATCGTATA 1          0.459459  \n",
      "chr-XII 457104 457105 +      629297  6                8                  43              AACTACGAGCT 1          0.428571  \n",
      "chr-XII 461274 461275 +      629440  5                7                  85              TTATCCGAATG 1          0.416667  \n",
      "chr-XII 463406 463407 +      629541  6                12                 29              TCTTTCGCCCC 1          0.333333  \n",
      "chr-XII 457887 457888 -      629906  6                12                 53              GCTGCCGGAAA 1          0.333333  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t629638\t000090\t-\n",
      "chr-XII\t451976\t451977\t629644\t000071\t-\n",
      "chr-XII\t451979\t451980\t629645\t000071\t-\n",
      "chr-XII\t452137\t452138\t629655\t000050\t-\n",
      "chr-XII\t452217\t452218\t629658\t000272\t-\n",
      "chr-XII\t452220\t452221\t629659\t000272\t-\n",
      "chr-XII\t452331\t452332\t629080\t000062\t+\n",
      "chr-XII\t452340\t452341\t629081\t000062\t+\n",
      "chr-XII\t452632\t452633\t629089\t000066\t+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\",\n",
    "    split_group=True)\n",
    "\n",
    "print(f)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Changing filtering threshold (not recommended)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T09:29:19.706841Z",
     "start_time": "2019-05-01T09:28:59.688176Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:19, 31412.70 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 340081/340081 [00:00<00:00, 523826.87 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read sites summary\n",
      "\ttotal: 605,248\n",
      "\tunmethylated: 571,328\n",
      "\tmethylated: 33,920\n",
      "Genomic positions summary\n",
      "\ttotal: 340,081\n",
      "\tlow_coverage: 339,355\n",
      "\tlow_meth_freq: 698\n",
      "\tvalid: 28\n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence       num_motifs meth_freq \n",
      "chr-XII 461541 461542 +      1012246 26               3                  52              AATTCCGAGGG    1          0.896552  \n",
      "chr-XII 455556 455557 +      1012101 16               6                  20              AGATCCGTTGT    1          0.727273  \n",
      "chr-XII 462606 462607 +      1012278 21               9                  28              AATTCCGGGGT    1          0.700000  \n",
      "chr-XII 458208 458209 -      1012544 25               15                 37              TTAAACGCAAA    1          0.625000  \n",
      "chr-XII 463513 463514 -      1012680 13               8                  29              TCTTTCGGGTC    1          0.619048  \n",
      "chr-XII 461032 461033 -      1012599 27               20                 60              CCCCACGCAAA    1          0.574468  \n",
      "chr-XII 456198 456202 -      1012492 17               13                 16              CAGCACGACGGAGT 2          0.566667  \n",
      "chr-XII 461274 461275 +      1012237 21               21                 55              TTATCCGAATG    1          0.500000  \n",
      "chr-XII 465464 465465 -      1012730 10               10                 19              TCCATCGGCTT    1          0.500000  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t455151\t455152\t1012091\t000409\t+\n",
      "chr-XII\t455556\t455557\t1012101\t000727\t+\n",
      "chr-XII\t456198\t456202\t1012492\t000566\t-\n",
      "chr-XII\t456451\t456452\t1012499\t000303\t-\n",
      "chr-XII\t456686\t456687\t1012128\t000321\t+\n",
      "chr-XII\t457045\t457046\t1012140\t000476\t+\n",
      "chr-XII\t457104\t457105\t1012142\t000447\t+\n",
      "chr-XII\t457887\t457888\t1012537\t000361\t-\n",
      "chr-XII\t458208\t458209\t1012544\t000625\t-\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\",\n",
    "    min_llr=1,\n",
    "    min_depth=20,\n",
    "    min_meth_freq=0.3)\n",
    "\n",
    "print(f)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python3.6",
   "language": "python",
   "name": "python3.6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
