{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Freq_meth_calculate usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate methylation frequency at genomic CpG sites from the output of `nanopolish call-methylation`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output files format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Freq_meth_calculate` can generates 2 files, a standard BED file and a tabulated file containing extra information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### BED file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Standard genomic BED6 (https://genome.ucsc.edu/FAQ/FAQformat.html#format1). The score correspond to the methylation frequency multiplied by 1000. The file is sorted by coordinates and can be rendered with a genome browser such as [IGV](https://software.broadinstitute.org/software/igv/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Tabulated TSV file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Contrary to the bed file, in the tabulated report, positions are ordered by decreasing methylation frequency.\n",
    "\n",
    "The file contains the following fields:\n",
    "\n",
    "* **chrom / start / end / strand**: Genomic coordinates of the motif or group of motifs in case split_group was not selected.\n",
    "* **site_id**: Unique integer identifier of the genomic position.\n",
    "* **methylated_reads / unmethylated_reads / ambiguous_reads**: Number of reads at a given genomic location with a higher likelyhood of being methylated or unmethylated or with an ambiguous methylation call.\n",
    "* **sequence**: -5 to +5 sequence of the motif or group of motifs in case split_group was not selected.\n",
    "* **num_motifs**: Number of motif in the group.\n",
    "* **meth_freq**: Methylation frequency (out of non anbiguous calls)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bash command line usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-10T15:20:48.144465Z",
     "start_time": "2019-09-10T15:20:47.185420Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: NanopolishComp Freq_meth_calculate [-h] [-i INPUT_FN]\n",
      "                                          [-b OUTPUT_BED_FN]\n",
      "                                          [-t OUTPUT_TSV_FN] [-d MIN_DEPTH]\n",
      "                                          [-f FASTA_INDEX] [-s SAMPLE_ID]\n",
      "                                          [--strand_specific]\n",
      "                                          [--min_llr MIN_LLR] [-v | -q]\n",
      "\n",
      "Calculate methylation frequency at genomic CpG sites from the output of\n",
      "nanopolish call-methylation\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -v, --verbose         Increase verbosity\n",
      "  -q, --quiet           Reduce verbosity\n",
      "\n",
      "Input/Output options:\n",
      "  -i INPUT_FN, --input_fn INPUT_FN\n",
      "                        Path to a nanopolish call_methylation tsv output file.\n",
      "                        If not specified read from std input\n",
      "  -b OUTPUT_BED_FN, --output_bed_fn OUTPUT_BED_FN\n",
      "                        Path to write a summary result file in BED format\n",
      "  -t OUTPUT_TSV_FN, --output_tsv_fn OUTPUT_TSV_FN\n",
      "                        Path to write an more extensive result report in TSV\n",
      "                        format\n",
      "\n",
      "Filtering options:\n",
      "  -d MIN_DEPTH, --min_depth MIN_DEPTH\n",
      "                        Minimal number of reads covering a site to be reported\n",
      "                        (default: 10)\n",
      "\n",
      "Other options:\n",
      "  -f FASTA_INDEX, --fasta_index FASTA_INDEX\n",
      "                        fasta index file obtained with samtools faidx.\n",
      "                        Required for coordinate sorting\n",
      "  -s SAMPLE_ID, --sample_id SAMPLE_ID\n",
      "                        Sample ID to be used for the bed track header\n",
      "                        (default: )\n",
      "  --strand_specific     Output strand specific sites\n",
      "  --min_llr MIN_LLR     Minimal log likelyhood ratio to consider a site\n",
      "                        significantly methylated or unmethylated (default: 2)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon NanopolishComp\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From an existing nanopolish call_methylation file output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-10T15:21:11.598368Z",
     "start_time": "2019-09-10T15:20:55.289497Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name=Methylation_Sample1 itemRgb=On\n",
      "chr-VIII\t138415\t138416\t7224149695747901416\t-3.200000\t.\t138415\t138416\t'8,121,207'\n",
      "chr-VIII\t138429\t138430\t7224167352505884938\t-4.722000\t.\t138429\t138430\t'8,121,207'\n",
      "chr-VIII\t212351\t212352\t7296895538640012056\t-4.125000\t.\t212351\t212352\t'8,121,207'\n",
      "chr-VIII\t212392\t212393\t7296795904077105039\t-1.866667\t.\t212392\t212393\t'100,100,100'\n",
      "chr-VIII\t212457\t212461\t7296716448666179190\t-5.546667\t.\t212457\t212461\t'8,121,207'\n",
      "chr-VIII\t212530\t212531\t7296001349967846549\t-1.850833\t.\t212530\t212531\t'100,100,100'\n",
      "chr-VIII\t212581\t212582\t7295894148222946594\t-2.307692\t.\t212581\t212582\t'8,121,207'\n",
      "chr-VIII\t212596\t212600\t7295913066177928939\t-5.900000\t.\t212596\t212600\t'8,121,207'\n",
      "chr-VIII\t212612\t212613\t7295852528721985435\t-3.675385\t.\t212612\t212613\t'8,121,207'\n",
      "chromosome\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tllr_list\n",
      "chr-VIII\t138415\t138416\t.\t7224149695747901416\t0\t7\t3\tGGTCTCGCTTT\t1\t-5.51,-9.63,-2.23,0.53,-0.47,-5.44,-2.38,-5.64,-2.29,1.06\n",
      "chr-VIII\t138429\t138430\t.\t7224167352505884938\t0\t8\t2\tAGCTTCGAGGA\t1\t-5.14,-3.64,-9.53,-8.07,-2.08,1.16,-5.42,-4.97,-9.1,-0.43\n",
      "chr-VIII\t212351\t212352\t.\t7296895538640012056\t0\t8\t4\tTGGGGCGACAT\t1\t-9.1,-6.06,-3.14,0.53,-2.48,-11.61,0.17,-2.76,-3.4,-12.44,0.66,0.13\n",
      "chr-VIII\t212392\t212393\t.\t7296795904077105039\t1\t5\t6\tATTAACGTATA\t1\t0.21,-1.82,-6.91,-4.89,-1.92,3.09,-3.07,-1.12,1.83,-1.82,-2.98,-3.0\n",
      "chr-VIII\t212457\t212461\t.\t7296716448666179190\t0\t8\t4\tAGAATCGTCGATTA\t2\t-3.48,0.08,-6.33,-0.33,-4.83,-13.56,-1.71,-10.95,-3.32,-12.85,-1.55,-7.73\n",
      "chr-VIII\t212530\t212531\t.\t7296001349967846549\t0\t5\t7\tCTATTCGTTTC\t1\t-1.48,-1.42,-5.41,-2.82,-0.49,1.29,0.15,-2.73,-1.09,-1.11,-4.52,-2.58\n",
      "chr-VIII\t212581\t212582\t.\t7295894148222946594\t1\t4\t8\tGTTACCGCAGG\t1\t0.87,-6.56,1.21,1.75,0.86,-0.03,7.31,-12.54,-0.77,-5.27,-18.42,0.23,1.36\n",
      "chr-VIII\t212596\t212600\t.\t7295913066177928939\t0\t10\t3\tTTTGTCGTCGCTGT\t2\t-4.4,-8.76,-13.95,-0.94,1.11,-1.47,-5.42,-3.37,-7.38,-9.96,-13.14,-6.26,-2.76\n",
      "chr-VIII\t212612\t212613\t.\t7295852528721985435\t0\t8\t5\tCACCCCGTTGG\t1\t0.97,-0.5,-7.73,-6.92,-3.65,-2.94,-9.09,-0.83,-7.41,0.14,1.11,-2.7,-8.23\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Options summary ##\n",
      "\tpackage_name: NanopolishComp\n",
      "\tpackage_version: 0.6.11\n",
      "\ttimestamp: 2019-09-10 16:20:56.119314\n",
      "\tquiet: False\n",
      "\tverbose: True\n",
      "\tmin_llr: 2\n",
      "\tstrand_specific: False\n",
      "\tsample_id: Sample1\n",
      "\tmin_depth: 10\n",
      "\toutput_tsv_fn: ./output/freq_meth_calculate/out_freq_meth_calculate.tsv\n",
      "\toutput_bed_fn: ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
      "\tfasta_index: \n",
      "\tinput_fn: data/freq_meth_calculate/methylation_calls.tsv\n",
      "## Checking arguments ##\n",
      "\tTesting input file readability\n",
      "\tCheck output file\n",
      "\t\tOutput results in bed format\n",
      "\t\tOutput results in tsv format\n",
      "## Parsing methylation_calls file ##\n",
      "\tStarting to parse file Nanopolish methylation call file\n",
      "\r",
      "\t: 0 lines [00:00, ? lines/s]\r",
      "\t: 4698 lines [00:00, 46976.57 lines/s]\r",
      "\t: 9389 lines [00:00, 46954.54 lines/s]\r",
      "\t: 13850 lines [00:00, 46224.86 lines/s]\r",
      "\t: 18323 lines [00:00, 45763.72 lines/s]\r",
      "\t: 22890 lines [00:00, 45735.41 lines/s]\r",
      "\t: 26901 lines [00:00, 43885.31 lines/s]\r",
      "\t: 31590 lines [00:00, 44745.17 lines/s]\r",
      "\t: 36213 lines [00:00, 45178.59 lines/s]\r",
      "\t: 40900 lines [00:00, 45672.84 lines/s]\r",
      "\t: 45581 lines [00:01, 46007.80 lines/s]\r",
      "\t: 50222 lines [00:01, 46125.94 lines/s]\r",
      "\t: 55035 lines [00:01, 46706.99 lines/s]\r",
      "\t: 59913 lines [00:01, 47308.04 lines/s]\r",
      "\t: 64647 lines [00:01, 47316.11 lines/s]\r",
      "\t: 69351 lines [00:01, 47045.36 lines/s]\r",
      "\t: 74037 lines [00:01, 46799.50 lines/s]\r",
      "\t: 78716 lines [00:01, 46796.22 lines/s]\r",
      "\t: 83387 lines [00:01, 46691.92 lines/s]\r",
      "\t: 88050 lines [00:01, 46515.56 lines/s]\r",
      "\t: 92698 lines [00:02, 46326.83 lines/s]\r",
      "\t: 97328 lines [00:02, 44279.63 lines/s]\r",
      "\t: 101796 lines [00:02, 44396.54 lines/s]\r",
      "\t: 106337 lines [00:02, 44694.74 lines/s]\r",
      "\t: 110941 lines [00:02, 45089.65 lines/s]\r",
      "\t: 115458 lines [00:02, 42034.66 lines/s]\r",
      "\t: 120019 lines [00:02, 43045.23 lines/s]\r",
      "\t: 124679 lines [00:02, 44051.34 lines/s]\r",
      "\t: 129371 lines [00:02, 44873.35 lines/s]\r",
      "\t: 133985 lines [00:02, 45245.97 lines/s]\r",
      "\t: 138574 lines [00:03, 45436.67 lines/s]\r",
      "\t: 143254 lines [00:03, 45836.27 lines/s]\r",
      "\t: 148109 lines [00:03, 46616.63 lines/s]\r",
      "\t: 153028 lines [00:03, 47359.64 lines/s]\r",
      "\t: 157799 lines [00:03, 47463.58 lines/s]\r",
      "\t: 162553 lines [00:03, 47109.26 lines/s]\r",
      "\t: 167270 lines [00:03, 43523.92 lines/s]\r",
      "\t: 171948 lines [00:03, 44450.40 lines/s]\r",
      "\t: 176581 lines [00:03, 44997.80 lines/s]\r",
      "\t: 181199 lines [00:03, 45345.81 lines/s]\r",
      "\t: 185813 lines [00:04, 45578.91 lines/s]\r",
      "\t: 190412 lines [00:04, 45700.53 lines/s]\r",
      "\t: 195028 lines [00:04, 45835.76 lines/s]\r",
      "\t: 199621 lines [00:04, 44930.05 lines/s]\r",
      "\t: 204124 lines [00:04, 44696.76 lines/s]\r",
      "\t: 208659 lines [00:04, 44890.06 lines/s]\r",
      "\t: 213292 lines [00:04, 45312.06 lines/s]\r",
      "\t: 217952 lines [00:04, 45689.25 lines/s]\r",
      "\t: 222528 lines [00:04, 45707.17 lines/s]\r",
      "\t: 227134 lines [00:04, 45810.28 lines/s]\r",
      "\t: 231782 lines [00:05, 46006.96 lines/s]\r",
      "\t: 236567 lines [00:05, 46542.11 lines/s]\r",
      "\t: 241385 lines [00:05, 47015.71 lines/s]\r",
      "\t: 246177 lines [00:05, 47280.77 lines/s]\r",
      "\t: 250908 lines [00:05, 46644.13 lines/s]\r",
      "\t: 255577 lines [00:05, 42189.07 lines/s]\r",
      "\t: 260252 lines [00:05, 43459.14 lines/s]\r",
      "\t: 264859 lines [00:05, 44210.04 lines/s]\r",
      "\t: 269480 lines [00:05, 44790.26 lines/s]\r",
      "\t: 274146 lines [00:06, 45334.54 lines/s]\r",
      "\t: 278733 lines [00:06, 45492.49 lines/s]\r",
      "\t: 283310 lines [00:06, 45575.14 lines/s]\r",
      "\t: 287967 lines [00:06, 45869.05 lines/s]\r",
      "\t: 292611 lines [00:06, 46036.62 lines/s]\r",
      "\t: 297257 lines [00:06, 46162.57 lines/s]\r",
      "\t: 301879 lines [00:06, 46051.65 lines/s]\r",
      "\t: 306556 lines [00:06, 46263.56 lines/s]\r",
      "\t: 311194 lines [00:06, 46297.88 lines/s]\r",
      "\t: 315826 lines [00:06, 46178.53 lines/s]\r",
      "\t: 320446 lines [00:07, 46161.88 lines/s]\r",
      "\t: 325064 lines [00:07, 41324.33 lines/s]\r",
      "\t: 329863 lines [00:07, 43120.83 lines/s]\r",
      "\t: 334764 lines [00:07, 44732.76 lines/s]\r",
      "\t: 339482 lines [00:07, 45436.61 lines/s]\r",
      "\t: 344114 lines [00:07, 45697.15 lines/s]\r",
      "\t: 348738 lines [00:07, 45858.31 lines/s]\r",
      "\t: 353352 lines [00:07, 45223.46 lines/s]\r",
      "\t: 357967 lines [00:07, 45495.81 lines/s]\r",
      "\t: 362652 lines [00:07, 45891.50 lines/s]\r",
      "\t: 367349 lines [00:08, 46209.05 lines/s]\r",
      "\t: 372040 lines [00:08, 46416.48 lines/s]\r",
      "\t: 376688 lines [00:08, 46408.17 lines/s]\r",
      "\t: 381334 lines [00:08, 46353.89 lines/s]\r",
      "\t: 385973 lines [00:08, 46295.26 lines/s]\r",
      "\t: 390605 lines [00:08, 46023.29 lines/s]\r",
      "\t: 395275 lines [00:08, 46222.61 lines/s]\r",
      "\t: 399976 lines [00:08, 46455.04 lines/s]\r",
      "\t: 404623 lines [00:08, 46286.00 lines/s]\r",
      "\t: 409253 lines [00:08, 46284.45 lines/s]\r",
      "\t: 413912 lines [00:09, 46373.33 lines/s]\r",
      "\t: 418640 lines [00:09, 46640.46 lines/s]\r",
      "\t: 423539 lines [00:09, 47320.83 lines/s]\r",
      "\t: 428360 lines [00:09, 47583.83 lines/s]\r",
      "\t: 433121 lines [00:09, 47458.92 lines/s]\r",
      "\t: 437869 lines [00:09, 47272.38 lines/s]\r",
      "\t: 442598 lines [00:09, 46846.06 lines/s]\r",
      "\t: 447285 lines [00:09, 46559.99 lines/s]\r",
      "\t: 451943 lines [00:09, 46534.46 lines/s]\r",
      "\t: 456598 lines [00:09, 46314.10 lines/s]\r",
      "\t: 461265 lines [00:10, 46418.53 lines/s]\r",
      "\t: 465935 lines [00:10, 46500.37 lines/s]\r",
      "\t: 470586 lines [00:10, 40083.66 lines/s]\r",
      "\t: 474805 lines [00:10, 40692.73 lines/s]\r",
      "\t: 478987 lines [00:10, 38424.14 lines/s]\r",
      "\t: 483642 lines [00:10, 40546.86 lines/s]\r",
      "\t: 488344 lines [00:10, 42291.14 lines/s]\r",
      "\t: 492974 lines [00:10, 43418.83 lines/s]\r",
      "\t: 497639 lines [00:10, 44338.23 lines/s]\r",
      "\t: 502323 lines [00:11, 45059.55 lines/s]\r",
      "\t: 507021 lines [00:11, 45618.91 lines/s]\r",
      "\t: 511886 lines [00:11, 46485.74 lines/s]\r",
      "\t: 516561 lines [00:11, 45713.06 lines/s]\r",
      "\t: 521154 lines [00:11, 42458.60 lines/s]\r",
      "\t: 525461 lines [00:11, 36311.37 lines/s]\r",
      "\t: 529657 lines [00:11, 37838.25 lines/s]\r",
      "\t: 533888 lines [00:11, 39076.87 lines/s]\r",
      "\t: 538190 lines [00:11, 40180.19 lines/s]\r",
      "\t: 542764 lines [00:12, 41699.28 lines/s]\r",
      "\t: 547398 lines [00:12, 42990.86 lines/s]\r",
      "\t: 552019 lines [00:12, 43907.59 lines/s]\r",
      "\t: 556461 lines [00:12, 43834.70 lines/s]\r",
      "\t: 560880 lines [00:12, 43117.44 lines/s]\r",
      "\t: 565220 lines [00:12, 43169.78 lines/s]\r",
      "\t: 569604 lines [00:12, 43367.22 lines/s]\r",
      "\t: 573955 lines [00:12, 42905.30 lines/s]\r",
      "\t: 578257 lines [00:12, 36218.82 lines/s]\r",
      "\t: 582163 lines [00:13, 37026.29 lines/s]\r",
      "\t: 586566 lines [00:13, 38881.71 lines/s]\r",
      "\t: 591115 lines [00:13, 40652.33 lines/s]\r",
      "\t: 595724 lines [00:13, 42141.89 lines/s]\r",
      "\t: 600228 lines [00:13, 42970.14 lines/s]\r",
      "\t: 604590 lines [00:13, 43136.93 lines/s]\r",
      "\t: 605248 lines [00:13, 44691.91 lines/s]\n",
      "\tFiltering out low coverage sites\n",
      "\tProcessing valid sites found\n",
      "\t\tWrite output file header\n",
      "\r",
      "\t:   0%|          | 0/804 [00:00<?, ? sites/s]\r",
      "\t:  16%|█▌        | 127/804 [00:00<00:00, 1247.04 sites/s]\r",
      "\t:  23%|██▎       | 182/804 [00:00<00:00, 888.36 sites/s] \r",
      "\t:  28%|██▊       | 225/804 [00:00<00:00, 617.72 sites/s]\r",
      "\t:  33%|███▎      | 266/804 [00:00<00:01, 444.86 sites/s]\r",
      "\t:  38%|███▊      | 303/804 [00:00<00:01, 313.09 sites/s]\r",
      "\t:  42%|████▏     | 334/804 [00:00<00:02, 206.64 sites/s]\r",
      "\t:  45%|████▍     | 360/804 [00:01<00:02, 192.39 sites/s]\r",
      "\t:  48%|████▊     | 383/804 [00:01<00:02, 198.90 sites/s]\r",
      "\t:  51%|█████     | 410/804 [00:01<00:01, 215.03 sites/s]\r",
      "\t:  55%|█████▌    | 443/804 [00:01<00:01, 239.67 sites/s]\r",
      "\t:  61%|██████    | 492/804 [00:01<00:01, 282.82 sites/s]\r",
      "\t:  76%|███████▋  | 614/804 [00:01<00:00, 367.50 sites/s]\r",
      "\t: 100%|██████████| 804/804 [00:01<00:00, 474.31 sites/s]\n",
      "## Results summary ##\n",
      "\tTotal read lines: 605,248\n",
      "\tValid read lines: 605,248\n",
      "\tTotal sites: 229,389\n",
      "\tLow coverage sites: 228,585\n",
      "\tValid sites: 1,608\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon NanopolishComp\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --verbose -i data/freq_meth_calculate/methylation_calls.tsv -b ./output/freq_meth_calculate/out_freq_meth_calculate.bed -t ./output/freq_meth_calculate/out_freq_meth_calculate.tsv -s Sample1\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Using a fasta index for output coordinates sorting and strand specific sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-10T15:22:07.152255Z",
     "start_time": "2019-09-10T15:21:50.468737Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name=Methylation_Sample1 itemRgb=On\n",
      "chr-VIII\t213351\t213352\t-8023785553745245992\t-0.500000\t-\t213351\t213352\t'100,100,100'\n",
      "chr-VIII\t213382\t213386\t-8023583102452360367\t-5.103000\t-\t213382\t213386\t'8,121,207'\n",
      "chr-VIII\t213427\t213428\t-8023686382448928188\t0.504545\t-\t213427\t213428\t'100,100,100'\n",
      "chr-VIII\t213516\t213517\t-8022685139567842841\t-2.447500\t-\t213516\t213517\t'8,121,207'\n",
      "chr-VIII\t213536\t213537\t-8022798660204149733\t-3.103333\t-\t213536\t213537\t'8,121,207'\n",
      "chr-VIII\t213549\t213550\t-8022802421743098730\t-2.724167\t-\t213549\t213550\t'8,121,207'\n",
      "chr-VIII\t213645\t213646\t-8022541247023329610\t-2.546667\t-\t213645\t213646\t'8,121,207'\n",
      "chr-VIII\t213668\t213669\t-8022652358379081217\t-3.728333\t-\t213668\t213669\t'8,121,207'\n",
      "chr-VIII\t213729\t213730\t-8022716548847329270\t-3.253846\t-\t213729\t213730\t'8,121,207'\n",
      "chromosome\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tllr_list\n",
      "chr-VIII\t213351\t213352\t-\t-8023785553745245992\t0\t1\t9\tACTAACGGGGA\t1\t-2.75,-0.46,1.48,-1.35,0.34,0.66,-1.58,-1.91,0.06,0.51\n",
      "chr-VIII\t213382\t213386\t-\t-8023583102452360367\t1\t7\t2\tTTCTTCGCCGACTG\t2\t-7.94,3.23,-0.77,0.05,-9.8,-5.7,-11.29,-5.61,-4.64,-8.56\n",
      "chr-VIII\t213427\t213428\t-\t-8023686382448928188\t1\t0\t10\tTTTCTCGCAAA\t1\t-0.28,0.35,-0.06,0.05,-0.11,2.63,-1.02,1.27,1.31,1.05,0.36\n",
      "chr-VIII\t213516\t213517\t-\t-8022685139567842841\t0\t6\t6\tACTTCCGTTGC\t1\t-2.01,-5.92,-0.28,-0.07,-0.59,0.87,-5.23,-0.84,-2.65,-7.24,0.47,-5.88\n",
      "chr-VIII\t213536\t213537\t-\t-8022798660204149733\t0\t11\t1\tCATCTCGTAAA\t1\t-3.33,-2.35,-4.14,-2.21,-1.92,-2.04,-2.17,-4.34,-2.79,-3.75,-4.01,-4.19\n",
      "chr-VIII\t213549\t213550\t-\t-8022802421743098730\t1\t7\t4\tGGATACGATAA\t1\t-3.3,-4.74,-0.15,-0.08,-4.84,-8.16,-3.61,2.21,-2.28,-1.69,-6.15,0.1\n",
      "chr-VIII\t213645\t213646\t-\t-8022541247023329610\t0\t6\t6\tAAATTCGTCAT\t1\t-9.16,-0.83,1.66,0.58,-2.18,-5.19,-2.13,-5.99,-3.47,-1.88,-1.55,-0.42\n",
      "chr-VIII\t213668\t213669\t-\t-8022652358379081217\t0\t8\t4\tCACAACGTTGG\t1\t-10.85,-2.37,0.81,-1.98,-4.37,-7.49,-2.94,-0.06,-4.53,-3.63,-0.61,-6.72\n",
      "chr-VIII\t213729\t213730\t-\t-8022716548847329270\t0\t9\t4\tATTTGCGAATA\t1\t-2.79,-4.55,1.21,-0.07,-6.66,-2.28,-7.42,-2.66,-1.86,-6.36,-5.87,0.39,-3.38\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Options summary ##\n",
      "\tpackage_name: NanopolishComp\n",
      "\tpackage_version: 0.6.11\n",
      "\ttimestamp: 2019-09-10 16:21:51.314487\n",
      "\tquiet: False\n",
      "\tverbose: True\n",
      "\tmin_llr: 2\n",
      "\tstrand_specific: True\n",
      "\tsample_id: Sample1\n",
      "\tmin_depth: 10\n",
      "\toutput_tsv_fn: ./output/freq_meth_calculate/out_freq_meth_calculate.tsv\n",
      "\toutput_bed_fn: ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
      "\tfasta_index: data/freq_meth_calculate/ref.fa.fai\n",
      "\tinput_fn: data/freq_meth_calculate/methylation_calls.tsv\n",
      "## Checking arguments ##\n",
      "\tTesting input file readability\n",
      "\tCheck output file\n",
      "\t\tOutput results in bed format\n",
      "\t\tOutput results in tsv format\n",
      "## Parsing methylation_calls file ##\n",
      "\tStarting to parse file Nanopolish methylation call file\n",
      "\r",
      "\t: 0 lines [00:00, ? lines/s]\r",
      "\t: 4631 lines [00:00, 46302.98 lines/s]\r",
      "\t: 9194 lines [00:00, 46098.59 lines/s]\r",
      "\t: 13789 lines [00:00, 46052.54 lines/s]\r",
      "\t: 17310 lines [00:00, 42156.49 lines/s]\r",
      "\t: 21883 lines [00:00, 43167.34 lines/s]\r",
      "\t: 26434 lines [00:00, 43843.43 lines/s]\r",
      "\t: 31080 lines [00:00, 44595.99 lines/s]\r",
      "\t: 35476 lines [00:00, 44402.99 lines/s]\r",
      "\t: 39671 lines [00:00, 43273.55 lines/s]\r",
      "\t: 43833 lines [00:01, 41946.73 lines/s]\r",
      "\t: 47950 lines [00:01, 41709.15 lines/s]\r",
      "\t: 52047 lines [00:01, 41348.23 lines/s]\r",
      "\t: 56601 lines [00:01, 42521.17 lines/s]\r",
      "\t: 61113 lines [00:01, 43267.77 lines/s]\r",
      "\t: 65424 lines [00:01, 41015.13 lines/s]\r",
      "\t: 69539 lines [00:01, 39778.79 lines/s]\r",
      "\t: 73566 lines [00:01, 39921.87 lines/s]\r",
      "\t: 77571 lines [00:01, 38655.87 lines/s]\r",
      "\t: 81977 lines [00:01, 40132.24 lines/s]\r",
      "\t: 86087 lines [00:02, 40413.75 lines/s]\r",
      "\t: 90228 lines [00:02, 40707.11 lines/s]\r",
      "\t: 94608 lines [00:02, 41584.69 lines/s]\r",
      "\t: 98880 lines [00:02, 41916.99 lines/s]\r",
      "\t: 103159 lines [00:02, 42174.52 lines/s]\r",
      "\t: 107385 lines [00:02, 42042.04 lines/s]\r",
      "\t: 111748 lines [00:02, 42504.13 lines/s]\r",
      "\t: 116004 lines [00:02, 38730.62 lines/s]\r",
      "\t: 120353 lines [00:02, 40044.70 lines/s]\r",
      "\t: 124823 lines [00:02, 41335.08 lines/s]\r",
      "\t: 129402 lines [00:03, 42576.43 lines/s]\r",
      "\t: 133990 lines [00:03, 43513.94 lines/s]\r",
      "\t: 138587 lines [00:03, 44221.49 lines/s]\r",
      "\t: 143378 lines [00:03, 45265.08 lines/s]\r",
      "\t: 147931 lines [00:03, 44719.41 lines/s]\r",
      "\t: 152459 lines [00:03, 44884.22 lines/s]\r",
      "\t: 156962 lines [00:03, 44726.34 lines/s]\r",
      "\t: 161445 lines [00:03, 44632.56 lines/s]\r",
      "\t: 165916 lines [00:03, 40676.15 lines/s]\r",
      "\t: 170449 lines [00:04, 41967.32 lines/s]\r",
      "\t: 174957 lines [00:04, 42853.85 lines/s]\r",
      "\t: 179439 lines [00:04, 43422.70 lines/s]\r",
      "\t: 184002 lines [00:04, 44061.45 lines/s]\r",
      "\t: 188669 lines [00:04, 44810.92 lines/s]\r",
      "\t: 193288 lines [00:04, 45214.32 lines/s]\r",
      "\t: 197882 lines [00:04, 45428.52 lines/s]\r",
      "\t: 202437 lines [00:04, 45306.26 lines/s]\r",
      "\t: 206976 lines [00:04, 45045.43 lines/s]\r",
      "\t: 211570 lines [00:04, 45306.05 lines/s]\r",
      "\t: 216106 lines [00:05, 39871.83 lines/s]\r",
      "\t: 220800 lines [00:05, 41756.16 lines/s]\r",
      "\t: 225424 lines [00:05, 43005.22 lines/s]\r",
      "\t: 230099 lines [00:05, 44063.47 lines/s]\r",
      "\t: 234633 lines [00:05, 44436.99 lines/s]\r",
      "\t: 239285 lines [00:05, 45041.27 lines/s]\r",
      "\t: 243908 lines [00:05, 45391.31 lines/s]\r",
      "\t: 248473 lines [00:05, 44522.62 lines/s]\r",
      "\t: 252998 lines [00:05, 44737.74 lines/s]\r",
      "\t: 257621 lines [00:05, 45173.29 lines/s]\r",
      "\t: 262196 lines [00:06, 45339.00 lines/s]\r",
      "\t: 266766 lines [00:06, 45445.78 lines/s]\r",
      "\t: 271317 lines [00:06, 45313.38 lines/s]\r",
      "\t: 275874 lines [00:06, 45386.35 lines/s]\r",
      "\t: 280416 lines [00:06, 45365.42 lines/s]\r",
      "\t: 284955 lines [00:06, 38503.83 lines/s]\r",
      "\t: 289517 lines [00:06, 40392.24 lines/s]\r",
      "\t: 294108 lines [00:06, 41902.42 lines/s]\r",
      "\t: 298764 lines [00:06, 43198.12 lines/s]\r",
      "\t: 303426 lines [00:07, 44168.01 lines/s]\r",
      "\t: 308114 lines [00:07, 44945.72 lines/s]\r",
      "\t: 312862 lines [00:07, 45675.20 lines/s]\r",
      "\t: 317470 lines [00:07, 45701.27 lines/s]\r",
      "\t: 322081 lines [00:07, 45823.04 lines/s]\r",
      "\t: 326684 lines [00:07, 45534.12 lines/s]\r",
      "\t: 331256 lines [00:07, 45588.31 lines/s]\r",
      "\t: 335826 lines [00:07, 45621.48 lines/s]\r",
      "\t: 340396 lines [00:07, 45047.41 lines/s]\r",
      "\t: 344908 lines [00:07, 44656.96 lines/s]\r",
      "\t: 349380 lines [00:08, 44298.46 lines/s]\r",
      "\t: 353909 lines [00:08, 44590.72 lines/s]\r",
      "\t: 358546 lines [00:08, 45108.44 lines/s]\r",
      "\t: 363236 lines [00:08, 45626.53 lines/s]\r",
      "\t: 367876 lines [00:08, 45854.30 lines/s]\r",
      "\t: 372465 lines [00:08, 45451.47 lines/s]\r",
      "\t: 377014 lines [00:08, 45078.19 lines/s]\r",
      "\t: 381525 lines [00:08, 42739.27 lines/s]\r",
      "\t: 385921 lines [00:08, 43097.56 lines/s]\r",
      "\t: 390666 lines [00:08, 44314.99 lines/s]\r",
      "\t: 395348 lines [00:09, 45035.64 lines/s]\r",
      "\t: 400274 lines [00:09, 46224.66 lines/s]\r",
      "\t: 404918 lines [00:09, 41082.22 lines/s]\r",
      "\t: 409147 lines [00:09, 35777.69 lines/s]\r",
      "\t: 413717 lines [00:09, 38269.57 lines/s]\r",
      "\t: 418496 lines [00:09, 40701.02 lines/s]\r",
      "\t: 422888 lines [00:09, 41614.68 lines/s]\r",
      "\t: 427415 lines [00:09, 42646.79 lines/s]\r",
      "\t: 432077 lines [00:09, 43764.67 lines/s]\r",
      "\t: 436529 lines [00:10, 42584.08 lines/s]\r",
      "\t: 440847 lines [00:10, 38164.63 lines/s]\r",
      "\t: 444792 lines [00:10, 37160.22 lines/s]\r",
      "\t: 449250 lines [00:10, 39112.00 lines/s]\r",
      "\t: 453762 lines [00:10, 40739.26 lines/s]\r",
      "\t: 458258 lines [00:10, 41919.37 lines/s]\r",
      "\t: 462726 lines [00:10, 42709.03 lines/s]\r",
      "\t: 467213 lines [00:10, 43334.59 lines/s]\r",
      "\t: 471683 lines [00:10, 43733.36 lines/s]\r",
      "\t: 476084 lines [00:11, 42901.69 lines/s]\r",
      "\t: 480530 lines [00:11, 43356.51 lines/s]\r",
      "\t: 485044 lines [00:11, 43874.80 lines/s]\r",
      "\t: 489681 lines [00:11, 44594.62 lines/s]\r",
      "\t: 494153 lines [00:11, 44631.28 lines/s]\r",
      "\t: 498625 lines [00:11, 42131.31 lines/s]\r",
      "\t: 503263 lines [00:11, 43320.18 lines/s]\r",
      "\t: 507628 lines [00:11, 36422.12 lines/s]\r",
      "\t: 511942 lines [00:11, 38206.58 lines/s]\r",
      "\t: 516013 lines [00:12, 38921.75 lines/s]\r",
      "\t: 520311 lines [00:12, 40056.24 lines/s]\r",
      "\t: 524930 lines [00:12, 41716.05 lines/s]\r",
      "\t: 529666 lines [00:12, 43261.83 lines/s]\r",
      "\t: 534381 lines [00:12, 44357.82 lines/s]\r",
      "\t: 539001 lines [00:12, 44894.44 lines/s]\r",
      "\t: 543656 lines [00:12, 45377.13 lines/s]\r",
      "\t: 548340 lines [00:12, 45805.36 lines/s]\r",
      "\t: 552944 lines [00:12, 44681.69 lines/s]\r",
      "\t: 557434 lines [00:12, 43418.17 lines/s]\r",
      "\t: 561799 lines [00:13, 43045.06 lines/s]\r",
      "\t: 566466 lines [00:13, 44070.86 lines/s]\r",
      "\t: 571322 lines [00:13, 45326.43 lines/s]\r",
      "\t: 575906 lines [00:13, 45477.92 lines/s]\r",
      "\t: 580469 lines [00:13, 43947.41 lines/s]\r",
      "\t: 584985 lines [00:13, 44303.43 lines/s]\r",
      "\t: 589432 lines [00:13, 43456.83 lines/s]\r",
      "\t: 593793 lines [00:13, 42339.89 lines/s]\r",
      "\t: 598392 lines [00:13, 43372.38 lines/s]\r",
      "\t: 603033 lines [00:13, 44240.85 lines/s]\r",
      "\t: 605248 lines [00:14, 43139.23 lines/s]\n",
      "\tFiltering out low coverage sites\n",
      "\tSorting by coordinates\n",
      "\tProcessing valid sites found\n",
      "\t\tWrite output file header\n",
      "\r",
      "\t:   0%|          | 0/999 [00:00<?, ? sites/s]\r",
      "\t:  17%|█▋        | 168/999 [00:00<00:00, 1671.03 sites/s]\r",
      "\t:  27%|██▋       | 266/999 [00:00<00:00, 1377.73 sites/s]\r",
      "\t:  33%|███▎      | 331/999 [00:00<00:00, 1017.13 sites/s]\r",
      "\t:  40%|███▉      | 396/999 [00:00<00:00, 783.58 sites/s] \r",
      "\t:  46%|████▌     | 457/999 [00:00<00:00, 617.00 sites/s]\r",
      "\t:  51%|█████▏    | 512/999 [00:00<00:01, 458.63 sites/s]\r",
      "\t:  56%|█████▌    | 559/999 [00:00<00:01, 397.74 sites/s]\r",
      "\t:  60%|██████    | 601/999 [00:01<00:01, 396.12 sites/s]\r",
      "\t:  65%|██████▌   | 650/999 [00:01<00:00, 420.15 sites/s]\r",
      "\t:  71%|███████   | 711/999 [00:01<00:00, 463.02 sites/s]\r",
      "\t:  79%|███████▉  | 789/999 [00:01<00:00, 527.27 sites/s]\r",
      "\t:  90%|████████▉ | 896/999 [00:01<00:00, 621.30 sites/s]\r",
      "\t: 100%|██████████| 999/999 [00:01<00:00, 668.32 sites/s]\n",
      "## Results summary ##\n",
      "\tTotal read lines: 605,248\n",
      "\tValid read lines: 605,248\n",
      "\tTotal sites: 340,081\n",
      "\tLow coverage sites: 339,082\n",
      "\tValid sites: 1,998\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon NanopolishComp\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --verbose --strand_specific -i data/freq_meth_calculate/methylation_calls.tsv -b ./output/freq_meth_calculate/out_freq_meth_calculate.bed -t ./output/freq_meth_calculate/out_freq_meth_calculate.tsv -s Sample1 -f data/freq_meth_calculate/ref.fa.fai\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Changing filtering thresholds (not recommended)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-10T15:22:42.054813Z",
     "start_time": "2019-09-10T15:22:22.524978Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name=Methylation_ itemRgb=On\n",
      "chr-I\t4399\t4400\t1191264914885793541\t-2.344000\t.\t4399\t4400\t'8,121,207'\n",
      "chr-I\t4534\t4535\t1191135011594914772\t-3.033333\t.\t4534\t4535\t'8,121,207'\n",
      "chr-I\t4591\t4592\t1191022765062019525\t-3.181667\t.\t4591\t4592\t'8,121,207'\n",
      "chr-I\t4654\t4655\t1192234775377888428\t-1.055000\t.\t4654\t4655\t'8,121,207'\n",
      "chr-I\t1755\t1756\t1195840537597523385\t-1.706000\t.\t1755\t1756\t'8,121,207'\n",
      "chr-I\t1782\t1787\t1195897291462470420\t-2.776000\t.\t1782\t1787\t'8,121,207'\n",
      "chr-I\t1797\t1798\t1195752253807605775\t-3.224000\t.\t1797\t1798\t'8,121,207'\n",
      "chr-I\t1814\t1815\t1195776216550583412\t-2.454000\t.\t1814\t1815\t'8,121,207'\n",
      "chr-I\t1829\t1830\t1195792612111568111\t-2.896000\t.\t1829\t1830\t'8,121,207'\n",
      "chromosome\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tllr_list\n",
      "chr-I\t4399\t4400\t.\t1191264914885793541\t1\t3\t1\tCTTGCCGAATA\t1\t-3.17,-11.29,5.14,-1.8,-0.6\n",
      "chr-I\t4534\t4535\t.\t1191135011594914772\t0\t5\t1\tTTATTCGTTTT\t1\t-3.31,0.02,-2.66,-5.82,-4.21,-2.22\n",
      "chr-I\t4591\t4592\t.\t1191022765062019525\t0\t3\t3\tTCAAACGCTTG\t1\t-6.16,-0.6,0.28,-0.38,-6.02,-6.21\n",
      "chr-I\t4654\t4655\t.\t1192234775377888428\t1\t4\t1\tTTAATCGACAA\t1\t-1.5,-0.67,-2.66,-2.94,-1.66,3.1\n",
      "chr-I\t1755\t1756\t.\t1195840537597523385\t1\t2\t2\tTTGTTCGTTAA\t1\t-3.37,-0.25,-0.17,-7.47,2.73\n",
      "chr-I\t1782\t1787\t.\t1195897291462470420\t0\t5\t0\tAAACCCGTTCGTAAA\t2\t-2.2,-3.83,-1.37,-4.8,-1.68\n",
      "chr-I\t1797\t1798\t.\t1195752253807605775\t0\t4\t1\tATTGGCGTTTG\t1\t-1.45,-5.09,-2.17,-6.66,-0.75\n",
      "chr-I\t1814\t1815\t.\t1195776216550583412\t0\t4\t1\tGTTTGCGATAG\t1\t-2.2,-4.1,-1.86,-0.4,-3.71\n",
      "chr-I\t1829\t1830\t.\t1195792612111568111\t0\t4\t1\tGATACCGTCCT\t1\t-0.16,-5.78,-3.32,-4.0,-1.22\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Options summary ##\n",
      "\tpackage_name: NanopolishComp\n",
      "\tpackage_version: 0.6.11\n",
      "\ttimestamp: 2019-09-10 16:22:23.393609\n",
      "\tquiet: False\n",
      "\tverbose: True\n",
      "\tmin_llr: 1.0\n",
      "\tstrand_specific: False\n",
      "\tsample_id: \n",
      "\tmin_depth: 5\n",
      "\toutput_tsv_fn: ./output/freq_meth_calculate/out_freq_meth_calculate.tsv\n",
      "\toutput_bed_fn: ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
      "\tfasta_index: \n",
      "\tinput_fn: data/freq_meth_calculate/methylation_calls.tsv\n",
      "## Checking arguments ##\n",
      "\tTesting input file readability\n",
      "\tCheck output file\n",
      "\t\tOutput results in bed format\n",
      "\t\tOutput results in tsv format\n",
      "## Parsing methylation_calls file ##\n",
      "\tStarting to parse file Nanopolish methylation call file\n",
      "\r",
      "\t: 0 lines [00:00, ? lines/s]\r",
      "\t: 4741 lines [00:00, 47407.78 lines/s]\r",
      "\t: 9273 lines [00:00, 46759.33 lines/s]\r",
      "\t: 13799 lines [00:00, 46298.27 lines/s]\r",
      "\t: 18453 lines [00:00, 46368.75 lines/s]\r",
      "\t: 23083 lines [00:00, 46345.23 lines/s]\r",
      "\t: 27188 lines [00:00, 44616.71 lines/s]\r",
      "\t: 31897 lines [00:00, 45328.43 lines/s]\r",
      "\t: 36494 lines [00:00, 45516.12 lines/s]\r",
      "\t: 41168 lines [00:00, 45871.11 lines/s]\r",
      "\t: 45791 lines [00:01, 45971.73 lines/s]\r",
      "\t: 50599 lines [00:01, 46583.68 lines/s]\r",
      "\t: 55487 lines [00:01, 47249.58 lines/s]\r",
      "\t: 60330 lines [00:01, 47573.99 lines/s]\r",
      "\t: 65120 lines [00:01, 47668.83 lines/s]\r",
      "\t: 69859 lines [00:01, 47520.48 lines/s]\r",
      "\t: 74593 lines [00:01, 47465.22 lines/s]\r",
      "\t: 79337 lines [00:01, 47456.07 lines/s]\r",
      "\t: 84073 lines [00:01, 47057.60 lines/s]\r",
      "\t: 88773 lines [00:01, 46346.86 lines/s]\r",
      "\t: 93406 lines [00:02, 44254.84 lines/s]\r",
      "\t: 97851 lines [00:02, 40573.31 lines/s]\r",
      "\t: 102116 lines [00:02, 41173.95 lines/s]\r",
      "\t: 106287 lines [00:02, 41032.08 lines/s]\r",
      "\t: 110862 lines [00:02, 42341.11 lines/s]\r",
      "\t: 115508 lines [00:02, 43496.26 lines/s]\r",
      "\t: 120253 lines [00:02, 44611.05 lines/s]\r",
      "\t: 124928 lines [00:02, 45230.16 lines/s]\r",
      "\t: 129594 lines [00:02, 45649.07 lines/s]\r",
      "\t: 134314 lines [00:02, 46101.36 lines/s]\r",
      "\t: 138956 lines [00:03, 46192.13 lines/s]\r",
      "\t: 143834 lines [00:03, 46938.01 lines/s]\r",
      "\t: 148626 lines [00:03, 47228.30 lines/s]\r",
      "\t: 153628 lines [00:03, 48030.47 lines/s]\r",
      "\t: 158474 lines [00:03, 48157.51 lines/s]\r",
      "\t: 163296 lines [00:03, 47851.03 lines/s]\r",
      "\t: 168086 lines [00:03, 44139.77 lines/s]\r",
      "\t: 172824 lines [00:03, 45064.16 lines/s]\r",
      "\t: 177641 lines [00:03, 45951.41 lines/s]\r",
      "\t: 182437 lines [00:03, 46534.79 lines/s]\r",
      "\t: 187223 lines [00:04, 46922.35 lines/s]\r",
      "\t: 191936 lines [00:04, 46846.68 lines/s]\r",
      "\t: 196635 lines [00:04, 46833.41 lines/s]\r",
      "\t: 201345 lines [00:04, 46909.17 lines/s]\r",
      "\t: 206043 lines [00:04, 46424.53 lines/s]\r",
      "\t: 210745 lines [00:04, 46598.06 lines/s]\r",
      "\t: 215455 lines [00:04, 46741.69 lines/s]\r",
      "\t: 220137 lines [00:04, 46764.50 lines/s]\r",
      "\t: 224825 lines [00:04, 46798.70 lines/s]\r",
      "\t: 229648 lines [00:04, 47215.83 lines/s]\r",
      "\t: 234444 lines [00:05, 47435.01 lines/s]\r",
      "\t: 239353 lines [00:05, 47919.25 lines/s]\r",
      "\t: 244148 lines [00:05, 46657.55 lines/s]\r",
      "\t: 248823 lines [00:05, 46680.89 lines/s]\r",
      "\t: 253498 lines [00:05, 41881.48 lines/s]\r",
      "\t: 258160 lines [00:05, 43197.04 lines/s]\r",
      "\t: 262856 lines [00:05, 44259.74 lines/s]\r",
      "\t: 267648 lines [00:05, 45297.13 lines/s]\r",
      "\t: 272452 lines [00:05, 46085.66 lines/s]\r",
      "\t: 277169 lines [00:06, 46404.01 lines/s]\r",
      "\t: 281913 lines [00:06, 46708.06 lines/s]\r",
      "\t: 286675 lines [00:06, 46976.64 lines/s]\r",
      "\t: 291437 lines [00:06, 47166.26 lines/s]\r",
      "\t: 296164 lines [00:06, 47186.87 lines/s]\r",
      "\t: 300890 lines [00:06, 46325.65 lines/s]\r",
      "\t: 305583 lines [00:06, 46504.37 lines/s]\r",
      "\t: 310240 lines [00:06, 46416.07 lines/s]\r",
      "\t: 314908 lines [00:06, 46493.85 lines/s]\r",
      "\t: 319602 lines [00:06, 46626.29 lines/s]\r",
      "\t: 324267 lines [00:07, 41585.79 lines/s]\r",
      "\t: 328857 lines [00:07, 42788.70 lines/s]\r",
      "\t: 333217 lines [00:07, 42263.57 lines/s]\r",
      "\t: 338144 lines [00:07, 44145.45 lines/s]\r",
      "\t: 342832 lines [00:07, 44930.63 lines/s]\r",
      "\t: 347574 lines [00:07, 45647.97 lines/s]\r",
      "\t: 352299 lines [00:07, 46116.68 lines/s]\r",
      "\t: 357097 lines [00:07, 46660.12 lines/s]\r",
      "\t: 361893 lines [00:07, 47040.46 lines/s]\r",
      "\t: 366637 lines [00:07, 47157.11 lines/s]\r",
      "\t: 371364 lines [00:08, 46742.65 lines/s]\r",
      "\t: 376208 lines [00:08, 47239.01 lines/s]\r",
      "\t: 381004 lines [00:08, 47450.95 lines/s]\r",
      "\t: 385878 lines [00:08, 47830.40 lines/s]\r",
      "\t: 390666 lines [00:08, 47619.83 lines/s]\r",
      "\t: 395432 lines [00:08, 47609.02 lines/s]\r",
      "\t: 400245 lines [00:08, 47763.66 lines/s]\r",
      "\t: 405024 lines [00:08, 47631.95 lines/s]\r",
      "\t: 409789 lines [00:08, 47543.78 lines/s]\r",
      "\t: 414562 lines [00:09, 47596.82 lines/s]\r",
      "\t: 419376 lines [00:09, 47757.79 lines/s]\r",
      "\t: 424153 lines [00:09, 45487.22 lines/s]\r",
      "\t: 428725 lines [00:09, 45246.05 lines/s]\r",
      "\t: 433603 lines [00:09, 46250.00 lines/s]\r",
      "\t: 438401 lines [00:09, 46754.99 lines/s]\r",
      "\t: 443151 lines [00:09, 46974.25 lines/s]\r",
      "\t: 447858 lines [00:09, 46974.90 lines/s]\r",
      "\t: 452563 lines [00:09, 46995.10 lines/s]\r",
      "\t: 457299 lines [00:09, 47102.75 lines/s]\r",
      "\t: 462098 lines [00:10, 47364.51 lines/s]\r",
      "\t: 466838 lines [00:10, 47174.23 lines/s]\r",
      "\t: 471558 lines [00:10, 40544.56 lines/s]\r",
      "\t: 475915 lines [00:10, 41406.84 lines/s]\r",
      "\t: 480694 lines [00:10, 43134.92 lines/s]\r",
      "\t: 485538 lines [00:10, 44598.54 lines/s]\r",
      "\t: 490264 lines [00:10, 45363.26 lines/s]\r",
      "\t: 495081 lines [00:10, 46168.16 lines/s]\r",
      "\t: 499881 lines [00:10, 46700.86 lines/s]\r",
      "\t: 504586 lines [00:10, 46674.72 lines/s]\r",
      "\t: 509376 lines [00:11, 47035.12 lines/s]\r",
      "\t: 514219 lines [00:11, 47444.12 lines/s]\r",
      "\t: 518977 lines [00:11, 47050.99 lines/s]\r",
      "\t: 523693 lines [00:11, 46980.79 lines/s]\r",
      "\t: 528399 lines [00:11, 46936.37 lines/s]\r",
      "\t: 533158 lines [00:11, 47129.88 lines/s]\r",
      "\t: 537918 lines [00:11, 47268.74 lines/s]\r",
      "\t: 542689 lines [00:11, 47398.62 lines/s]\r",
      "\t: 547447 lines [00:11, 47450.87 lines/s]\r",
      "\t: 552194 lines [00:11, 46868.10 lines/s]\r",
      "\t: 556941 lines [00:12, 47043.07 lines/s]\r",
      "\t: 561757 lines [00:12, 47370.75 lines/s]\r",
      "\t: 566496 lines [00:12, 47182.41 lines/s]\r",
      "\t: 571269 lines [00:12, 47344.20 lines/s]\r",
      "\t: 576013 lines [00:12, 47371.29 lines/s]\r",
      "\t: 580751 lines [00:12, 39738.21 lines/s]\r",
      "\t: 585379 lines [00:12, 41495.92 lines/s]\r",
      "\t: 590087 lines [00:12, 43026.60 lines/s]\r",
      "\t: 594813 lines [00:12, 44212.79 lines/s]\r",
      "\t: 599353 lines [00:13, 44562.27 lines/s]\r",
      "\t: 604032 lines [00:13, 45207.41 lines/s]\r",
      "\t: 605248 lines [00:13, 45934.72 lines/s]\n",
      "\tFiltering out low coverage sites\n",
      "\tProcessing valid sites found\n",
      "\t\tWrite output file header\n",
      "\r",
      "\t:   0%|          | 0/18531 [00:00<?, ? sites/s]\r",
      "\t:   2%|▏         | 441/18531 [00:00<00:04, 4400.54 sites/s]\r",
      "\t:   5%|▍         | 915/18531 [00:00<00:03, 4496.24 sites/s]\r",
      "\t:   8%|▊         | 1421/18531 [00:00<00:03, 4650.24 sites/s]\r",
      "\t:  10%|█         | 1876/18531 [00:00<00:03, 4617.95 sites/s]\r",
      "\t:  13%|█▎        | 2353/18531 [00:00<00:03, 4661.88 sites/s]\r",
      "\t:  15%|█▌        | 2830/18531 [00:00<00:03, 4693.08 sites/s]\r",
      "\t:  18%|█▊        | 3338/18531 [00:00<00:03, 4800.74 sites/s]\r",
      "\t:  21%|██        | 3830/18531 [00:00<00:03, 4833.62 sites/s]\r",
      "\t:  23%|██▎       | 4312/18531 [00:00<00:02, 4827.03 sites/s]\r",
      "\t:  26%|██▌       | 4816/18531 [00:01<00:02, 4888.10 sites/s]\r",
      "\t:  29%|██▊       | 5314/18531 [00:01<00:02, 4912.95 sites/s]\r",
      "\t:  31%|███▏      | 5798/18531 [00:01<00:02, 4840.09 sites/s]\r",
      "\t:  34%|███▍      | 6293/18531 [00:01<00:02, 4871.85 sites/s]\r",
      "\t:  37%|███▋      | 6777/18531 [00:01<00:02, 4752.55 sites/s]\r",
      "\t:  39%|███▉      | 7251/18531 [00:01<00:02, 4428.79 sites/s]\r",
      "\t:  42%|████▏     | 7732/18531 [00:01<00:02, 4534.76 sites/s]\r",
      "\t:  44%|████▍     | 8189/18531 [00:01<00:02, 4452.73 sites/s]\r",
      "\t:  47%|████▋     | 8679/18531 [00:01<00:02, 4577.42 sites/s]\r",
      "\t:  50%|████▉     | 9192/18531 [00:01<00:01, 4728.90 sites/s]\r",
      "\t:  52%|█████▏    | 9698/18531 [00:02<00:01, 4822.79 sites/s]\r",
      "\t:  55%|█████▍    | 10183/18531 [00:02<00:01, 4808.49 sites/s]\r",
      "\t:  58%|█████▊    | 10680/18531 [00:02<00:01, 4854.88 sites/s]\r",
      "\t:  60%|██████    | 11167/18531 [00:02<00:01, 4808.97 sites/s]\r",
      "\t:  63%|██████▎   | 11649/18531 [00:02<00:01, 4679.59 sites/s]\r",
      "\t:  65%|██████▌   | 12119/18531 [00:03<00:06, 994.75 sites/s] \r",
      "\t:  67%|██████▋   | 12459/18531 [00:03<00:04, 1247.89 sites/s]\r",
      "\t:  70%|██████▉   | 12908/18531 [00:04<00:03, 1592.69 sites/s]\r",
      "\t:  72%|███████▏  | 13374/18531 [00:04<00:02, 1984.38 sites/s]\r",
      "\t:  75%|███████▍  | 13869/18531 [00:04<00:01, 2418.84 sites/s]\r",
      "\t:  78%|███████▊  | 14363/18531 [00:04<00:01, 2855.83 sites/s]\r",
      "\t:  80%|████████  | 14865/18531 [00:04<00:01, 3280.04 sites/s]\r",
      "\t:  83%|████████▎ | 15362/18531 [00:04<00:00, 3652.34 sites/s]\r",
      "\t:  86%|████████▌ | 15854/18531 [00:04<00:00, 3957.17 sites/s]\r",
      "\t:  88%|████████▊ | 16370/18531 [00:04<00:00, 4253.81 sites/s]\r",
      "\t:  91%|█████████ | 16860/18531 [00:04<00:00, 4355.07 sites/s]\r",
      "\t:  94%|█████████▎| 17342/18531 [00:04<00:00, 4317.58 sites/s]\r",
      "\t:  96%|█████████▌| 17806/18531 [00:05<00:00, 4401.75 sites/s]\r",
      "\t:  99%|█████████▊| 18269/18531 [00:05<00:00, 4140.44 sites/s]\r",
      "\t: 100%|██████████| 18531/18531 [00:05<00:00, 3528.30 sites/s]\n",
      "## Results summary ##\n",
      "\tTotal read lines: 605,248\n",
      "\tValid read lines: 605,248\n",
      "\tTotal sites: 229,389\n",
      "\tLow coverage sites: 210,858\n",
      "\tValid sites: 37,062\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon NanopolishComp\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --verbose -i data/freq_meth_calculate/methylation_calls.tsv -b ./output/freq_meth_calculate/out_freq_meth_calculate.bed -t ./output/freq_meth_calculate/out_freq_meth_calculate.tsv --min_depth 5 --min_llr 1\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import the package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-10T15:23:02.187329Z",
     "start_time": "2019-09-10T15:23:02.021086Z"
    }
   },
   "outputs": [],
   "source": [
    "# Import main program\n",
    "from NanopolishComp.Freq_meth_calculate import Freq_meth_calculate\n",
    "\n",
    "# Import helper functions\n",
    "from NanopolishComp.common import jhelp, head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### python API help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-10T15:23:03.716887Z",
     "start_time": "2019-09-10T15:23:03.699591Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "**NanopolishComp.Freq_meth_calculate.__init__**\n",
       "\n",
       "Calculate methylation frequency at genomic CpG sites from the output of nanopolish call-methylation\n",
       "\n",
       "---\n",
       "\n",
       "* **input_fn** *: str (required)*\n",
       "\n",
       "Path to a nanopolish call_methylation tsv output file\n",
       "\n",
       "* **fasta_index** *: str (default = )*\n",
       "\n",
       "fasta index file obtained with samtools faidx. Required for coordinate sorting\n",
       "\n",
       "* **output_bed_fn** *: str (default = )*\n",
       "\n",
       "Path to write a summary result file in BED format\n",
       "\n",
       "* **output_tsv_fn** *: str (default = )*\n",
       "\n",
       "Path to write an more extensive result report in TSV format\n",
       "\n",
       "* **min_depth** *: int (default = 10)*\n",
       "\n",
       "Minimal number of reads covering a site to be reported\n",
       "\n",
       "* **sample_id** *: str (default = )*\n",
       "\n",
       "Sample ID to be used for the bed track header\n",
       "\n",
       "* **strand_specific** *: bool (default = False)*\n",
       "\n",
       "If True, output strand specific sites\n",
       "\n",
       "* **min_llr** *: float (default = 2)*\n",
       "\n",
       "Minimal log likelyhood ratio to consider a site significantly methylated or unmethylated\n",
       "\n",
       "* **verbose** *: bool (default = False)*\n",
       "\n",
       "Increase verbosity\n",
       "\n",
       "* **quiet** *: bool (default = False)*\n",
       "\n",
       "Reduce verbosity\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(Freq_meth_calculate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-23T09:19:41.284517Z",
     "start_time": "2019-04-23T09:19:41.165804Z"
    },
    "scrolled": false
   },
   "source": [
    "#### basic setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-10T15:23:24.852979Z",
     "start_time": "2019-09-10T15:23:07.631105Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Options summary ##\n",
      "\tpackage_name: NanopolishComp\n",
      "\tpackage_version: 0.6.11\n",
      "\ttimestamp: 2019-09-10 16:23:07.646062\n",
      "\tquiet: False\n",
      "\tverbose: True\n",
      "\tmin_llr: 2\n",
      "\tstrand_specific: False\n",
      "\tsample_id: Sample1\n",
      "\tmin_depth: 10\n",
      "\toutput_tsv_fn: \n",
      "\toutput_bed_fn: ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
      "\tfasta_index: \n",
      "\tinput_fn: ./data/freq_meth_calculate/methylation_calls.tsv\n",
      "## Checking arguments ##\n",
      "\tTesting input file readability\n",
      "\tCheck output file\n",
      "\t\tOutput results in bed format\n",
      "## Parsing methylation_calls file ##\n",
      "\tStarting to parse file Nanopolish methylation call file\n",
      "\t: 605248 lines [00:15, 39476.23 lines/s]\n",
      "\tFiltering out low coverage sites\n",
      "\tProcessing valid sites found\n",
      "\t\tWrite output file header\n",
      "\t: 100%|██████████| 804/804 [00:01<00:00, 492.95 sites/s] \n",
      "## Results summary ##\n",
      "\tTotal read lines: 605,248\n",
      "\tValid read lines: 605,248\n",
      "\tTotal sites: 229,389\n",
      "\tLow coverage sites: 228,585\n",
      "\tValid sites: 1,608\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name=Methylation_Sample1 itemRgb=On\n",
      "chr-VIII\t138415\t138416\t-4387463463888030503\t-3.200000\t.\t138415\t138416\t'8,121,207'\n",
      "chr-VIII\t138429\t138430\t-4387486165434009317\t-4.722000\t.\t138429\t138430\t'8,121,207'\n",
      "chr-VIII\t212351\t212352\t-4459725007132593111\t-4.125000\t.\t212351\t212352\t'8,121,207'\n",
      "chr-VIII\t212392\t212393\t-4459458894565841458\t-1.866667\t.\t212392\t212393\t'100,100,100'\n",
      "chr-VIII\t212457\t212461\t-4459540872370764953\t-5.546667\t.\t212457\t212461\t'8,121,207'\n",
      "chr-VIII\t212530\t212531\t-4460596494258779804\t-1.850833\t.\t212530\t212531\t'100,100,100'\n",
      "chr-VIII\t212581\t212582\t-4460665860093715069\t-2.307692\t.\t212581\t212582\t'8,121,207'\n",
      "chr-VIII\t212596\t212600\t-4460684778048697414\t-5.900000\t.\t212596\t212600\t'8,121,207'\n",
      "chr-VIII\t212612\t212613\t-4460462807376904566\t-3.675385\t.\t212612\t212613\t'8,121,207'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    output_bed_fn=\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\",\n",
    "    sample_id=\"Sample1\",\n",
    "    verbose=True)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Using a fasta index for output coordinates sorting and strand specific sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-10T15:24:13.688289Z",
     "start_time": "2019-09-10T15:23:56.770922Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Options summary ##\n",
      "\tpackage_name: NanopolishComp\n",
      "\tpackage_version: 0.6.11\n",
      "\ttimestamp: 2019-09-10 16:23:56.781302\n",
      "\tquiet: False\n",
      "\tverbose: True\n",
      "\tmin_llr: 2\n",
      "\tstrand_specific: True\n",
      "\tsample_id: Sample1\n",
      "\tmin_depth: 10\n",
      "\toutput_tsv_fn: \n",
      "\toutput_bed_fn: ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
      "\tfasta_index: ./data/freq_meth_calculate/ref.fa.fai\n",
      "\tinput_fn: ./data/freq_meth_calculate/methylation_calls.tsv\n",
      "## Checking arguments ##\n",
      "\tTesting input file readability\n",
      "\tCheck output file\n",
      "\t\tOutput results in bed format\n",
      "## Parsing methylation_calls file ##\n",
      "\tStarting to parse file Nanopolish methylation call file\n",
      "\t: 605248 lines [00:15, 40181.70 lines/s]\n",
      "\tFiltering out low coverage sites\n",
      "\tSorting by coordinates\n",
      "\tProcessing valid sites found\n",
      "\t\tWrite output file header\n",
      "\t: 100%|██████████| 999/999 [00:01<00:00, 637.85 sites/s] \n",
      "## Results summary ##\n",
      "\tTotal read lines: 605,248\n",
      "\tValid read lines: 605,248\n",
      "\tTotal sites: 340,081\n",
      "\tLow coverage sites: 339,082\n",
      "\tValid sites: 1,998\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name=Methylation_Sample1 itemRgb=On\n",
      "chr-VIII\t213351\t213352\t3852928671350919267\t-0.500000\t-\t213351\t213352\t'100,100,100'\n",
      "chr-VIII\t213382\t213386\t3853101183011884366\t-5.103000\t-\t213382\t213386\t'8,121,207'\n",
      "chr-VIII\t213427\t213428\t3853155687118280343\t0.504545\t-\t213427\t213428\t'100,100,100'\n",
      "chr-VIII\t213516\t213517\t3846995005950784724\t-2.447500\t-\t213516\t213517\t'8,121,207'\n",
      "chr-VIII\t213536\t213537\t3846697452562010016\t-3.103333\t-\t213536\t213537\t'8,121,207'\n",
      "chr-VIII\t213549\t213550\t3846721371280116041\t-2.724167\t-\t213549\t213550\t'8,121,207'\n",
      "chr-VIII\t213645\t213646\t3846832673276330665\t-2.546667\t-\t213645\t213646\t'8,121,207'\n",
      "chr-VIII\t213668\t213669\t3847166582036626108\t-3.728333\t-\t213668\t213669\t'8,121,207'\n",
      "chr-VIII\t213729\t213730\t3846776908639343893\t-3.253846\t-\t213729\t213730\t'8,121,207'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    fasta_index=\"./data/freq_meth_calculate/ref.fa.fai\",\n",
    "    output_bed_fn=\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\",\n",
    "    sample_id=\"Sample1\",\n",
    "    verbose=True,\n",
    "    strand_specific=True)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Changing filtering threshold (not recommended)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-09-10T15:24:48.726621Z",
     "start_time": "2019-09-10T15:24:28.343368Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "## Parsing methylation_calls file ##\n",
      "\tStarting to parse file Nanopolish methylation call file\n",
      "\t: 605248 lines [00:14, 40874.57 lines/s]\n",
      "\tFiltering out low coverage sites\n",
      "\tProcessing valid sites found\n",
      "\t: 100%|██████████| 18531/18531 [00:05<00:00, 3458.71 sites/s]\n",
      "## Results summary ##\n",
      "\tTotal read lines: 605,248\n",
      "\tValid read lines: 605,248\n",
      "\tTotal sites: 229,389\n",
      "\tLow coverage sites: 210,858\n",
      "\tValid sites: 37,062\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome start end  strand site_id              methylated_reads unmethylated_reads ambiguous_reads sequence        num_motifs llr_list                           \n",
      "chr-I      4399  4400 .      -2730676911920365501 1                3                  1               CTTGCCGAATA     1          -3.17,-11.29,5.14,-1.8,-0.6        \n",
      "chr-I      4534  4535 .      -2730544486235489086 0                5                  1               TTATTCGTTTT     1          -3.31,0.02,-2.66,-5.82,-4.21,-2.22 \n",
      "chr-I      4591  4592 .      -2730596195312440829 0                3                  3               TCAAACGCTTG     1          -6.16,-0.6,0.28,-0.38,-6.02,-6.21  \n",
      "chr-I      4654  4655 .      -2731644250018462742 1                4                  1               TTAATCGACAA     1          -1.5,-0.67,-2.66,-2.94,-1.66,3.1   \n",
      "chr-I      1755  1756 .      -2725162958641511345 1                2                  2               TTGTTCGTTAA     1          -3.37,-0.25,-0.17,-7.47,2.73       \n",
      "chr-I      1782  1787 .      -2725136473504536062 0                5                  0               AAACCCGTTCGTAAA 2          -2.2,-3.83,-1.37,-4.8,-1.68        \n",
      "chr-I      1797  1798 .      -2724913241635744391 0                4                  1               ATTGGCGTTTG     1          -1.45,-5.09,-2.17,-6.66,-0.75      \n",
      "chr-I      1814  1815 .      -2724934681984724382 0                4                  1               GTTTGCGATAG     1          -2.2,-4.1,-1.86,-0.4,-3.71         \n",
      "chr-I      1829  1830 .      -2724872883331782055 0                4                  1               GATACCGTCCT     1          -0.16,-5.78,-3.32,-4.0,-1.22       \n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    output_tsv_fn=\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\",\n",
    "    min_llr=1,\n",
    "    min_depth=5)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "NanopolishComp",
   "language": "python",
   "name": "nanopolishcomp"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
