{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Freq_meth_calculate usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate methylation frequency at genomic CpG sites from the output of `nanopolish call-methylation`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Options"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **input_fn**\n",
    "\n",
    "Path to a `nanopolish call-methylation` tsv output file (read access required). In command line mode it is also possible to pipe the output of the `nanopolish call-methylation` directly into `Freq_meth_calculate`.\n",
    "\n",
    "* **outdir**\n",
    "\n",
    "Path to an existing directory where to write all the output files generated by `Eventalign_collapse`(write access required). If the directory does not exist an error is raised.\n",
    "\n",
    "* **outprefix**\n",
    "\n",
    "Prefix for all the files generated by the program\n",
    "\n",
    "Filtering options:\n",
    "\n",
    "* **min_llr**\n",
    "\n",
    "Set the minimal required log likelihood ratio between methylated and unmethylated state. Otherwise the value is considered ambiguous. \n",
    "\n",
    "* **min_depth**\n",
    "\n",
    "Set the minimal number of reads with non -ambiguous call per genomic position to be written in the output files.\n",
    "\n",
    "* **min_meth_freq**\n",
    "\n",
    "Set the minimal methylation frequency per genomic position to be written in the output files.\n",
    "\n",
    "* **split_group**\n",
    "\n",
    "In CpG rich regions it is not sometimes hard to distinguish which kmers is methylated in this case `nanopolish` outputs a group containing multiple CpG (or other motif). This option splits each CpG in individual genomic site. This can be useful for visualisation in a genome browser.\n",
    "\n",
    "* **motif**\n",
    "\n",
    "Specify the motif type , akin the one used for `Nanopolish call-methylation`. This is only useful if the split_group option is also selected. \n",
    "\n",
    "!!! note \"Methylation motif to sequence correspondence\"\n",
    "    * cpg → CG\n",
    "    * gpc → GC\n",
    "    * dam → GATC\n",
    "    * dcm → CCAGG\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output files format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Freq_meth_calculate` generates 2 files, a standard BED file and a tabulated file containing extra information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### BED file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Standard genomic BED6 (https://genome.ucsc.edu/FAQ/FAQformat.html#format1). The score correspond to the methylation frequency multiplied by 1000. The file is sorted by coordinates and can be rendered with a genome browser such as [IGV](https://software.broadinstitute.org/software/igv/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Tabulated TSV file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Contrary to the bed file, in the tabulated report, positions are ordered by decreasing methylation frequency.\n",
    "\n",
    "The file contains the following fields:\n",
    "\n",
    "* **chrom / start / end / strand**: Genomic coordinates of the motif or group of motifs in case split_group was not selected.\n",
    "* **site_id**: Unique integer identifier of the genomic position.\n",
    "* **methylated_reads / unmethylated_reads / ambiguous_reads**: Number of reads at a given genomic location with a higher likelyhood of being methylated or unmethylated or with an ambiguous methylation call.\n",
    "* **sequence**: -5 to +5 sequence of the motif or group of motifs in case split_group was not selected.\n",
    "* **num_motifs**: Number of motif in the group.\n",
    "* **meth_freq**: Methylation frequency (out of non anbiguous calls)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Log YAML file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The program also generates a simple log file containing the sites and positions count formated in YAML.\n",
    "\n",
    "Example file content:\n",
    "\n",
    "```\n",
    "Read sites summary:\n",
    "    total: 605,248\n",
    "    unmethylated: 571,328\n",
    "    methylated: 33,920\n",
    "Genomic positions summary:\n",
    "    total: 340,081\n",
    "    low_coverage: 339,355\n",
    "    low_meth_freq: 698\n",
    "    valid: 28\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bash command line usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T13:02:19.149219Z",
     "start_time": "2019-05-01T13:02:18.222250Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: NanopolishComp Freq_meth_calculate [-h] [-i INPUT_FN] [-o OUTDIR]\n",
      "                                          [-p OUTPREFIX] [-l MIN_LLR]\n",
      "                                          [-d MIN_DEPTH] [-f MIN_METH_FREQ]\n",
      "                                          [-s] [-m {cpg,gpc,dam,dcm}]\n",
      "                                          [-v | -q]\n",
      "\n",
      "Calculate methylation frequency at genomic CpG sites from the output of\n",
      "nanopolish call-methylation\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -v, --verbose         Increase verbosity (default: False)\n",
      "  -q, --quiet           Reduce verbosity (default: False)\n",
      "\n",
      "Input/Output options:\n",
      "  -i INPUT_FN, --input_fn INPUT_FN\n",
      "                        Path to a nanopolish call_methylation tsv output file.\n",
      "                        If not specified read from std input\n",
      "  -o OUTDIR, --outdir OUTDIR\n",
      "                        Path to the output folder (default: ./)\n",
      "  -p OUTPREFIX, --outprefix OUTPREFIX\n",
      "                        text outprefix for all the files generated (default:\n",
      "                        out)\n",
      "\n",
      "Filtering options:\n",
      "  -l MIN_LLR, --min_llr MIN_LLR\n",
      "                        Log likelihood ratio threshold (default: 2.5)\n",
      "  -d MIN_DEPTH, --min_depth MIN_DEPTH\n",
      "                        Minimal number of reads covering a site to be reported\n",
      "                        (default: 10)\n",
      "  -f MIN_METH_FREQ, --min_meth_freq MIN_METH_FREQ\n",
      "                        Minimal methylation frequency of a site to be reported\n",
      "                        (default: 0.05)\n",
      "\n",
      "Motif options:\n",
      "  -s, --split_group     If True, multi motif groups (sequence with close\n",
      "                        motifs) are split in individual site (default: False)\n",
      "  -m {cpg,gpc,dam,dcm}, --motif {cpg,gpc,dam,dcm}\n",
      "                        Methylation motif type (default: cpg)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From an existing nanopolish call_methylation file output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T13:02:39.328276Z",
     "start_time": "2019-05-01T13:02:19.152991Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t216740\t000090\t-\n",
      "chr-XII\t451976\t451980\t216746\t000071\t-\n",
      "chr-XII\t452137\t452138\t216753\t000050\t-\n",
      "chr-XII\t452217\t452221\t216756\t000272\t-\n",
      "chr-XII\t452331\t452341\t216384\t000062\t+\n",
      "chr-XII\t452632\t452633\t216391\t000066\t+\n",
      "chr-XII\t452654\t452655\t216769\t000100\t-\n",
      "chr-XII\t453086\t453087\t216781\t000055\t-\n",
      "chr-XII\t453336\t453337\t216790\t000066\t-\n",
      "chrom\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tmeth_freq\n",
      "chr-XII\t465335\t465339\t-\t217103\t6\t5\t29\tCAGCACGACGGAGT\t2\t0.545455\n",
      "chr-XII\t456198\t456202\t-\t216869\t7\t7\t32\tCAGCACGACGGAGT\t2\t0.500000\n",
      "chr-XII\t460773\t460774\t+\t216597\t17\t20\t77\tCTCATCGTATA\t1\t0.459459\n",
      "chr-XII\t457104\t457105\t+\t216519\t6\t8\t43\tAACTACGAGCT\t1\t0.428571\n",
      "chr-XII\t461274\t461275\t+\t216614\t5\t7\t85\tTTATCCGAATG\t1\t0.416667\n",
      "chr-XII\t463406\t463407\t+\t216682\t6\t12\t29\tTCTTTCGCCCC\t1\t0.333333\n",
      "chr-XII\t457887\t457888\t-\t216914\t6\t12\t53\tGCTGCCGGAAA\t1\t0.333333\n",
      "chr-XII\t461077\t461078\t+\t216605\t8\t17\t74\tCTTTCCGCCAA\t1\t0.320000\n",
      "chr-XII\t462283\t462284\t-\t217018\t5\t11\t57\tAGGCACGCAAG\t1\t0.312500\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "\r",
      "0 lines [00:00, ? lines/s]\r",
      "3694 lines [00:00, 36936.86 lines/s]\r",
      "7482 lines [00:00, 37212.28 lines/s]\r",
      "10734 lines [00:00, 35667.60 lines/s]\r",
      "14440 lines [00:00, 36072.69 lines/s]\r",
      "18037 lines [00:00, 36040.67 lines/s]\r",
      "21625 lines [00:00, 35991.14 lines/s]\r",
      "25336 lines [00:00, 36317.90 lines/s]\r",
      "28669 lines [00:00, 34892.13 lines/s]\r",
      "32294 lines [00:00, 35288.45 lines/s]\r",
      "35778 lines [00:01, 35152.49 lines/s]\r",
      "39417 lines [00:01, 35514.70 lines/s]\r",
      "43273 lines [00:01, 36376.40 lines/s]\r",
      "47048 lines [00:01, 36775.78 lines/s]\r",
      "50790 lines [00:01, 36966.16 lines/s]\r",
      "54468 lines [00:01, 35695.92 lines/s]\r",
      "58036 lines [00:01, 35689.11 lines/s]\r",
      "61667 lines [00:01, 35871.98 lines/s]\r",
      "65444 lines [00:01, 36420.06 lines/s]\r",
      "69087 lines [00:01, 35966.06 lines/s]\r",
      "72709 lines [00:02, 36040.00 lines/s]\r",
      "76315 lines [00:02, 35726.97 lines/s]\r",
      "79890 lines [00:02, 35371.23 lines/s]\r",
      "83509 lines [00:02, 35611.67 lines/s]\r",
      "87072 lines [00:02, 35528.83 lines/s]\r",
      "90688 lines [00:02, 35710.92 lines/s]\r",
      "94365 lines [00:02, 36016.51 lines/s]\r",
      "97983 lines [00:02, 36064.36 lines/s]\r",
      "101591 lines [00:02, 33760.63 lines/s]\r",
      "105174 lines [00:02, 34353.89 lines/s]\r",
      "108792 lines [00:03, 34879.50 lines/s]\r",
      "112567 lines [00:03, 35691.17 lines/s]\r",
      "116261 lines [00:03, 36054.73 lines/s]\r",
      "120017 lines [00:03, 36491.47 lines/s]\r",
      "123677 lines [00:03, 36430.93 lines/s]\r",
      "127328 lines [00:03, 36100.36 lines/s]\r",
      "130959 lines [00:03, 36161.31 lines/s]\r",
      "134580 lines [00:03, 33640.59 lines/s]\r",
      "138049 lines [00:03, 33423.56 lines/s]\r",
      "141674 lines [00:03, 34222.01 lines/s]\r",
      "145337 lines [00:04, 34908.34 lines/s]\r",
      "148868 lines [00:04, 35026.25 lines/s]\r",
      "152514 lines [00:04, 35443.32 lines/s]\r",
      "156112 lines [00:04, 35600.85 lines/s]\r",
      "159680 lines [00:04, 35429.19 lines/s]\r",
      "163270 lines [00:04, 35567.61 lines/s]\r",
      "166845 lines [00:04, 35622.11 lines/s]\r",
      "170411 lines [00:04, 35273.52 lines/s]\r",
      "173942 lines [00:04, 32431.49 lines/s]\r",
      "177639 lines [00:05, 33669.34 lines/s]\r",
      "181415 lines [00:05, 34799.97 lines/s]\r",
      "185157 lines [00:05, 35545.07 lines/s]\r",
      "188925 lines [00:05, 36158.67 lines/s]\r",
      "192567 lines [00:05, 36218.82 lines/s]\r",
      "196207 lines [00:05, 35699.74 lines/s]\r",
      "199945 lines [00:05, 36185.90 lines/s]\r",
      "203576 lines [00:05, 36182.46 lines/s]\r",
      "207305 lines [00:05, 36507.00 lines/s]\r",
      "211027 lines [00:05, 36717.97 lines/s]\r",
      "214833 lines [00:06, 37109.69 lines/s]\r",
      "218549 lines [00:06, 37059.64 lines/s]\r",
      "222258 lines [00:06, 36918.24 lines/s]\r",
      "225970 lines [00:06, 36977.42 lines/s]\r",
      "229670 lines [00:06, 35383.21 lines/s]\r",
      "233232 lines [00:06, 35453.12 lines/s]\r",
      "236872 lines [00:06, 35730.43 lines/s]\r",
      "240454 lines [00:06, 35540.80 lines/s]\r",
      "244015 lines [00:06, 35449.02 lines/s]\r",
      "247737 lines [00:06, 35960.06 lines/s]\r",
      "251443 lines [00:07, 36281.61 lines/s]\r",
      "255091 lines [00:07, 36339.07 lines/s]\r",
      "258728 lines [00:07, 32742.21 lines/s]\r",
      "262144 lines [00:07, 33154.93 lines/s]\r",
      "265693 lines [00:07, 33820.17 lines/s]\r",
      "269159 lines [00:07, 34065.43 lines/s]\r",
      "272821 lines [00:07, 34792.55 lines/s]\r",
      "276433 lines [00:07, 35179.58 lines/s]\r",
      "280126 lines [00:07, 35684.45 lines/s]\r",
      "283709 lines [00:07, 35661.31 lines/s]\r",
      "287285 lines [00:08, 33375.47 lines/s]\r",
      "290659 lines [00:08, 33333.65 lines/s]\r",
      "294302 lines [00:08, 34205.14 lines/s]\r",
      "298022 lines [00:08, 35050.44 lines/s]\r",
      "301549 lines [00:08, 35071.13 lines/s]\r",
      "305118 lines [00:08, 35253.09 lines/s]\r",
      "308771 lines [00:08, 35625.90 lines/s]\r",
      "312343 lines [00:08, 35633.99 lines/s]\r",
      "315943 lines [00:08, 35738.50 lines/s]\r",
      "319569 lines [00:09, 35892.24 lines/s]\r",
      "323247 lines [00:09, 36151.43 lines/s]\r",
      "326981 lines [00:09, 36496.16 lines/s]\r",
      "330634 lines [00:09, 31874.39 lines/s]\r",
      "334275 lines [00:09, 33110.19 lines/s]\r",
      "337674 lines [00:09, 33053.07 lines/s]\r",
      "341275 lines [00:09, 33884.76 lines/s]\r",
      "344879 lines [00:09, 34502.15 lines/s]\r",
      "348424 lines [00:09, 34780.29 lines/s]\r",
      "352198 lines [00:09, 35615.65 lines/s]\r",
      "355878 lines [00:10, 35961.00 lines/s]\r",
      "359819 lines [00:10, 36929.30 lines/s]\r",
      "363741 lines [00:10, 37587.85 lines/s]\r",
      "367650 lines [00:10, 38025.18 lines/s]\r",
      "371688 lines [00:10, 38700.32 lines/s]\r",
      "375669 lines [00:10, 39022.03 lines/s]\r",
      "379580 lines [00:10, 38665.41 lines/s]\r",
      "383454 lines [00:10, 37909.85 lines/s]\r",
      "387254 lines [00:10, 37566.73 lines/s]\r",
      "391018 lines [00:11, 33373.23 lines/s]\r",
      "394447 lines [00:11, 31057.74 lines/s]\r",
      "397649 lines [00:11, 30072.43 lines/s]\r",
      "400731 lines [00:11, 30017.29 lines/s]\r",
      "404120 lines [00:11, 31081.55 lines/s]\r",
      "407546 lines [00:11, 31970.25 lines/s]\r",
      "410782 lines [00:11, 32008.58 lines/s]\r",
      "414132 lines [00:11, 32440.38 lines/s]\r",
      "417397 lines [00:11, 31907.29 lines/s]\r",
      "420604 lines [00:11, 31911.86 lines/s]\r",
      "423807 lines [00:12, 31935.99 lines/s]\r",
      "427009 lines [00:12, 28558.08 lines/s]\r",
      "429938 lines [00:12, 25030.87 lines/s]\r",
      "432571 lines [00:12, 25389.36 lines/s]\r",
      "435769 lines [00:12, 27059.71 lines/s]\r",
      "439129 lines [00:12, 28737.11 lines/s]\r",
      "442252 lines [00:12, 29437.09 lines/s]\r",
      "445597 lines [00:12, 30534.80 lines/s]\r",
      "449148 lines [00:12, 31867.77 lines/s]\r",
      "452498 lines [00:13, 32339.00 lines/s]\r",
      "455774 lines [00:13, 27941.41 lines/s]\r",
      "458704 lines [00:13, 22947.88 lines/s]\r",
      "461241 lines [00:13, 19850.32 lines/s]\r",
      "463470 lines [00:13, 16689.86 lines/s]\r",
      "465395 lines [00:13, 16843.70 lines/s]\r",
      "468681 lines [00:13, 19727.59 lines/s]\r",
      "470972 lines [00:14, 20083.04 lines/s]\r",
      "473205 lines [00:14, 19638.85 lines/s]\r",
      "475328 lines [00:14, 16863.04 lines/s]\r",
      "478677 lines [00:14, 19812.92 lines/s]\r",
      "482100 lines [00:14, 22675.51 lines/s]\r",
      "485442 lines [00:14, 25095.07 lines/s]\r",
      "488618 lines [00:14, 26779.94 lines/s]\r",
      "491807 lines [00:14, 28129.99 lines/s]\r",
      "495278 lines [00:14, 29826.05 lines/s]\r",
      "498607 lines [00:15, 30786.46 lines/s]\r",
      "502111 lines [00:15, 31943.33 lines/s]\r",
      "505407 lines [00:15, 31808.01 lines/s]\r",
      "508660 lines [00:15, 28152.30 lines/s]\r",
      "511603 lines [00:15, 22895.36 lines/s]\r",
      "514142 lines [00:15, 19749.39 lines/s]\r",
      "517566 lines [00:15, 22621.03 lines/s]\r",
      "520968 lines [00:15, 25147.59 lines/s]\r",
      "524491 lines [00:16, 27509.06 lines/s]\r",
      "528192 lines [00:16, 29802.64 lines/s]\r",
      "531429 lines [00:16, 30348.87 lines/s]\r",
      "534646 lines [00:16, 28593.62 lines/s]\r",
      "538028 lines [00:16, 29982.41 lines/s]\r",
      "541653 lines [00:16, 31621.69 lines/s]\r",
      "544925 lines [00:16, 30578.96 lines/s]\r",
      "548067 lines [00:16, 30175.37 lines/s]\r",
      "551405 lines [00:16, 31068.80 lines/s]\r",
      "554561 lines [00:17, 31195.66 lines/s]\r",
      "558077 lines [00:17, 32286.50 lines/s]\r",
      "561718 lines [00:17, 33420.60 lines/s]\r",
      "565178 lines [00:17, 33762.17 lines/s]\r",
      "568579 lines [00:17, 33364.23 lines/s]\r",
      "572187 lines [00:17, 34133.27 lines/s]\r",
      "575719 lines [00:17, 34479.28 lines/s]\r",
      "579417 lines [00:17, 35192.23 lines/s]\r",
      "582949 lines [00:17, 31020.06 lines/s]\r",
      "586151 lines [00:18, 28249.02 lines/s]\r",
      "589415 lines [00:18, 29435.73 lines/s]\r",
      "592937 lines [00:18, 30959.07 lines/s]\r",
      "596121 lines [00:18, 23711.43 lines/s]\r",
      "599915 lines [00:18, 26716.08 lines/s]\r",
      "603872 lines [00:18, 29599.60 lines/s]\r",
      "605248 lines [00:18, 32458.39 lines/s]\n",
      "Read sites summary\n",
      "\ttotal: 605,248\n",
      "\tunmethylated: 596,963\n",
      "\tmethylated: 8,285\n",
      "\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "\r",
      "  0%|          | 0/340081 [00:00<?, ? positions/s]\r",
      " 17%|█▋        | 58949/340081 [00:00<00:00, 589483.53 positions/s]\r",
      " 34%|███▎      | 114624/340081 [00:00<00:00, 579266.07 positions/s]\r",
      " 53%|█████▎    | 180685/340081 [00:00<00:00, 601484.14 positions/s]\r",
      " 72%|███████▏  | 243494/340081 [00:00<00:00, 609224.65 positions/s]\r",
      " 89%|████████▉ | 303393/340081 [00:00<00:00, 606117.02 positions/s]\r",
      "100%|██████████| 340081/340081 [00:00<00:00, 611981.36 positions/s]\n",
      "Genomic positions summary\n",
      "\ttotal: 340,081\n",
      "\tlow_coverage: 339,292\n",
      "\tlow_meth_freq: 647\n",
      "\tvalid: 142\n",
      "\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --verbose -i data/freq_meth_calculate/methylation_calls.tsv -o ./output/freq_meth_calculate/\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From standard input to a file (the same can be done with Nanopolish call-methylation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example the split_group option was also selected leading to motif groups to be split in individual positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T13:03:02.252682Z",
     "start_time": "2019-05-01T13:02:39.331625Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t289557\t000090\t-\n",
      "chr-XII\t451976\t451977\t289563\t000071\t-\n",
      "chr-XII\t451979\t451980\t289564\t000071\t-\n",
      "chr-XII\t452137\t452138\t289574\t000050\t-\n",
      "chr-XII\t452217\t452218\t289577\t000272\t-\n",
      "chr-XII\t452220\t452221\t289578\t000272\t-\n",
      "chr-XII\t452331\t452332\t288999\t000062\t+\n",
      "chr-XII\t452340\t452341\t289000\t000062\t+\n",
      "chr-XII\t452632\t452633\t289008\t000066\t+\n",
      "chrom\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tmeth_freq\n",
      "chr-XII\t465335\t465336\t-\t290122\t6\t5\t29\tCAGCACGACGG\t2\t0.545455\n",
      "chr-XII\t465338\t465339\t-\t290123\t6\t5\t29\tCACGACGGAGT\t2\t0.545455\n",
      "chr-XII\t456198\t456199\t-\t289760\t7\t7\t32\tCAGCACGACGG\t2\t0.500000\n",
      "chr-XII\t456201\t456202\t-\t289761\t7\t7\t32\tCACGACGGAGT\t2\t0.500000\n",
      "chr-XII\t460773\t460774\t+\t289338\t17\t20\t77\tCTCATCGTATA\t1\t0.459459\n",
      "chr-XII\t457104\t457105\t+\t289216\t6\t8\t43\tAACTACGAGCT\t1\t0.428571\n",
      "chr-XII\t461274\t461275\t+\t289359\t5\t7\t85\tTTATCCGAATG\t1\t0.416667\n",
      "chr-XII\t463406\t463407\t+\t289460\t6\t12\t29\tTCTTTCGCCCC\t1\t0.333333\n",
      "chr-XII\t457887\t457888\t-\t289825\t6\t12\t53\tGCTGCCGGAAA\t1\t0.333333\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "\r",
      "0 lines [00:00, ? lines/s]\r",
      "3282 lines [00:00, 32814.09 lines/s]\r",
      "6234 lines [00:00, 31750.91 lines/s]\r",
      "9278 lines [00:00, 31344.65 lines/s]\r",
      "12110 lines [00:00, 30367.92 lines/s]\r",
      "15364 lines [00:00, 30987.47 lines/s]\r",
      "18809 lines [00:00, 31949.70 lines/s]\r",
      "21634 lines [00:00, 30414.05 lines/s]\r",
      "25200 lines [00:00, 31817.59 lines/s]\r",
      "28261 lines [00:00, 31441.90 lines/s]\r",
      "31505 lines [00:01, 31732.81 lines/s]\r",
      "34605 lines [00:01, 30795.90 lines/s]\r",
      "37639 lines [00:01, 30308.92 lines/s]\r",
      "40944 lines [00:01, 31079.95 lines/s]\r",
      "44179 lines [00:01, 31448.08 lines/s]\r",
      "47314 lines [00:01, 31174.16 lines/s]\r",
      "50425 lines [00:01, 27709.43 lines/s]\r",
      "53263 lines [00:01, 27204.76 lines/s]\r",
      "56372 lines [00:01, 28263.75 lines/s]\r",
      "59320 lines [00:01, 28616.13 lines/s]\r",
      "62480 lines [00:02, 29448.88 lines/s]\r",
      "65787 lines [00:02, 30443.95 lines/s]\r",
      "69004 lines [00:02, 30940.12 lines/s]\r",
      "72120 lines [00:02, 30005.95 lines/s]\r",
      "75143 lines [00:02, 28290.67 lines/s]\r",
      "78194 lines [00:02, 28916.94 lines/s]\r",
      "81114 lines [00:02, 24024.93 lines/s]\r",
      "84192 lines [00:02, 25716.14 lines/s]\r",
      "87393 lines [00:02, 27328.07 lines/s]\r",
      "90503 lines [00:03, 28356.96 lines/s]\r",
      "93439 lines [00:03, 28068.85 lines/s]\r",
      "96317 lines [00:03, 28089.15 lines/s]\r",
      "99176 lines [00:03, 25173.86 lines/s]\r",
      "101785 lines [00:03, 24153.22 lines/s]\r",
      "104740 lines [00:03, 25551.48 lines/s]\r",
      "108121 lines [00:03, 27571.58 lines/s]\r",
      "111498 lines [00:03, 29176.71 lines/s]\r",
      "114748 lines [00:03, 30098.43 lines/s]\r",
      "117980 lines [00:04, 30731.91 lines/s]\r",
      "121338 lines [00:04, 31529.11 lines/s]\r",
      "124534 lines [00:04, 30097.86 lines/s]\r",
      "127588 lines [00:04, 24138.21 lines/s]\r",
      "130221 lines [00:04, 22936.40 lines/s]\r",
      "133453 lines [00:04, 25123.94 lines/s]\r",
      "136800 lines [00:04, 27153.38 lines/s]\r",
      "140159 lines [00:04, 28807.95 lines/s]\r",
      "143497 lines [00:04, 30039.88 lines/s]\r",
      "146672 lines [00:05, 30528.83 lines/s]\r",
      "149818 lines [00:05, 30801.32 lines/s]\r",
      "152958 lines [00:05, 30432.86 lines/s]\r",
      "156044 lines [00:05, 29857.42 lines/s]\r",
      "159384 lines [00:05, 30838.13 lines/s]\r",
      "162703 lines [00:05, 31506.08 lines/s]\r",
      "165879 lines [00:05, 30730.05 lines/s]\r",
      "168974 lines [00:05, 29823.46 lines/s]\r",
      "172208 lines [00:05, 30535.50 lines/s]\r",
      "175596 lines [00:06, 31465.46 lines/s]\r",
      "178763 lines [00:06, 30884.74 lines/s]\r",
      "181876 lines [00:06, 30957.70 lines/s]\r",
      "184984 lines [00:06, 24718.86 lines/s]\r",
      "188028 lines [00:06, 26195.52 lines/s]\r",
      "191388 lines [00:06, 28048.74 lines/s]\r",
      "194413 lines [00:06, 28673.42 lines/s]\r",
      "197776 lines [00:06, 29999.28 lines/s]\r",
      "201426 lines [00:06, 31692.15 lines/s]\r",
      "204688 lines [00:07, 31119.05 lines/s]\r",
      "207867 lines [00:07, 29530.65 lines/s]\r",
      "210928 lines [00:07, 29844.04 lines/s]\r",
      "213958 lines [00:07, 28334.75 lines/s]\r",
      "217139 lines [00:07, 29294.38 lines/s]\r",
      "220156 lines [00:07, 29549.18 lines/s]\r",
      "223629 lines [00:07, 30932.37 lines/s]\r",
      "226797 lines [00:07, 31151.34 lines/s]\r",
      "229939 lines [00:07, 30306.25 lines/s]\r",
      "233104 lines [00:07, 30696.08 lines/s]\r",
      "236192 lines [00:08, 30518.64 lines/s]\r",
      "239257 lines [00:08, 30479.17 lines/s]\r",
      "242369 lines [00:08, 30668.44 lines/s]\r",
      "245443 lines [00:08, 30542.27 lines/s]\r",
      "248747 lines [00:08, 31250.12 lines/s]\r",
      "251880 lines [00:08, 26934.98 lines/s]\r",
      "255002 lines [00:08, 28091.47 lines/s]\r",
      "258398 lines [00:08, 29625.78 lines/s]\r",
      "261521 lines [00:08, 30087.20 lines/s]\r",
      "264667 lines [00:09, 30486.24 lines/s]\r",
      "267760 lines [00:09, 29674.64 lines/s]\r",
      "270848 lines [00:09, 30024.21 lines/s]\r",
      "274075 lines [00:09, 30660.52 lines/s]\r",
      "277163 lines [00:09, 30380.00 lines/s]\r",
      "280414 lines [00:09, 30987.70 lines/s]\r",
      "283791 lines [00:09, 31771.99 lines/s]\r",
      "286983 lines [00:09, 31713.61 lines/s]\r",
      "290260 lines [00:09, 32022.96 lines/s]\r",
      "293626 lines [00:09, 32495.85 lines/s]\r",
      "296883 lines [00:10, 32005.31 lines/s]\r",
      "300418 lines [00:10, 32939.03 lines/s]\r",
      "303724 lines [00:10, 30819.76 lines/s]\r",
      "307165 lines [00:10, 31814.35 lines/s]\r",
      "310591 lines [00:10, 32508.96 lines/s]\r",
      "313869 lines [00:10, 31854.02 lines/s]\r",
      "317077 lines [00:10, 31058.70 lines/s]\r",
      "320203 lines [00:10, 26259.21 lines/s]\r",
      "323638 lines [00:10, 28255.18 lines/s]\r",
      "326606 lines [00:11, 27723.79 lines/s]\r",
      "329727 lines [00:11, 28683.07 lines/s]\r",
      "332858 lines [00:11, 29422.83 lines/s]\r",
      "335904 lines [00:11, 29725.71 lines/s]\r",
      "339373 lines [00:11, 31058.59 lines/s]\r",
      "342627 lines [00:11, 31488.07 lines/s]\r",
      "346156 lines [00:11, 32539.58 lines/s]\r",
      "349443 lines [00:11, 31582.80 lines/s]\r",
      "352853 lines [00:11, 32295.77 lines/s]\r",
      "356321 lines [00:11, 32974.89 lines/s]\r",
      "359639 lines [00:12, 31763.10 lines/s]\r",
      "363011 lines [00:12, 32323.60 lines/s]\r",
      "366263 lines [00:12, 31565.45 lines/s]\r",
      "369814 lines [00:12, 32642.75 lines/s]\r",
      "373100 lines [00:12, 30212.59 lines/s]\r",
      "376829 lines [00:12, 32036.39 lines/s]\r",
      "380219 lines [00:12, 32571.86 lines/s]\r",
      "383779 lines [00:12, 33424.43 lines/s]\r",
      "387726 lines [00:12, 35032.41 lines/s]\r",
      "391395 lines [00:13, 35512.01 lines/s]\r",
      "394982 lines [00:13, 35334.08 lines/s]\r",
      "398541 lines [00:13, 32888.23 lines/s]\r",
      "401904 lines [00:13, 33104.91 lines/s]\r",
      "405251 lines [00:13, 33164.72 lines/s]\r",
      "408593 lines [00:13, 32386.69 lines/s]\r",
      "411854 lines [00:13, 31203.65 lines/s]\r",
      "414999 lines [00:13, 30493.53 lines/s]\r",
      "418070 lines [00:13, 29606.67 lines/s]\r",
      "421051 lines [00:14, 26293.03 lines/s]\r",
      "423885 lines [00:14, 26873.51 lines/s]\r",
      "426777 lines [00:14, 27454.83 lines/s]\r",
      "429569 lines [00:14, 24275.93 lines/s]\r",
      "432833 lines [00:14, 26295.63 lines/s]\r",
      "435607 lines [00:14, 26706.65 lines/s]\r",
      "438363 lines [00:14, 23947.07 lines/s]\r",
      "440871 lines [00:14, 24270.50 lines/s]\r",
      "444162 lines [00:14, 26344.96 lines/s]\r",
      "447595 lines [00:15, 28320.27 lines/s]\r",
      "450902 lines [00:15, 29592.98 lines/s]\r",
      "454188 lines [00:15, 30500.95 lines/s]\r",
      "457538 lines [00:15, 31337.45 lines/s]\r",
      "460730 lines [00:15, 26328.89 lines/s]\r",
      "463986 lines [00:15, 27932.49 lines/s]\r",
      "467108 lines [00:15, 28842.01 lines/s]\r",
      "470105 lines [00:15, 28704.76 lines/s]\r",
      "473055 lines [00:15, 24677.92 lines/s]\r",
      "475678 lines [00:16, 18569.86 lines/s]\r",
      "477867 lines [00:16, 18374.53 lines/s]\r",
      "479938 lines [00:16, 18357.07 lines/s]\r",
      "481937 lines [00:16, 18270.97 lines/s]\r",
      "484002 lines [00:16, 18924.52 lines/s]\r",
      "487129 lines [00:16, 21465.64 lines/s]\r",
      "489662 lines [00:16, 22494.12 lines/s]\r",
      "492793 lines [00:16, 24568.67 lines/s]\r",
      "496167 lines [00:17, 26749.67 lines/s]\r",
      "499376 lines [00:17, 28154.17 lines/s]\r",
      "502839 lines [00:17, 29824.69 lines/s]\r",
      "506227 lines [00:17, 30934.21 lines/s]\r",
      "509504 lines [00:17, 31462.21 lines/s]\r",
      "512719 lines [00:17, 30783.85 lines/s]\r",
      "515849 lines [00:17, 30875.66 lines/s]\r",
      "519236 lines [00:17, 31714.53 lines/s]\r",
      "522712 lines [00:17, 32569.89 lines/s]\r",
      "526167 lines [00:17, 33137.66 lines/s]\r",
      "529694 lines [00:18, 33749.38 lines/s]\r",
      "533122 lines [00:18, 33905.29 lines/s]\r",
      "536541 lines [00:18, 33989.82 lines/s]\r",
      "539950 lines [00:18, 34018.00 lines/s]\r",
      "543359 lines [00:18, 33593.37 lines/s]\r",
      "546733 lines [00:18, 33636.35 lines/s]\r",
      "550101 lines [00:18, 33558.36 lines/s]\r",
      "553463 lines [00:18, 33575.74 lines/s]\r",
      "557078 lines [00:18, 34308.13 lines/s]\r",
      "560599 lines [00:18, 34572.94 lines/s]\r",
      "564061 lines [00:19, 34354.57 lines/s]\r",
      "567500 lines [00:19, 34061.57 lines/s]\r",
      "570910 lines [00:19, 26652.70 lines/s]\r",
      "574427 lines [00:19, 28738.58 lines/s]\r",
      "577882 lines [00:19, 30265.41 lines/s]\r",
      "581370 lines [00:19, 31511.44 lines/s]\r",
      "584659 lines [00:19, 30870.82 lines/s]\r",
      "587846 lines [00:19, 26775.87 lines/s]\r",
      "590690 lines [00:20, 25281.87 lines/s]\r",
      "593352 lines [00:20, 25216.99 lines/s]\r",
      "595968 lines [00:20, 22122.82 lines/s]\r",
      "598315 lines [00:20, 19582.04 lines/s]\r",
      "600421 lines [00:20, 19155.34 lines/s]\r",
      "603093 lines [00:20, 20932.77 lines/s]\r",
      "605248 lines [00:20, 29158.68 lines/s]\n",
      "Read sites summary\n",
      "\ttotal: 605,248\n",
      "\tunmethylated: 596,963\n",
      "\tmethylated: 8,285\n",
      "\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "\r",
      "  0%|          | 0/455542 [00:00<?, ? positions/s]\r",
      "  8%|▊         | 34772/455542 [00:00<00:01, 347714.53 positions/s]\r",
      " 16%|█▌        | 71612/455542 [00:00<00:01, 353669.75 positions/s]\r",
      " 27%|██▋       | 123116/455542 [00:00<00:00, 390360.74 positions/s]\r",
      " 41%|████▏     | 188103/455542 [00:00<00:00, 443488.68 positions/s]\r",
      " 55%|█████▌    | 252143/455542 [00:00<00:00, 488555.07 positions/s]\r",
      " 68%|██████▊   | 311299/455542 [00:00<00:00, 515479.70 positions/s]\r",
      " 79%|███████▉  | 361754/455542 [00:00<00:00, 448347.45 positions/s]\r",
      " 89%|████████▉ | 407388/455542 [00:00<00:00, 416954.91 positions/s]\r",
      "100%|██████████| 455542/455542 [00:00<00:00, 469478.36 positions/s]\n",
      "Genomic positions summary\n",
      "\ttotal: 455,542\n",
      "\tlow_coverage: 454,219\n",
      "\tlow_meth_freq: 1,148\n",
      "\tvalid: 175\n",
      "\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "cat data/freq_meth_calculate/methylation_calls.tsv | NanopolishComp Freq_meth_calculate --verbose -o ./output/freq_meth_calculate/ --split_group\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import the package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T13:03:02.266812Z",
     "start_time": "2019-05-01T13:03:02.257612Z"
    }
   },
   "outputs": [],
   "source": [
    "# Import main program\n",
    "from NanopolishComp.Freq_meth_calculate import Freq_meth_calculate\n",
    "\n",
    "# Import helper functions\n",
    "from NanopolishComp.common import jhelp, head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### python API help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T13:03:02.298535Z",
     "start_time": "2019-05-01T13:03:02.269396Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "**NanopolishComp.Freq_meth_calculate.__init__**\n",
       "\n",
       "Calculate methylation frequency at genomic CpG sites from the output of nanopolish call-methylation\n",
       "\n",
       "---\n",
       "\n",
       "* **input_fn** *: str (required)*\n",
       "\n",
       "Path to a nanopolish call_methylation tsv output file\n",
       "\n",
       "* **outdir** *: str (default = ./)*\n",
       "\n",
       "Path to the output folder\n",
       "\n",
       "* **outprefix** *: str (default = out)*\n",
       "\n",
       "text outprefix for all the files generated\n",
       "\n",
       "* **min_llr** *: float (default = 2.5)*\n",
       "\n",
       "Log likelihood ratio threshold\n",
       "\n",
       "* **min_depth** *: int (default = 10)*\n",
       "\n",
       "Minimal number of reads covering a site to be reported\n",
       "\n",
       "* **min_meth_freq** *: float (default = 0.05)*\n",
       "\n",
       "Minimal methylation frequency of a site to be reported\n",
       "\n",
       "* **split_group** *: bool (default = False)*\n",
       "\n",
       "If True, multi motif groups (sequence with close motifs) are split in individual site\n",
       "\n",
       "* **motif** *: {cpg,gpc,dam,dcm} (default = cpg)*\n",
       "\n",
       "Methylation motif type\n",
       "\n",
       "* **verbose** *: bool (default = False)*\n",
       "\n",
       "Increase verbosity\n",
       "\n",
       "* **quiet** *: bool (default = False)*\n",
       "\n",
       "Reduce verbosity\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(Freq_meth_calculate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-23T09:19:41.284517Z",
     "start_time": "2019-04-23T09:19:41.165804Z"
    },
    "scrolled": false
   },
   "source": [
    "#### basic setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T13:03:25.161440Z",
     "start_time": "2019-05-01T13:03:02.306825Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:22, 27351.10 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 340081/340081 [00:00<00:00, 544098.47 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read sites summary:        \n",
      "total: 605,248             \n",
      "unmethylated: 596,963      \n",
      "methylated: 8,285          \n",
      "Genomic positions summary: \n",
      "total: 340,081             \n",
      "low_coverage: 339,292      \n",
      "low_meth_freq: 647         \n",
      "valid: 142                 \n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence       num_motifs meth_freq \n",
      "chr-XII 465335 465339 -      217103  6                5                  29              CAGCACGACGGAGT 2          0.545455  \n",
      "chr-XII 456198 456202 -      216869  7                7                  32              CAGCACGACGGAGT 2          0.500000  \n",
      "chr-XII 460773 460774 +      216597  17               20                 77              CTCATCGTATA    1          0.459459  \n",
      "chr-XII 457104 457105 +      216519  6                8                  43              AACTACGAGCT    1          0.428571  \n",
      "chr-XII 461274 461275 +      216614  5                7                  85              TTATCCGAATG    1          0.416667  \n",
      "chr-XII 463406 463407 +      216682  6                12                 29              TCTTTCGCCCC    1          0.333333  \n",
      "chr-XII 457887 457888 -      216914  6                12                 53              GCTGCCGGAAA    1          0.333333  \n",
      "chr-XII 461077 461078 +      216605  8                17                 74              CTTTCCGCCAA    1          0.320000  \n",
      "chr-XII 462283 462284 -      217018  5                11                 57              AGGCACGCAAG    1          0.312500  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t216740\t000090\t-\n",
      "chr-XII\t451976\t451980\t216746\t000071\t-\n",
      "chr-XII\t452137\t452138\t216753\t000050\t-\n",
      "chr-XII\t452217\t452221\t216756\t000272\t-\n",
      "chr-XII\t452331\t452341\t216384\t000062\t+\n",
      "chr-XII\t452632\t452633\t216391\t000066\t+\n",
      "chr-XII\t452654\t452655\t216769\t000100\t-\n",
      "chr-XII\t453086\t453087\t216781\t000055\t-\n",
      "chr-XII\t453336\t453337\t216790\t000066\t-\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\")\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.log\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### split motif groups in individual genomic position "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T13:03:45.918768Z",
     "start_time": "2019-05-01T13:03:25.163340Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:19, 30511.55 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 455542/455542 [00:00<00:00, 569486.09 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read sites summary:        \n",
      "total: 605,248             \n",
      "unmethylated: 596,963      \n",
      "methylated: 8,285          \n",
      "Genomic positions summary: \n",
      "total: 455,542             \n",
      "low_coverage: 454,219      \n",
      "low_meth_freq: 1,148       \n",
      "valid: 175                 \n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence    num_motifs meth_freq \n",
      "chr-XII 465335 465336 -      630203  6                5                  29              CAGCACGACGG 2          0.545455  \n",
      "chr-XII 465338 465339 -      630204  6                5                  29              CACGACGGAGT 2          0.545455  \n",
      "chr-XII 456198 456199 -      629841  7                7                  32              CAGCACGACGG 2          0.500000  \n",
      "chr-XII 456201 456202 -      629842  7                7                  32              CACGACGGAGT 2          0.500000  \n",
      "chr-XII 460773 460774 +      629419  17               20                 77              CTCATCGTATA 1          0.459459  \n",
      "chr-XII 457104 457105 +      629297  6                8                  43              AACTACGAGCT 1          0.428571  \n",
      "chr-XII 461274 461275 +      629440  5                7                  85              TTATCCGAATG 1          0.416667  \n",
      "chr-XII 463406 463407 +      629541  6                12                 29              TCTTTCGCCCC 1          0.333333  \n",
      "chr-XII 457887 457888 -      629906  6                12                 53              GCTGCCGGAAA 1          0.333333  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t629638\t000090\t-\n",
      "chr-XII\t451976\t451977\t629644\t000071\t-\n",
      "chr-XII\t451979\t451980\t629645\t000071\t-\n",
      "chr-XII\t452137\t452138\t629655\t000050\t-\n",
      "chr-XII\t452217\t452218\t629658\t000272\t-\n",
      "chr-XII\t452220\t452221\t629659\t000272\t-\n",
      "chr-XII\t452331\t452332\t629080\t000062\t+\n",
      "chr-XII\t452340\t452341\t629081\t000062\t+\n",
      "chr-XII\t452632\t452633\t629089\t000066\t+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\",\n",
    "    split_group=True)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.log\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Changing filtering threshold (not recommended)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-01T13:04:04.903462Z",
     "start_time": "2019-05-01T13:03:45.921018Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:18, 33172.24 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 340081/340081 [00:00<00:00, 530049.95 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read sites summary:        \n",
      "total: 605,248             \n",
      "unmethylated: 571,328      \n",
      "methylated: 33,920         \n",
      "Genomic positions summary: \n",
      "total: 340,081             \n",
      "low_coverage: 339,355      \n",
      "low_meth_freq: 698         \n",
      "valid: 28                  \n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence       num_motifs meth_freq \n",
      "chr-XII 461541 461542 +      1012246 26               3                  52              AATTCCGAGGG    1          0.896552  \n",
      "chr-XII 455556 455557 +      1012101 16               6                  20              AGATCCGTTGT    1          0.727273  \n",
      "chr-XII 462606 462607 +      1012278 21               9                  28              AATTCCGGGGT    1          0.700000  \n",
      "chr-XII 458208 458209 -      1012544 25               15                 37              TTAAACGCAAA    1          0.625000  \n",
      "chr-XII 463513 463514 -      1012680 13               8                  29              TCTTTCGGGTC    1          0.619048  \n",
      "chr-XII 461032 461033 -      1012599 27               20                 60              CCCCACGCAAA    1          0.574468  \n",
      "chr-XII 456198 456202 -      1012492 17               13                 16              CAGCACGACGGAGT 2          0.566667  \n",
      "chr-XII 461274 461275 +      1012237 21               21                 55              TTATCCGAATG    1          0.500000  \n",
      "chr-XII 465464 465465 -      1012730 10               10                 19              TCCATCGGCTT    1          0.500000  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t455151\t455152\t1012091\t000409\t+\n",
      "chr-XII\t455556\t455557\t1012101\t000727\t+\n",
      "chr-XII\t456198\t456202\t1012492\t000566\t-\n",
      "chr-XII\t456451\t456452\t1012499\t000303\t-\n",
      "chr-XII\t456686\t456687\t1012128\t000321\t+\n",
      "chr-XII\t457045\t457046\t1012140\t000476\t+\n",
      "chr-XII\t457104\t457105\t1012142\t000447\t+\n",
      "chr-XII\t457887\t457888\t1012537\t000361\t-\n",
      "chr-XII\t458208\t458209\t1012544\t000625\t-\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\",\n",
    "    min_llr=1,\n",
    "    min_depth=20,\n",
    "    min_meth_freq=0.3)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.log\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python3.6",
   "language": "python",
   "name": "python3.6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
