{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Freq_meth_calculate usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate methylation frequency at genomic CpG sites from the output of `nanopolish call-methylation`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Options"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **input_fn**\n",
    "\n",
    "Path to a `nanopolish call-methylation` tsv output file (read access required). In command line mode it is also possible to pipe the output of the `nanopolish call-methylation` directly into `Freq_meth_calculate`.\n",
    "\n",
    "* **outdir**\n",
    "\n",
    "Path to an existing directory where to write all the output files generated by `Eventalign_collapse`(write access required). If the directory does not exist an error is raised.\n",
    "\n",
    "* **outprefix**\n",
    "\n",
    "Prefix for all the files generated by the program\n",
    "\n",
    "Filtering options:\n",
    "\n",
    "* **min_llr**\n",
    "\n",
    "Set the minimal required log likelihood ratio between methylated and unmethylated state. Otherwise the value is considered ambiguous. \n",
    "\n",
    "* **min_depth**\n",
    "\n",
    "Set the minimal number of reads with non -ambiguous call per genomic position to be written in the output files.\n",
    "\n",
    "* **min_meth_freq**\n",
    "\n",
    "Set the minimal methylation frequency per genomic position to be written in the output files.\n",
    "\n",
    "* **split_group**\n",
    "\n",
    "In CpG rich regions it is not sometimes hard to distinguish which kmers is methylated in this case `nanopolish` outputs a group containing multiple CpG (or other motif). This option splits each CpG in individual genomic site. This can be useful for visualisation in a genome browser.\n",
    "\n",
    "* **motif**\n",
    "\n",
    "Specify the motif type , akin the one used for `Nanopolish call-methylation`. This is only useful if the split_group option is also selected. \n",
    "\n",
    "!!! note \"Methylation motif to sequence correspondence\"\n",
    "    * cpg → CG\n",
    "    * gpc → GC\n",
    "    * dam → GATC\n",
    "    * dcm → CCAGG\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output files format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Freq_meth_calculate` generates 2 files, a standard BED file and a tabulated file containing extra information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### BED file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Standard genomic BED6 (https://genome.ucsc.edu/FAQ/FAQformat.html#format1). The score correspond to the methylation frequency multiplied by 1000. The file is sorted by coordinates and can be rendered with a genome browser such as [IGV](https://software.broadinstitute.org/software/igv/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Tabulated TSV file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Contrary to the bed file, in the tabulated report, positions are ordered by decreasing methylation frequency.\n",
    "\n",
    "The file contains the following fields:\n",
    "\n",
    "* **chrom / start / end / strand**: Genomic coordinates of the motif or group of motifs in case split_group was not selected.\n",
    "* **site_id**: Unique integer identifier of the genomic position.\n",
    "* **methylated_reads / unmethylated_reads / ambiguous_reads**: Number of reads at a given genomic location with a higher likelyhood of being methylated or unmethylated or with an ambiguous methylation call.\n",
    "* **sequence**: -5 to +5 sequence of the motif or group of motifs in case split_group was not selected.\n",
    "* **num_motifs**: Number of motif in the group.\n",
    "* **meth_freq**: Methylation frequency (out of non anbiguous calls)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Log YAML file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The program also generates a simple log file containing the sites and positions count formated in YAML.\n",
    "\n",
    "Example file content:\n",
    "\n",
    "```\n",
    "Read sites summary:\n",
    "    total: 605,248\n",
    "    unmethylated: 571,328\n",
    "    methylated: 33,920\n",
    "Genomic positions summary:\n",
    "    total: 340,081\n",
    "    low_coverage: 339,355\n",
    "    low_meth_freq: 698\n",
    "    valid: 28\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bash command line usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-08T11:50:06.436703Z",
     "start_time": "2019-05-08T11:50:05.441542Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: NanopolishComp Freq_meth_calculate [-h] [-i INPUT_FN] [-o OUTDIR]\n",
      "                                          [-p OUTPREFIX] [-l MIN_LLR]\n",
      "                                          [-d MIN_DEPTH] [-f MIN_METH_FREQ]\n",
      "                                          [-s] [-m {cpg,gpc,dam,dcm}]\n",
      "                                          [-v | -q]\n",
      "\n",
      "Calculate methylation frequency at genomic CpG sites from the output of\n",
      "nanopolish call-methylation\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -v, --verbose         Increase verbosity (default: False)\n",
      "  -q, --quiet           Reduce verbosity (default: False)\n",
      "\n",
      "Input/Output options:\n",
      "  -i INPUT_FN, --input_fn INPUT_FN\n",
      "                        Path to a nanopolish call_methylation tsv output file.\n",
      "                        If not specified read from std input\n",
      "  -o OUTDIR, --outdir OUTDIR\n",
      "                        Path to the output folder (will be created if it does\n",
      "                        exist yet) (default: ./)\n",
      "  -p OUTPREFIX, --outprefix OUTPREFIX\n",
      "                        text outprefix for all the files generated (default:\n",
      "                        out)\n",
      "\n",
      "Filtering options:\n",
      "  -l MIN_LLR, --min_llr MIN_LLR\n",
      "                        Log likelihood ratio threshold (default: 2.5)\n",
      "  -d MIN_DEPTH, --min_depth MIN_DEPTH\n",
      "                        Minimal number of reads covering a site to be reported\n",
      "                        (default: 10)\n",
      "  -f MIN_METH_FREQ, --min_meth_freq MIN_METH_FREQ\n",
      "                        Minimal methylation frequency of a site to be reported\n",
      "                        (default: 0.05)\n",
      "\n",
      "Motif options:\n",
      "  -s, --split_group     If True, multi motif groups (sequence with close\n",
      "                        motifs) are split in individual site (default: False)\n",
      "  -m {cpg,gpc,dam,dcm}, --motif {cpg,gpc,dam,dcm}\n",
      "                        Methylation motif type (default: cpg)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From an existing nanopolish call_methylation file output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-08T11:52:10.999583Z",
     "start_time": "2019-05-08T11:51:51.857073Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t216740\t000090\t-\n",
      "chr-XII\t451976\t451980\t216746\t000071\t-\n",
      "chr-XII\t452137\t452138\t216753\t000050\t-\n",
      "chr-XII\t452217\t452221\t216756\t000272\t-\n",
      "chr-XII\t452331\t452341\t216384\t000062\t+\n",
      "chr-XII\t452632\t452633\t216391\t000066\t+\n",
      "chr-XII\t452654\t452655\t216769\t000100\t-\n",
      "chr-XII\t453086\t453087\t216781\t000055\t-\n",
      "chr-XII\t453336\t453337\t216790\t000066\t-\n",
      "chrom\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tmeth_freq\n",
      "chr-XII\t465335\t465339\t-\t217103\t6\t5\t29\tCAGCACGACGGAGT\t2\t0.545455\n",
      "chr-XII\t456198\t456202\t-\t216869\t7\t7\t32\tCAGCACGACGGAGT\t2\t0.500000\n",
      "chr-XII\t460773\t460774\t+\t216597\t17\t20\t77\tCTCATCGTATA\t1\t0.459459\n",
      "chr-XII\t457104\t457105\t+\t216519\t6\t8\t43\tAACTACGAGCT\t1\t0.428571\n",
      "chr-XII\t461274\t461275\t+\t216614\t5\t7\t85\tTTATCCGAATG\t1\t0.416667\n",
      "chr-XII\t463406\t463407\t+\t216682\t6\t12\t29\tTCTTTCGCCCC\t1\t0.333333\n",
      "chr-XII\t457887\t457888\t-\t216914\t6\t12\t53\tGCTGCCGGAAA\t1\t0.333333\n",
      "chr-XII\t461077\t461078\t+\t216605\t8\t17\t74\tCTTTCCGCCAA\t1\t0.320000\n",
      "chr-XII\t462283\t462284\t-\t217018\t5\t11\t57\tAGGCACGCAAG\t1\t0.312500\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Options summary\n",
      "\tpackage_name: NanopolishComp\n",
      "\tpackage_version: 0.6.1\n",
      "\ttimestamp: 2019-05-08 12:51:52.765199\n",
      "\tquiet: False\n",
      "\tverbose: True\n",
      "\tmotif: cpg\n",
      "\tsplit_group: False\n",
      "\tmin_meth_freq: 0.05\n",
      "\tmin_depth: 10\n",
      "\tmin_llr: 2.5\n",
      "\toutprefix: out\n",
      "\toutdir: ./output/freq_meth_calculate/\n",
      "\tinput_fn: data/freq_meth_calculate/methylation_calls.tsv\n",
      "\n",
      "## Checking arguments ##\n",
      "\tTesting input file readability\n",
      "\tCreating output folder\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "\r",
      "0 lines [00:00, ? lines/s]\r",
      "3372 lines [00:00, 33717.94 lines/s]\r",
      "7149 lines [00:00, 34838.21 lines/s]\r",
      "10719 lines [00:00, 35087.64 lines/s]\r",
      "14524 lines [00:00, 35925.41 lines/s]\r",
      "18273 lines [00:00, 36378.56 lines/s]\r",
      "21966 lines [00:00, 36541.96 lines/s]\r",
      "25638 lines [00:00, 36594.83 lines/s]\r",
      "29024 lines [00:00, 34807.15 lines/s]\r",
      "32549 lines [00:00, 34936.32 lines/s]\r",
      "36011 lines [00:01, 34840.02 lines/s]\r",
      "39521 lines [00:01, 34915.98 lines/s]\r",
      "43365 lines [00:01, 35902.09 lines/s]\r",
      "47290 lines [00:01, 36843.77 lines/s]\r",
      "51090 lines [00:01, 37180.46 lines/s]\r",
      "54797 lines [00:01, 36649.63 lines/s]\r",
      "58456 lines [00:01, 36208.79 lines/s]\r",
      "62074 lines [00:01, 36003.85 lines/s]\r",
      "65673 lines [00:01, 35564.70 lines/s]\r",
      "69230 lines [00:01, 35433.86 lines/s]\r",
      "72774 lines [00:02, 35125.90 lines/s]\r",
      "76288 lines [00:02, 33329.14 lines/s]\r",
      "79641 lines [00:02, 32917.36 lines/s]\r",
      "82948 lines [00:02, 32123.71 lines/s]\r",
      "86509 lines [00:02, 33095.04 lines/s]\r",
      "89967 lines [00:02, 33526.76 lines/s]\r",
      "93434 lines [00:02, 33861.11 lines/s]\r",
      "97091 lines [00:02, 34629.28 lines/s]\r",
      "100566 lines [00:02, 32572.17 lines/s]\r",
      "104045 lines [00:02, 33205.88 lines/s]\r",
      "107392 lines [00:03, 32238.25 lines/s]\r",
      "110641 lines [00:03, 30385.51 lines/s]\r",
      "113719 lines [00:03, 29811.26 lines/s]\r",
      "116739 lines [00:03, 29924.42 lines/s]\r",
      "119753 lines [00:03, 29431.38 lines/s]\r",
      "122900 lines [00:03, 30013.75 lines/s]\r",
      "126380 lines [00:03, 31303.59 lines/s]\r",
      "129681 lines [00:03, 31796.05 lines/s]\r",
      "133489 lines [00:03, 33450.50 lines/s]\r",
      "137286 lines [00:04, 34687.89 lines/s]\r",
      "140792 lines [00:04, 34581.21 lines/s]\r",
      "144478 lines [00:04, 35228.50 lines/s]\r",
      "148216 lines [00:04, 35846.02 lines/s]\r",
      "151819 lines [00:04, 33562.97 lines/s]\r",
      "155218 lines [00:04, 33559.23 lines/s]\r",
      "158905 lines [00:04, 34488.01 lines/s]\r",
      "162605 lines [00:04, 35204.41 lines/s]\r",
      "166148 lines [00:04, 34490.05 lines/s]\r",
      "169616 lines [00:04, 34116.24 lines/s]\r",
      "173042 lines [00:05, 31750.62 lines/s]\r",
      "176854 lines [00:05, 33425.59 lines/s]\r",
      "180743 lines [00:05, 34895.42 lines/s]\r",
      "184287 lines [00:05, 35041.00 lines/s]\r",
      "187931 lines [00:05, 35448.90 lines/s]\r",
      "191689 lines [00:05, 36061.83 lines/s]\r",
      "195458 lines [00:05, 36533.48 lines/s]\r",
      "199193 lines [00:05, 36774.05 lines/s]\r",
      "202929 lines [00:05, 36946.08 lines/s]\r",
      "206734 lines [00:05, 37269.30 lines/s]\r",
      "210468 lines [00:06, 36914.51 lines/s]\r",
      "214165 lines [00:06, 35872.09 lines/s]\r",
      "217763 lines [00:06, 34527.99 lines/s]\r",
      "221234 lines [00:06, 31747.01 lines/s]\r",
      "224466 lines [00:06, 29450.81 lines/s]\r",
      "227545 lines [00:06, 29838.26 lines/s]\r",
      "230837 lines [00:06, 30699.90 lines/s]\r",
      "234000 lines [00:06, 30972.52 lines/s]\r",
      "237628 lines [00:06, 32393.47 lines/s]\r",
      "241160 lines [00:07, 33217.50 lines/s]\r",
      "244517 lines [00:07, 33321.34 lines/s]\r",
      "247872 lines [00:07, 32573.76 lines/s]\r",
      "251149 lines [00:07, 32209.47 lines/s]\r",
      "254385 lines [00:07, 31406.15 lines/s]\r",
      "257541 lines [00:07, 27011.37 lines/s]\r",
      "260500 lines [00:07, 27736.23 lines/s]\r",
      "263547 lines [00:07, 28503.23 lines/s]\r",
      "266465 lines [00:07, 28663.42 lines/s]\r",
      "269396 lines [00:08, 28853.53 lines/s]\r",
      "272479 lines [00:08, 29414.55 lines/s]\r",
      "275544 lines [00:08, 29773.08 lines/s]\r",
      "278541 lines [00:08, 29553.85 lines/s]\r",
      "281607 lines [00:08, 29876.61 lines/s]\r",
      "284683 lines [00:08, 29085.97 lines/s]\r",
      "288244 lines [00:08, 30776.14 lines/s]\r",
      "291918 lines [00:08, 32350.32 lines/s]\r",
      "295598 lines [00:08, 33566.61 lines/s]\r",
      "299361 lines [00:08, 34688.60 lines/s]\r",
      "303023 lines [00:09, 35245.86 lines/s]\r",
      "306955 lines [00:09, 36374.55 lines/s]\r",
      "310678 lines [00:09, 36625.21 lines/s]\r",
      "314363 lines [00:09, 36261.65 lines/s]\r",
      "318088 lines [00:09, 36551.49 lines/s]\r",
      "321792 lines [00:09, 36695.31 lines/s]\r",
      "325617 lines [00:09, 37147.19 lines/s]\r",
      "329339 lines [00:09, 31990.27 lines/s]\r",
      "332945 lines [00:09, 33110.54 lines/s]\r",
      "336655 lines [00:10, 34211.11 lines/s]\r",
      "340373 lines [00:10, 35046.83 lines/s]\r",
      "344143 lines [00:10, 35802.25 lines/s]\r",
      "347851 lines [00:10, 36173.74 lines/s]\r",
      "351632 lines [00:10, 36648.07 lines/s]\r",
      "355327 lines [00:10, 36736.28 lines/s]\r",
      "359181 lines [00:10, 37259.25 lines/s]\r",
      "363200 lines [00:10, 38091.18 lines/s]\r",
      "367269 lines [00:10, 38834.93 lines/s]\r",
      "371289 lines [00:10, 39234.54 lines/s]\r",
      "375274 lines [00:11, 39414.93 lines/s]\r",
      "379438 lines [00:11, 40055.02 lines/s]\r",
      "383642 lines [00:11, 40629.66 lines/s]\r",
      "387841 lines [00:11, 41025.02 lines/s]\r",
      "391956 lines [00:11, 41061.65 lines/s]\r",
      "396067 lines [00:11, 40880.09 lines/s]\r",
      "400159 lines [00:11, 40765.90 lines/s]\r",
      "404238 lines [00:11, 39454.07 lines/s]\r",
      "408195 lines [00:11, 38780.00 lines/s]\r",
      "412084 lines [00:11, 38326.58 lines/s]\r",
      "415925 lines [00:12, 37540.12 lines/s]\r",
      "419689 lines [00:12, 37406.68 lines/s]\r",
      "423483 lines [00:12, 37564.15 lines/s]\r",
      "427245 lines [00:12, 37316.43 lines/s]\r",
      "430981 lines [00:12, 37310.27 lines/s]\r",
      "434715 lines [00:12, 37284.63 lines/s]\r",
      "438495 lines [00:12, 37435.54 lines/s]\r",
      "442261 lines [00:12, 37500.58 lines/s]\r",
      "446013 lines [00:12, 37433.72 lines/s]\r",
      "449758 lines [00:12, 36125.29 lines/s]\r",
      "453461 lines [00:13, 36389.73 lines/s]\r",
      "457289 lines [00:13, 36936.57 lines/s]\r",
      "461135 lines [00:13, 37379.93 lines/s]\r",
      "464962 lines [00:13, 37642.19 lines/s]\r",
      "468732 lines [00:13, 37379.28 lines/s]\r",
      "472474 lines [00:13, 37341.27 lines/s]\r",
      "476211 lines [00:13, 30920.47 lines/s]\r",
      "479833 lines [00:13, 32338.41 lines/s]\r",
      "483505 lines [00:13, 33537.91 lines/s]\r",
      "487161 lines [00:14, 34385.71 lines/s]\r",
      "490833 lines [00:14, 35053.11 lines/s]\r",
      "494517 lines [00:14, 35568.39 lines/s]\r",
      "498120 lines [00:14, 35386.04 lines/s]\r",
      "501703 lines [00:14, 35516.11 lines/s]\r",
      "505278 lines [00:14, 35341.14 lines/s]\r",
      "508896 lines [00:14, 35588.65 lines/s]\r",
      "512638 lines [00:14, 36116.99 lines/s]\r",
      "516260 lines [00:14, 36015.36 lines/s]\r",
      "519869 lines [00:14, 35910.01 lines/s]\r",
      "523582 lines [00:15, 36266.32 lines/s]\r",
      "527214 lines [00:15, 35415.97 lines/s]\r",
      "531044 lines [00:15, 36232.15 lines/s]\r",
      "534678 lines [00:15, 34902.47 lines/s]\r",
      "538186 lines [00:15, 34520.34 lines/s]\r",
      "541652 lines [00:15, 34054.68 lines/s]\r",
      "545391 lines [00:15, 34990.15 lines/s]\r",
      "549065 lines [00:15, 35495.33 lines/s]\r",
      "552626 lines [00:15, 34344.66 lines/s]\r",
      "556322 lines [00:15, 35088.21 lines/s]\r",
      "559917 lines [00:16, 35339.70 lines/s]\r",
      "563617 lines [00:16, 35821.22 lines/s]\r",
      "567209 lines [00:16, 34874.87 lines/s]\r",
      "570792 lines [00:16, 35155.71 lines/s]\r",
      "574317 lines [00:16, 33441.09 lines/s]\r",
      "577694 lines [00:16, 33537.36 lines/s]\r",
      "581065 lines [00:16, 30567.57 lines/s]\r",
      "584572 lines [00:16, 31790.60 lines/s]\r",
      "588180 lines [00:16, 32965.62 lines/s]\r",
      "591650 lines [00:17, 33463.26 lines/s]\r",
      "595034 lines [00:17, 27365.42 lines/s]\r",
      "598554 lines [00:17, 29322.83 lines/s]\r",
      "601862 lines [00:17, 30355.94 lines/s]\r",
      "605248 lines [00:17, 34539.19 lines/s]\n",
      "Read sites summary\n",
      "\ttotal: 605,248\n",
      "\tunmethylated: 596,963\n",
      "\tmethylated: 8,285\n",
      "\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "\r",
      "  0%|          | 0/340081 [00:00<?, ? positions/s]\r",
      " 19%|█▉        | 64155/340081 [00:00<00:00, 641549.08 positions/s]\r",
      " 37%|███▋      | 124873/340081 [00:00<00:00, 630836.36 positions/s]\r",
      " 54%|█████▍    | 184799/340081 [00:00<00:00, 621017.36 positions/s]\r",
      " 71%|███████   | 241229/340081 [00:00<00:00, 602839.33 positions/s]\r",
      " 88%|████████▊ | 298796/340081 [00:00<00:00, 594420.99 positions/s]\r",
      "100%|██████████| 340081/340081 [00:00<00:00, 591113.30 positions/s]\n",
      "Genomic positions summary\n",
      "\ttotal: 340,081\n",
      "\tlow_coverage: 339,292\n",
      "\tlow_meth_freq: 647\n",
      "\tvalid: 142\n",
      "\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --verbose -i data/freq_meth_calculate/methylation_calls.tsv -o ./output/freq_meth_calculate/ -v\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From standard input to a file (the same can be done with Nanopolish call-methylation directly)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example the split_group option was also selected leading to motif groups to be split in individual positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-08T11:52:32.742961Z",
     "start_time": "2019-05-08T11:52:11.003619Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t289557\t000090\t-\n",
      "chr-XII\t451976\t451977\t289563\t000071\t-\n",
      "chr-XII\t451979\t451980\t289564\t000071\t-\n",
      "chr-XII\t452137\t452138\t289574\t000050\t-\n",
      "chr-XII\t452217\t452218\t289577\t000272\t-\n",
      "chr-XII\t452220\t452221\t289578\t000272\t-\n",
      "chr-XII\t452331\t452332\t288999\t000062\t+\n",
      "chr-XII\t452340\t452341\t289000\t000062\t+\n",
      "chr-XII\t452632\t452633\t289008\t000066\t+\n",
      "chrom\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tmeth_freq\n",
      "chr-XII\t465335\t465336\t-\t290122\t6\t5\t29\tCAGCACGACGG\t2\t0.545455\n",
      "chr-XII\t465338\t465339\t-\t290123\t6\t5\t29\tCACGACGGAGT\t2\t0.545455\n",
      "chr-XII\t456198\t456199\t-\t289760\t7\t7\t32\tCAGCACGACGG\t2\t0.500000\n",
      "chr-XII\t456201\t456202\t-\t289761\t7\t7\t32\tCACGACGGAGT\t2\t0.500000\n",
      "chr-XII\t460773\t460774\t+\t289338\t17\t20\t77\tCTCATCGTATA\t1\t0.459459\n",
      "chr-XII\t457104\t457105\t+\t289216\t6\t8\t43\tAACTACGAGCT\t1\t0.428571\n",
      "chr-XII\t461274\t461275\t+\t289359\t5\t7\t85\tTTATCCGAATG\t1\t0.416667\n",
      "chr-XII\t463406\t463407\t+\t289460\t6\t12\t29\tTCTTTCGCCCC\t1\t0.333333\n",
      "chr-XII\t457887\t457888\t-\t289825\t6\t12\t53\tGCTGCCGGAAA\t1\t0.333333\n",
      "General options:\n",
      "\tpackage_name: NanopolishComp\n",
      "\tpackage_version: 0.6.1\n",
      "\ttimestamp: 2019-05-08 12:52:11.951019\n",
      "\tquiet: False\n",
      "\tverbose: False\n",
      "\tmotif: cpg\n",
      "\tsplit_group: True\n",
      "\tmin_meth_freq: 0.05\n",
      "\tmin_depth: 10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "\r",
      "0 lines [00:00, ? lines/s]\r",
      "3375 lines [00:00, 33745.29 lines/s]\r",
      "6650 lines [00:00, 33439.63 lines/s]\r",
      "9827 lines [00:00, 32919.78 lines/s]\r",
      "13123 lines [00:00, 32930.08 lines/s]\r",
      "16686 lines [00:00, 33695.15 lines/s]\r",
      "20025 lines [00:00, 33599.18 lines/s]\r",
      "23133 lines [00:00, 32800.81 lines/s]\r",
      "26245 lines [00:00, 32277.46 lines/s]\r",
      "29541 lines [00:00, 32477.20 lines/s]\r",
      "32875 lines [00:01, 32729.33 lines/s]\r",
      "36409 lines [00:01, 33470.30 lines/s]\r",
      "39803 lines [00:01, 33609.09 lines/s]\r",
      "43305 lines [00:01, 34018.21 lines/s]\r",
      "46679 lines [00:01, 32977.04 lines/s]\r",
      "49965 lines [00:01, 31220.97 lines/s]\r",
      "53098 lines [00:01, 30159.17 lines/s]\r",
      "56579 lines [00:01, 31417.21 lines/s]\r",
      "59976 lines [00:01, 32141.65 lines/s]\r",
      "63212 lines [00:01, 32097.75 lines/s]\r",
      "66437 lines [00:02, 31629.41 lines/s]\r",
      "69940 lines [00:02, 32577.80 lines/s]\r",
      "73214 lines [00:02, 31136.91 lines/s]\r",
      "76539 lines [00:02, 31741.38 lines/s]\r",
      "80004 lines [00:02, 32559.04 lines/s]\r",
      "83363 lines [00:02, 32856.64 lines/s]\r",
      "86663 lines [00:02, 32667.40 lines/s]\r",
      "89940 lines [00:02, 32380.18 lines/s]\r",
      "93276 lines [00:02, 32667.72 lines/s]\r",
      "96832 lines [00:02, 33483.85 lines/s]\r",
      "100190 lines [00:03, 31660.04 lines/s]\r",
      "103770 lines [00:03, 32796.49 lines/s]\r",
      "107079 lines [00:03, 31844.38 lines/s]\r",
      "110290 lines [00:03, 31269.87 lines/s]\r",
      "113655 lines [00:03, 31946.91 lines/s]\r",
      "116868 lines [00:03, 31358.99 lines/s]\r",
      "120405 lines [00:03, 32462.63 lines/s]\r",
      "123862 lines [00:03, 33065.20 lines/s]\r",
      "127414 lines [00:03, 33763.22 lines/s]\r",
      "130806 lines [00:04, 29028.69 lines/s]\r",
      "133837 lines [00:04, 24465.49 lines/s]\r",
      "136496 lines [00:04, 22992.53 lines/s]\r",
      "138962 lines [00:04, 23020.95 lines/s]\r",
      "141381 lines [00:04, 22376.08 lines/s]\r",
      "143705 lines [00:04, 21885.91 lines/s]\r",
      "146155 lines [00:04, 22607.83 lines/s]\r",
      "148700 lines [00:04, 23389.98 lines/s]\r",
      "151081 lines [00:05, 23020.02 lines/s]\r",
      "153761 lines [00:05, 24037.10 lines/s]\r",
      "156280 lines [00:05, 24369.36 lines/s]\r",
      "158959 lines [00:05, 25047.44 lines/s]\r",
      "161595 lines [00:05, 25425.43 lines/s]\r",
      "164216 lines [00:05, 25654.68 lines/s]\r",
      "167275 lines [00:05, 26955.88 lines/s]\r",
      "170647 lines [00:05, 28679.93 lines/s]\r",
      "173715 lines [00:05, 29249.64 lines/s]\r",
      "177041 lines [00:05, 30343.57 lines/s]\r",
      "180411 lines [00:06, 31276.76 lines/s]\r",
      "183572 lines [00:06, 28400.49 lines/s]\r",
      "186801 lines [00:06, 29463.75 lines/s]\r",
      "189897 lines [00:06, 29896.69 lines/s]\r",
      "192932 lines [00:06, 29725.75 lines/s]\r",
      "196365 lines [00:06, 30971.86 lines/s]\r",
      "199581 lines [00:06, 31317.24 lines/s]\r",
      "203082 lines [00:06, 32340.20 lines/s]\r",
      "206625 lines [00:06, 33207.02 lines/s]\r",
      "209971 lines [00:06, 30717.62 lines/s]\r",
      "213312 lines [00:07, 31478.35 lines/s]\r",
      "216949 lines [00:07, 32800.98 lines/s]\r",
      "220549 lines [00:07, 33695.32 lines/s]\r",
      "223956 lines [00:07, 32713.00 lines/s]\r",
      "227260 lines [00:07, 31834.62 lines/s]\r",
      "230773 lines [00:07, 32755.57 lines/s]\r",
      "234163 lines [00:07, 33085.33 lines/s]\r",
      "237730 lines [00:07, 33819.12 lines/s]\r",
      "241177 lines [00:07, 34010.31 lines/s]\r",
      "244613 lines [00:07, 34114.45 lines/s]\r",
      "248145 lines [00:08, 34467.11 lines/s]\r",
      "251599 lines [00:08, 28800.91 lines/s]\r",
      "254993 lines [00:08, 30171.09 lines/s]\r",
      "258303 lines [00:08, 30992.49 lines/s]\r",
      "261500 lines [00:08, 31088.53 lines/s]\r",
      "264679 lines [00:08, 31295.18 lines/s]\r",
      "267857 lines [00:08, 29137.13 lines/s]\r",
      "270835 lines [00:08, 26249.09 lines/s]\r",
      "273565 lines [00:09, 26554.81 lines/s]\r",
      "276809 lines [00:09, 28082.41 lines/s]\r",
      "280373 lines [00:09, 29989.83 lines/s]\r",
      "283569 lines [00:09, 30553.27 lines/s]\r",
      "286944 lines [00:09, 31444.21 lines/s]\r",
      "290385 lines [00:09, 32278.66 lines/s]\r",
      "293786 lines [00:09, 32777.53 lines/s]\r",
      "297160 lines [00:09, 33059.63 lines/s]\r",
      "300488 lines [00:09, 29374.35 lines/s]\r",
      "303517 lines [00:10, 25870.95 lines/s]\r",
      "306245 lines [00:10, 24461.90 lines/s]\r",
      "308806 lines [00:10, 23815.66 lines/s]\r",
      "311272 lines [00:10, 22651.67 lines/s]\r",
      "313608 lines [00:10, 21735.97 lines/s]\r",
      "316164 lines [00:10, 22756.92 lines/s]\r",
      "318493 lines [00:10, 20918.75 lines/s]\r",
      "321779 lines [00:10, 23477.89 lines/s]\r",
      "324894 lines [00:10, 25350.14 lines/s]\r",
      "328240 lines [00:11, 27337.11 lines/s]\r",
      "331525 lines [00:11, 28784.79 lines/s]\r",
      "334876 lines [00:11, 30056.03 lines/s]\r",
      "338177 lines [00:11, 30883.43 lines/s]\r",
      "341342 lines [00:11, 30052.83 lines/s]\r",
      "344743 lines [00:11, 31138.46 lines/s]\r",
      "347915 lines [00:11, 31310.28 lines/s]\r",
      "351082 lines [00:11, 31386.50 lines/s]\r",
      "354341 lines [00:11, 31735.58 lines/s]\r",
      "357603 lines [00:11, 31994.09 lines/s]\r",
      "360857 lines [00:12, 32155.64 lines/s]\r",
      "364083 lines [00:12, 31726.23 lines/s]\r",
      "367430 lines [00:12, 32227.99 lines/s]\r",
      "370862 lines [00:12, 32825.83 lines/s]\r",
      "374383 lines [00:12, 33505.67 lines/s]\r",
      "378053 lines [00:12, 34402.57 lines/s]\r",
      "381824 lines [00:12, 35331.37 lines/s]\r",
      "385656 lines [00:12, 36176.41 lines/s]\r",
      "389313 lines [00:12, 36291.85 lines/s]\r",
      "392954 lines [00:12, 36190.76 lines/s]\r",
      "396844 lines [00:13, 36962.24 lines/s]\r",
      "400821 lines [00:13, 37761.28 lines/s]\r",
      "404608 lines [00:13, 37108.79 lines/s]\r",
      "408329 lines [00:13, 36450.16 lines/s]\r",
      "411984 lines [00:13, 35751.32 lines/s]\r",
      "415569 lines [00:13, 35517.77 lines/s]\r",
      "419128 lines [00:13, 34858.69 lines/s]\r",
      "422622 lines [00:13, 33738.11 lines/s]\r",
      "426010 lines [00:13, 32586.52 lines/s]\r",
      "429287 lines [00:14, 31368.48 lines/s]\r",
      "432447 lines [00:14, 30608.71 lines/s]\r",
      "435528 lines [00:14, 30352.54 lines/s]\r",
      "438578 lines [00:14, 30210.60 lines/s]\r",
      "441817 lines [00:14, 30829.43 lines/s]\r",
      "444911 lines [00:14, 30474.94 lines/s]\r",
      "448083 lines [00:14, 30837.33 lines/s]\r",
      "451174 lines [00:14, 30616.02 lines/s]\r",
      "454370 lines [00:14, 31005.41 lines/s]\r",
      "457956 lines [00:14, 32315.67 lines/s]\r",
      "461206 lines [00:15, 26416.23 lines/s]\r",
      "464711 lines [00:15, 28522.84 lines/s]\r",
      "467798 lines [00:15, 29186.02 lines/s]\r",
      "471148 lines [00:15, 30357.96 lines/s]\r",
      "474287 lines [00:15, 30640.05 lines/s]\r",
      "477424 lines [00:15, 26224.47 lines/s]\r",
      "480810 lines [00:15, 28126.91 lines/s]\r",
      "484228 lines [00:15, 29704.05 lines/s]\r",
      "487686 lines [00:15, 31016.02 lines/s]\r",
      "491068 lines [00:16, 31805.58 lines/s]\r",
      "494496 lines [00:16, 32508.32 lines/s]\r",
      "497896 lines [00:16, 32940.08 lines/s]\r",
      "501359 lines [00:16, 33428.93 lines/s]\r",
      "504834 lines [00:16, 33812.74 lines/s]\r",
      "508327 lines [00:16, 34139.97 lines/s]\r",
      "511834 lines [00:16, 34413.40 lines/s]\r",
      "515288 lines [00:16, 34189.33 lines/s]\r",
      "518717 lines [00:16, 33656.70 lines/s]\r",
      "522091 lines [00:16, 33321.33 lines/s]\r",
      "525540 lines [00:17, 33662.17 lines/s]\r",
      "529000 lines [00:17, 33936.44 lines/s]\r",
      "532413 lines [00:17, 33993.27 lines/s]\r",
      "535816 lines [00:17, 33778.27 lines/s]\r",
      "539197 lines [00:17, 30021.70 lines/s]\r",
      "542279 lines [00:17, 25260.34 lines/s]\r",
      "544990 lines [00:17, 24905.87 lines/s]\r",
      "547933 lines [00:17, 26108.22 lines/s]\r",
      "551386 lines [00:18, 28169.03 lines/s]\r",
      "554324 lines [00:18, 27641.44 lines/s]\r",
      "557175 lines [00:18, 26848.91 lines/s]\r",
      "559926 lines [00:18, 26868.21 lines/s]\r",
      "562659 lines [00:18, 26396.22 lines/s]\r",
      "565577 lines [00:18, 27171.31 lines/s]\r",
      "568324 lines [00:18, 18668.81 lines/s]\r",
      "570759 lines [00:18, 20073.09 lines/s]\r",
      "573460 lines [00:19, 21748.30 lines/s]\r",
      "576035 lines [00:19, 22811.60 lines/s]\r",
      "578713 lines [00:19, 23872.84 lines/s]\r",
      "581750 lines [00:19, 25508.89 lines/s]\r",
      "585067 lines [00:19, 27407.97 lines/s]\r",
      "588605 lines [00:19, 29393.26 lines/s]\r",
      "592150 lines [00:19, 30980.35 lines/s]\r",
      "595671 lines [00:19, 32138.00 lines/s]\r",
      "599239 lines [00:19, 33123.36 lines/s]\r",
      "602872 lines [00:19, 34022.32 lines/s]\r",
      "605248 lines [00:19, 30307.05 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "\r",
      "  0%|          | 0/455542 [00:00<?, ? positions/s]\r",
      " 15%|█▍        | 66940/455542 [00:00<00:00, 669392.66 positions/s]\r",
      " 30%|██▉       | 134827/455542 [00:00<00:00, 672206.72 positions/s]\r",
      " 44%|████▍     | 202638/455542 [00:00<00:00, 673965.16 positions/s]\r",
      " 59%|█████▉    | 269205/455542 [00:00<00:00, 671454.69 positions/s]\r",
      " 74%|███████▎  | 335611/455542 [00:00<00:00, 669218.28 positions/s]\r",
      " 88%|████████▊ | 398904/455542 [00:00<00:00, 657900.51 positions/s]\r",
      "100%|██████████| 455542/455542 [00:00<00:00, 660202.32 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "cat data/freq_meth_calculate/methylation_calls.tsv | NanopolishComp Freq_meth_calculate -o ./output/freq_meth_calculate/ --split_group\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import the package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-08T11:00:30.216219Z",
     "start_time": "2019-05-08T11:00:30.206007Z"
    }
   },
   "outputs": [],
   "source": [
    "# Import main program\n",
    "from NanopolishComp.Freq_meth_calculate import Freq_meth_calculate\n",
    "\n",
    "# Import helper functions\n",
    "from NanopolishComp.common import jhelp, head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### python API help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-08T11:00:30.233919Z",
     "start_time": "2019-05-08T11:00:30.219992Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "**NanopolishComp.Freq_meth_calculate.__init__**\n",
       "\n",
       "Calculate methylation frequency at genomic CpG sites from the output of nanopolish call-methylation\n",
       "\n",
       "---\n",
       "\n",
       "* **input_fn** *: str (required)*\n",
       "\n",
       "Path to a nanopolish call_methylation tsv output file\n",
       "\n",
       "* **outdir** *: str (default = ./)*\n",
       "\n",
       "Path to the output folder (will be created if it does exist yet)\n",
       "\n",
       "* **outprefix** *: str (default = out)*\n",
       "\n",
       "text outprefix for all the files generated\n",
       "\n",
       "* **min_llr** *: float (default = 2.5)*\n",
       "\n",
       "Log likelihood ratio threshold\n",
       "\n",
       "* **min_depth** *: int (default = 10)*\n",
       "\n",
       "Minimal number of reads covering a site to be reported\n",
       "\n",
       "* **min_meth_freq** *: float (default = 0.05)*\n",
       "\n",
       "Minimal methylation frequency of a site to be reported\n",
       "\n",
       "* **split_group** *: bool (default = False)*\n",
       "\n",
       "If True, multi motif groups (sequence with close motifs) are split in individual site\n",
       "\n",
       "* **motif** *: {cpg,gpc,dam,dcm} (default = cpg)*\n",
       "\n",
       "Methylation motif type\n",
       "\n",
       "* **verbose** *: bool (default = False)*\n",
       "\n",
       "Increase verbosity\n",
       "\n",
       "* **quiet** *: bool (default = False)*\n",
       "\n",
       "Reduce verbosity\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(Freq_meth_calculate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-23T09:19:41.284517Z",
     "start_time": "2019-04-23T09:19:41.165804Z"
    },
    "scrolled": false
   },
   "source": [
    "#### basic setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-08T11:00:48.887439Z",
     "start_time": "2019-05-08T11:00:30.235898Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:17, 33626.64 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 340081/340081 [00:00<00:00, 613704.15 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "General options:                      \n",
      "package_name: NanopolishComp          \n",
      "package_version: 0.6.1                \n",
      "timestamp: 2019-05-08 12:00:30.240793 \n",
      "quiet: False                          \n",
      "verbose: False                        \n",
      "motif: cpg                            \n",
      "split_group: False                    \n",
      "min_meth_freq: 0.05                   \n",
      "min_depth: 10                         \n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence       num_motifs meth_freq \n",
      "chr-XII 465335 465339 -      217103  6                5                  29              CAGCACGACGGAGT 2          0.545455  \n",
      "chr-XII 456198 456202 -      216869  7                7                  32              CAGCACGACGGAGT 2          0.500000  \n",
      "chr-XII 460773 460774 +      216597  17               20                 77              CTCATCGTATA    1          0.459459  \n",
      "chr-XII 457104 457105 +      216519  6                8                  43              AACTACGAGCT    1          0.428571  \n",
      "chr-XII 461274 461275 +      216614  5                7                  85              TTATCCGAATG    1          0.416667  \n",
      "chr-XII 463406 463407 +      216682  6                12                 29              TCTTTCGCCCC    1          0.333333  \n",
      "chr-XII 457887 457888 -      216914  6                12                 53              GCTGCCGGAAA    1          0.333333  \n",
      "chr-XII 461077 461078 +      216605  8                17                 74              CTTTCCGCCAA    1          0.320000  \n",
      "chr-XII 462283 462284 -      217018  5                11                 57              AGGCACGCAAG    1          0.312500  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t216740\t000090\t-\n",
      "chr-XII\t451976\t451980\t216746\t000071\t-\n",
      "chr-XII\t452137\t452138\t216753\t000050\t-\n",
      "chr-XII\t452217\t452221\t216756\t000272\t-\n",
      "chr-XII\t452331\t452341\t216384\t000062\t+\n",
      "chr-XII\t452632\t452633\t216391\t000066\t+\n",
      "chr-XII\t452654\t452655\t216769\t000100\t-\n",
      "chr-XII\t453086\t453087\t216781\t000055\t-\n",
      "chr-XII\t453336\t453337\t216790\t000066\t-\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\")\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.log\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### split motif groups in individual genomic position "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-08T11:01:07.753387Z",
     "start_time": "2019-05-08T11:00:48.889538Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:18, 33565.56 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 455542/455542 [00:00<00:00, 627132.15 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "General options:                      \n",
      "package_name: NanopolishComp          \n",
      "package_version: 0.6.1                \n",
      "timestamp: 2019-05-08 12:00:48.890953 \n",
      "quiet: False                          \n",
      "verbose: False                        \n",
      "motif: cpg                            \n",
      "split_group: True                     \n",
      "min_meth_freq: 0.05                   \n",
      "min_depth: 10                         \n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence    num_motifs meth_freq \n",
      "chr-XII 465335 465336 -      630203  6                5                  29              CAGCACGACGG 2          0.545455  \n",
      "chr-XII 465338 465339 -      630204  6                5                  29              CACGACGGAGT 2          0.545455  \n",
      "chr-XII 456198 456199 -      629841  7                7                  32              CAGCACGACGG 2          0.500000  \n",
      "chr-XII 456201 456202 -      629842  7                7                  32              CACGACGGAGT 2          0.500000  \n",
      "chr-XII 460773 460774 +      629419  17               20                 77              CTCATCGTATA 1          0.459459  \n",
      "chr-XII 457104 457105 +      629297  6                8                  43              AACTACGAGCT 1          0.428571  \n",
      "chr-XII 461274 461275 +      629440  5                7                  85              TTATCCGAATG 1          0.416667  \n",
      "chr-XII 463406 463407 +      629541  6                12                 29              TCTTTCGCCCC 1          0.333333  \n",
      "chr-XII 457887 457888 -      629906  6                12                 53              GCTGCCGGAAA 1          0.333333  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t629638\t000090\t-\n",
      "chr-XII\t451976\t451977\t629644\t000071\t-\n",
      "chr-XII\t451979\t451980\t629645\t000071\t-\n",
      "chr-XII\t452137\t452138\t629655\t000050\t-\n",
      "chr-XII\t452217\t452218\t629658\t000272\t-\n",
      "chr-XII\t452220\t452221\t629659\t000272\t-\n",
      "chr-XII\t452331\t452332\t629080\t000062\t+\n",
      "chr-XII\t452340\t452341\t629081\t000062\t+\n",
      "chr-XII\t452632\t452633\t629089\t000066\t+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\",\n",
    "    split_group=True)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.log\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Changing filtering threshold (not recommended)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-08T11:01:25.326910Z",
     "start_time": "2019-05-08T11:01:07.755366Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:16, 35714.56 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 340081/340081 [00:00<00:00, 622699.90 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "General options:                      \n",
      "package_name: NanopolishComp          \n",
      "package_version: 0.6.1                \n",
      "timestamp: 2019-05-08 12:01:07.757086 \n",
      "quiet: False                          \n",
      "verbose: False                        \n",
      "motif: cpg                            \n",
      "split_group: False                    \n",
      "min_meth_freq: 0.3                    \n",
      "min_depth: 20                         \n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence       num_motifs meth_freq \n",
      "chr-XII 461541 461542 +      1012246 26               3                  52              AATTCCGAGGG    1          0.896552  \n",
      "chr-XII 455556 455557 +      1012101 16               6                  20              AGATCCGTTGT    1          0.727273  \n",
      "chr-XII 462606 462607 +      1012278 21               9                  28              AATTCCGGGGT    1          0.700000  \n",
      "chr-XII 458208 458209 -      1012544 25               15                 37              TTAAACGCAAA    1          0.625000  \n",
      "chr-XII 463513 463514 -      1012680 13               8                  29              TCTTTCGGGTC    1          0.619048  \n",
      "chr-XII 461032 461033 -      1012599 27               20                 60              CCCCACGCAAA    1          0.574468  \n",
      "chr-XII 456198 456202 -      1012492 17               13                 16              CAGCACGACGGAGT 2          0.566667  \n",
      "chr-XII 461274 461275 +      1012237 21               21                 55              TTATCCGAATG    1          0.500000  \n",
      "chr-XII 465464 465465 -      1012730 10               10                 19              TCCATCGGCTT    1          0.500000  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t455151\t455152\t1012091\t000409\t+\n",
      "chr-XII\t455556\t455557\t1012101\t000727\t+\n",
      "chr-XII\t456198\t456202\t1012492\t000566\t-\n",
      "chr-XII\t456451\t456452\t1012499\t000303\t-\n",
      "chr-XII\t456686\t456687\t1012128\t000321\t+\n",
      "chr-XII\t457045\t457046\t1012140\t000476\t+\n",
      "chr-XII\t457104\t457105\t1012142\t000447\t+\n",
      "chr-XII\t457887\t457888\t1012537\t000361\t-\n",
      "chr-XII\t458208\t458209\t1012544\t000625\t-\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\",\n",
    "    min_llr=1,\n",
    "    min_depth=20,\n",
    "    min_meth_freq=0.3)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.log\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python3.6",
   "language": "python",
   "name": "python3.6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
