{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-10T10:03:00.595966Z",
     "start_time": "2018-08-10T10:03:00.587806Z"
    }
   },
   "source": [
    "# Freq_meth_calculate usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate methylation frequency at genomic CpG sites from the output of `nanopolish call-methylation`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Options"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **input_fn**\n",
    "\n",
    "Path to a `nanopolish call-methylation` tsv output file (read access required). In command line mode it is also possible to pipe the output of the `nanopolish call-methylation` directly into `Freq_meth_calculate`.\n",
    "\n",
    "* **outdir**\n",
    "\n",
    "Path to an existing directory where to write all the output files generated by `Eventalign_collapse`(write access required). If the directory does not exist an error is raised.\n",
    "\n",
    "* **outprefix**\n",
    "\n",
    "Prefix for all the files generated by the program\n",
    "\n",
    "Filtering options:\n",
    "\n",
    "* **min_llr**\n",
    "\n",
    "Set the minimal required log likelihood ratio between methylated and unmethylated state. Otherwise the value is considered ambiguous. \n",
    "\n",
    "* **min_depth**\n",
    "\n",
    "Set the minimal number of reads with non -ambiguous call per genomic position to be written in the output files.\n",
    "\n",
    "* **min_meth_freq**\n",
    "\n",
    "Set the minimal methylation frequency per genomic position to be written in the output files.\n",
    "\n",
    "* **split_group**\n",
    "\n",
    "In CpG rich regions it is not sometimes hard to distinguish which kmers is methylated in this case `nanopolish` outputs a group containing multiple CpG (or other motif). This option splits each CpG in individual genomic site. This can be useful for visualisation in a genome browser.\n",
    "\n",
    "* **motif**\n",
    "\n",
    "Specify the motif type , akin the one used for `Nanopolish call-methylation`. This is only useful if the split_group option is also selected. \n",
    "\n",
    "!!! note \"Methylation motif to sequence correspondence\"\n",
    "    * cpg → CG\n",
    "    * gpc → GC\n",
    "    * dam → GATC\n",
    "    * dcm → CCAGG\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output files format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`Freq_meth_calculate` generates 2 files, a standard BED file and a tabulated file containing extra information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### BED file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Standard genomic BED6 (https://genome.ucsc.edu/FAQ/FAQformat.html#format1). The score correspond to the methylation frequency multiplied by 1000. The file is sorted by coordinates and can be rendered with a genome browser such as [IGV](https://software.broadinstitute.org/software/igv/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Tabulated TSV file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Contrary to the bed file, in the tabulated report, positions are ordered by decreasing methylation frequency.\n",
    "\n",
    "The file contains the following fields:\n",
    "\n",
    "* **chrom / start / end / strand**: Genomic coordinates of the motif or group of motifs in case split_group was not selected.\n",
    "* **site_id**: Unique integer identifier of the genomic position.\n",
    "* **methylated_reads / unmethylated_reads / ambiguous_reads**: Number of reads at a given genomic location with a higher likelyhood of being methylated or unmethylated or with an ambiguous methylation call.\n",
    "* **sequence**: -5 to +5 sequence of the motif or group of motifs in case split_group was not selected.\n",
    "* **num_motifs**: Number of motif in the group.\n",
    "* **meth_freq**: Methylation frequency (out of non anbiguous calls)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Log YAML file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The program also generates a simple log file containing the sites and positions count formated in YAML.\n",
    "\n",
    "Example file content:\n",
    "\n",
    "```\n",
    "Read sites summary:\n",
    "    total: 605,248\n",
    "    unmethylated: 571,328\n",
    "    methylated: 33,920\n",
    "Genomic positions summary:\n",
    "    total: 340,081\n",
    "    low_coverage: 339,355\n",
    "    low_meth_freq: 698\n",
    "    valid: 28\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bash command line usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:36:28.617634Z",
     "start_time": "2019-05-02T09:36:27.566472Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: NanopolishComp Freq_meth_calculate [-h] [-i INPUT_FN] [-o OUTDIR]\n",
      "                                          [-p OUTPREFIX] [-l MIN_LLR]\n",
      "                                          [-d MIN_DEPTH] [-f MIN_METH_FREQ]\n",
      "                                          [-s] [-m {cpg,gpc,dam,dcm}]\n",
      "                                          [-v | -q]\n",
      "\n",
      "Calculate methylation frequency at genomic CpG sites from the output of\n",
      "nanopolish call-methylation\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -v, --verbose         Increase verbosity (default: False)\n",
      "  -q, --quiet           Reduce verbosity (default: False)\n",
      "\n",
      "Input/Output options:\n",
      "  -i INPUT_FN, --input_fn INPUT_FN\n",
      "                        Path to a nanopolish call_methylation tsv output file.\n",
      "                        If not specified read from std input\n",
      "  -o OUTDIR, --outdir OUTDIR\n",
      "                        Path to the output folder (default: ./)\n",
      "  -p OUTPREFIX, --outprefix OUTPREFIX\n",
      "                        text outprefix for all the files generated (default:\n",
      "                        out)\n",
      "\n",
      "Filtering options:\n",
      "  -l MIN_LLR, --min_llr MIN_LLR\n",
      "                        Log likelihood ratio threshold (default: 2.5)\n",
      "  -d MIN_DEPTH, --min_depth MIN_DEPTH\n",
      "                        Minimal number of reads covering a site to be reported\n",
      "                        (default: 10)\n",
      "  -f MIN_METH_FREQ, --min_meth_freq MIN_METH_FREQ\n",
      "                        Minimal methylation frequency of a site to be reported\n",
      "                        (default: 0.05)\n",
      "\n",
      "Motif options:\n",
      "  -s, --split_group     If True, multi motif groups (sequence with close\n",
      "                        motifs) are split in individual site (default: False)\n",
      "  -m {cpg,gpc,dam,dcm}, --motif {cpg,gpc,dam,dcm}\n",
      "                        Methylation motif type (default: cpg)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From an existing nanopolish call_methylation file output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:36:48.448839Z",
     "start_time": "2019-05-02T09:36:28.620951Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t216740\t000090\t-\n",
      "chr-XII\t451976\t451980\t216746\t000071\t-\n",
      "chr-XII\t452137\t452138\t216753\t000050\t-\n",
      "chr-XII\t452217\t452221\t216756\t000272\t-\n",
      "chr-XII\t452331\t452341\t216384\t000062\t+\n",
      "chr-XII\t452632\t452633\t216391\t000066\t+\n",
      "chr-XII\t452654\t452655\t216769\t000100\t-\n",
      "chr-XII\t453086\t453087\t216781\t000055\t-\n",
      "chr-XII\t453336\t453337\t216790\t000066\t-\n",
      "chrom\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tmeth_freq\n",
      "chr-XII\t465335\t465339\t-\t217103\t6\t5\t29\tCAGCACGACGGAGT\t2\t0.545455\n",
      "chr-XII\t456198\t456202\t-\t216869\t7\t7\t32\tCAGCACGACGGAGT\t2\t0.500000\n",
      "chr-XII\t460773\t460774\t+\t216597\t17\t20\t77\tCTCATCGTATA\t1\t0.459459\n",
      "chr-XII\t457104\t457105\t+\t216519\t6\t8\t43\tAACTACGAGCT\t1\t0.428571\n",
      "chr-XII\t461274\t461275\t+\t216614\t5\t7\t85\tTTATCCGAATG\t1\t0.416667\n",
      "chr-XII\t463406\t463407\t+\t216682\t6\t12\t29\tTCTTTCGCCCC\t1\t0.333333\n",
      "chr-XII\t457887\t457888\t-\t216914\t6\t12\t53\tGCTGCCGGAAA\t1\t0.333333\n",
      "chr-XII\t461077\t461078\t+\t216605\t8\t17\t74\tCTTTCCGCCAA\t1\t0.320000\n",
      "chr-XII\t462283\t462284\t-\t217018\t5\t11\t57\tAGGCACGCAAG\t1\t0.312500\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Options summary\n",
      "\tpackage_name: NanopolishComp\n",
      "\tpackage_version: 0.6.1\n",
      "\ttimestamp: 2019-05-02 10:36:29.585341\n",
      "\tquiet: False\n",
      "\tverbose: True\n",
      "\tmotif: cpg\n",
      "\tsplit_group: False\n",
      "\tmin_meth_freq: 0.05\n",
      "\tmin_depth: 10\n",
      "\tmin_llr: 2.5\n",
      "\toutprefix: out\n",
      "\toutdir: ./output/freq_meth_calculate/\n",
      "\tinput_fn: data/freq_meth_calculate/methylation_calls.tsv\n",
      "\n",
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "\r",
      "0 lines [00:00, ? lines/s]\r",
      "3972 lines [00:00, 39719.66 lines/s]\r",
      "7974 lines [00:00, 39807.26 lines/s]\r",
      "11925 lines [00:00, 39714.71 lines/s]\r",
      "15922 lines [00:00, 39789.10 lines/s]\r",
      "19677 lines [00:00, 39088.44 lines/s]\r",
      "22922 lines [00:00, 36697.98 lines/s]\r",
      "26157 lines [00:00, 29681.15 lines/s]\r",
      "29386 lines [00:00, 30416.83 lines/s]\r",
      "33098 lines [00:00, 32157.87 lines/s]\r",
      "36729 lines [00:01, 33298.15 lines/s]\r",
      "40558 lines [00:01, 34651.22 lines/s]\r",
      "44481 lines [00:01, 35905.91 lines/s]\r",
      "48332 lines [00:01, 36649.03 lines/s]\r",
      "52240 lines [00:01, 37344.28 lines/s]\r",
      "56156 lines [00:01, 37870.58 lines/s]\r",
      "60072 lines [00:01, 38247.05 lines/s]\r",
      "64005 lines [00:01, 38563.25 lines/s]\r",
      "67892 lines [00:01, 38654.04 lines/s]\r",
      "71763 lines [00:01, 38124.82 lines/s]\r",
      "75581 lines [00:02, 32395.50 lines/s]\r",
      "78974 lines [00:02, 27110.51 lines/s]\r",
      "82500 lines [00:02, 29130.13 lines/s]\r",
      "85745 lines [00:02, 30047.66 lines/s]\r",
      "88911 lines [00:02, 26951.26 lines/s]\r",
      "91779 lines [00:02, 26365.74 lines/s]\r",
      "95469 lines [00:02, 28833.09 lines/s]\r",
      "98741 lines [00:02, 29897.35 lines/s]\r",
      "101850 lines [00:03, 25426.34 lines/s]\r",
      "105371 lines [00:03, 27733.79 lines/s]\r",
      "108346 lines [00:03, 26868.81 lines/s]\r",
      "111180 lines [00:03, 21167.71 lines/s]\r",
      "114625 lines [00:03, 23935.91 lines/s]\r",
      "118268 lines [00:03, 26680.03 lines/s]\r",
      "121922 lines [00:03, 29028.45 lines/s]\r",
      "125482 lines [00:03, 30730.17 lines/s]\r",
      "128940 lines [00:04, 31790.40 lines/s]\r",
      "132407 lines [00:04, 32600.32 lines/s]\r",
      "135792 lines [00:04, 25016.46 lines/s]\r",
      "138648 lines [00:04, 21498.39 lines/s]\r",
      "142083 lines [00:04, 24215.84 lines/s]\r",
      "144853 lines [00:04, 21830.15 lines/s]\r",
      "147989 lines [00:04, 24017.97 lines/s]\r",
      "151092 lines [00:04, 25764.01 lines/s]\r",
      "154225 lines [00:05, 27214.11 lines/s]\r",
      "157125 lines [00:05, 27158.76 lines/s]\r",
      "160148 lines [00:05, 28011.78 lines/s]\r",
      "163422 lines [00:05, 29279.59 lines/s]\r",
      "166932 lines [00:05, 30811.06 lines/s]\r",
      "170383 lines [00:05, 31834.16 lines/s]\r",
      "173630 lines [00:05, 30157.04 lines/s]\r",
      "177112 lines [00:05, 31419.03 lines/s]\r",
      "180564 lines [00:05, 32284.12 lines/s]\r",
      "183839 lines [00:06, 29498.66 lines/s]\r",
      "186868 lines [00:06, 25090.81 lines/s]\r",
      "189547 lines [00:06, 23488.05 lines/s]\r",
      "192466 lines [00:06, 24948.89 lines/s]\r",
      "195084 lines [00:06, 25013.84 lines/s]\r",
      "198543 lines [00:06, 27278.88 lines/s]\r",
      "201919 lines [00:06, 28944.67 lines/s]\r",
      "204926 lines [00:06, 28974.94 lines/s]\r",
      "207902 lines [00:06, 29186.25 lines/s]\r",
      "210876 lines [00:07, 28575.09 lines/s]\r",
      "213838 lines [00:07, 28880.26 lines/s]\r",
      "217201 lines [00:07, 30156.07 lines/s]\r",
      "220673 lines [00:07, 31391.06 lines/s]\r",
      "223851 lines [00:07, 27826.68 lines/s]\r",
      "227300 lines [00:07, 29538.19 lines/s]\r",
      "230866 lines [00:07, 31141.01 lines/s]\r",
      "234699 lines [00:07, 32996.48 lines/s]\r",
      "238602 lines [00:07, 34600.63 lines/s]\r",
      "242235 lines [00:07, 35099.36 lines/s]\r",
      "245936 lines [00:08, 35649.09 lines/s]\r",
      "249747 lines [00:08, 36351.39 lines/s]\r",
      "253506 lines [00:08, 36714.12 lines/s]\r",
      "257204 lines [00:08, 33073.24 lines/s]\r",
      "260932 lines [00:08, 34230.86 lines/s]\r",
      "264803 lines [00:08, 35459.93 lines/s]\r",
      "268469 lines [00:08, 35811.48 lines/s]\r",
      "272427 lines [00:08, 36862.49 lines/s]\r",
      "276343 lines [00:08, 37521.14 lines/s]\r",
      "280266 lines [00:09, 38016.07 lines/s]\r",
      "284090 lines [00:09, 37761.02 lines/s]\r",
      "287883 lines [00:09, 33629.95 lines/s]\r",
      "291342 lines [00:09, 33608.55 lines/s]\r",
      "294770 lines [00:09, 33571.47 lines/s]\r",
      "298403 lines [00:09, 34353.75 lines/s]\r",
      "302015 lines [00:09, 34864.31 lines/s]\r",
      "305562 lines [00:09, 35042.38 lines/s]\r",
      "309086 lines [00:09, 31551.18 lines/s]\r",
      "312530 lines [00:10, 32365.33 lines/s]\r",
      "315971 lines [00:10, 32948.88 lines/s]\r",
      "319312 lines [00:10, 31673.65 lines/s]\r",
      "322522 lines [00:10, 31366.85 lines/s]\r",
      "325689 lines [00:10, 30344.76 lines/s]\r",
      "328865 lines [00:10, 27638.77 lines/s]\r",
      "332249 lines [00:10, 29245.00 lines/s]\r",
      "336174 lines [00:10, 31665.71 lines/s]\r",
      "339964 lines [00:10, 33307.38 lines/s]\r",
      "343558 lines [00:10, 34054.15 lines/s]\r",
      "347123 lines [00:11, 34516.28 lines/s]\r",
      "350628 lines [00:11, 33981.96 lines/s]\r",
      "354201 lines [00:11, 34484.13 lines/s]\r",
      "358005 lines [00:11, 35477.81 lines/s]\r",
      "362031 lines [00:11, 36788.32 lines/s]\r",
      "366050 lines [00:11, 37746.64 lines/s]\r",
      "369853 lines [00:11, 37319.53 lines/s]\r",
      "373703 lines [00:11, 37664.96 lines/s]\r",
      "377485 lines [00:11, 37170.12 lines/s]\r",
      "381215 lines [00:12, 36958.38 lines/s]\r",
      "384920 lines [00:12, 36086.81 lines/s]\r",
      "388540 lines [00:12, 36041.09 lines/s]\r",
      "392639 lines [00:12, 37393.68 lines/s]\r",
      "396673 lines [00:12, 38231.33 lines/s]\r",
      "400856 lines [00:12, 39241.04 lines/s]\r",
      "404799 lines [00:12, 38005.94 lines/s]\r",
      "408622 lines [00:12, 36691.00 lines/s]\r",
      "412317 lines [00:12, 36733.16 lines/s]\r",
      "416008 lines [00:12, 36575.74 lines/s]\r",
      "419678 lines [00:13, 33549.12 lines/s]\r",
      "423204 lines [00:13, 34042.37 lines/s]\r",
      "426768 lines [00:13, 34504.83 lines/s]\r",
      "430387 lines [00:13, 34993.10 lines/s]\r",
      "433910 lines [00:13, 34385.02 lines/s]\r",
      "437368 lines [00:13, 34039.32 lines/s]\r",
      "440786 lines [00:13, 33517.56 lines/s]\r",
      "444283 lines [00:13, 33939.43 lines/s]\r",
      "448047 lines [00:13, 34969.31 lines/s]\r",
      "451594 lines [00:13, 35115.70 lines/s]\r",
      "455457 lines [00:14, 36100.57 lines/s]\r",
      "459319 lines [00:14, 36820.52 lines/s]\r",
      "463015 lines [00:14, 36781.63 lines/s]\r",
      "466952 lines [00:14, 37519.70 lines/s]\r",
      "470764 lines [00:14, 37697.68 lines/s]\r",
      "474542 lines [00:14, 31041.78 lines/s]\r",
      "478141 lines [00:14, 32375.45 lines/s]\r",
      "481967 lines [00:14, 33940.30 lines/s]\r",
      "485815 lines [00:14, 35183.75 lines/s]\r",
      "489820 lines [00:15, 36513.85 lines/s]\r",
      "493673 lines [00:15, 37094.94 lines/s]\r",
      "497444 lines [00:15, 37215.96 lines/s]\r",
      "501417 lines [00:15, 37933.85 lines/s]\r",
      "505318 lines [00:15, 38249.40 lines/s]\r",
      "509331 lines [00:15, 38792.16 lines/s]\r",
      "513252 lines [00:15, 38915.43 lines/s]\r",
      "517157 lines [00:15, 38860.38 lines/s]\r",
      "521053 lines [00:15, 38880.76 lines/s]\r",
      "524964 lines [00:15, 38947.82 lines/s]\r",
      "528864 lines [00:16, 38818.64 lines/s]\r",
      "532760 lines [00:16, 38858.74 lines/s]\r",
      "536696 lines [00:16, 39005.54 lines/s]\r",
      "540646 lines [00:16, 39150.28 lines/s]\r",
      "544563 lines [00:16, 39153.94 lines/s]\r",
      "548480 lines [00:16, 38975.30 lines/s]\r",
      "552379 lines [00:16, 38757.84 lines/s]\r",
      "556256 lines [00:16, 38555.73 lines/s]\r",
      "560113 lines [00:16, 38118.62 lines/s]\r",
      "563927 lines [00:16, 37990.75 lines/s]\r",
      "567728 lines [00:17, 37973.07 lines/s]\r",
      "571689 lines [00:17, 38449.70 lines/s]\r",
      "575537 lines [00:17, 38125.34 lines/s]\r",
      "579352 lines [00:17, 37955.82 lines/s]\r",
      "583150 lines [00:17, 37533.33 lines/s]\r",
      "586906 lines [00:17, 37167.12 lines/s]\r",
      "590655 lines [00:17, 37263.09 lines/s]\r",
      "594384 lines [00:17, 29802.36 lines/s]\r",
      "597603 lines [00:17, 30480.09 lines/s]\r",
      "600918 lines [00:18, 31233.28 lines/s]\r",
      "604362 lines [00:18, 32128.63 lines/s]\r",
      "605248 lines [00:18, 33248.80 lines/s]\n",
      "Read sites summary\n",
      "\ttotal: 605,248\n",
      "\tunmethylated: 596,963\n",
      "\tmethylated: 8,285\n",
      "\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "\r",
      "  0%|          | 0/340081 [00:00<?, ? positions/s]\r",
      " 17%|█▋        | 57573/340081 [00:00<00:00, 575727.80 positions/s]\r",
      " 35%|███▌      | 120274/340081 [00:00<00:00, 590207.28 positions/s]\r",
      " 54%|█████▍    | 184011/340081 [00:00<00:00, 603605.12 positions/s]\r",
      " 73%|███████▎  | 247852/340081 [00:00<00:00, 613639.96 positions/s]\r",
      " 91%|█████████ | 309644/340081 [00:00<00:00, 614915.72 positions/s]\r",
      "100%|██████████| 340081/340081 [00:00<00:00, 620659.10 positions/s]\n",
      "Genomic positions summary\n",
      "\ttotal: 340,081\n",
      "\tlow_coverage: 339,292\n",
      "\tlow_meth_freq: 647\n",
      "\tvalid: 142\n",
      "\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "NanopolishComp Freq_meth_calculate --verbose -i data/freq_meth_calculate/methylation_calls.tsv -o ./output/freq_meth_calculate/\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### From standard input to a file (the same can be done with Nanopolish call-methylation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example the split_group option was also selected leading to motif groups to be split in individual positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:37:09.151215Z",
     "start_time": "2019-05-02T09:36:48.451652Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t289557\t000090\t-\n",
      "chr-XII\t451976\t451977\t289563\t000071\t-\n",
      "chr-XII\t451979\t451980\t289564\t000071\t-\n",
      "chr-XII\t452137\t452138\t289574\t000050\t-\n",
      "chr-XII\t452217\t452218\t289577\t000272\t-\n",
      "chr-XII\t452220\t452221\t289578\t000272\t-\n",
      "chr-XII\t452331\t452332\t288999\t000062\t+\n",
      "chr-XII\t452340\t452341\t289000\t000062\t+\n",
      "chr-XII\t452632\t452633\t289008\t000066\t+\n",
      "chrom\tstart\tend\tstrand\tsite_id\tmethylated_reads\tunmethylated_reads\tambiguous_reads\tsequence\tnum_motifs\tmeth_freq\n",
      "chr-XII\t465335\t465336\t-\t290122\t6\t5\t29\tCAGCACGACGG\t2\t0.545455\n",
      "chr-XII\t465338\t465339\t-\t290123\t6\t5\t29\tCACGACGGAGT\t2\t0.545455\n",
      "chr-XII\t456198\t456199\t-\t289760\t7\t7\t32\tCAGCACGACGG\t2\t0.500000\n",
      "chr-XII\t456201\t456202\t-\t289761\t7\t7\t32\tCACGACGGAGT\t2\t0.500000\n",
      "chr-XII\t460773\t460774\t+\t289338\t17\t20\t77\tCTCATCGTATA\t1\t0.459459\n",
      "chr-XII\t457104\t457105\t+\t289216\t6\t8\t43\tAACTACGAGCT\t1\t0.428571\n",
      "chr-XII\t461274\t461275\t+\t289359\t5\t7\t85\tTTATCCGAATG\t1\t0.416667\n",
      "chr-XII\t463406\t463407\t+\t289460\t6\t12\t29\tTCTTTCGCCCC\t1\t0.333333\n",
      "chr-XII\t457887\t457888\t-\t289825\t6\t12\t53\tGCTGCCGGAAA\t1\t0.333333\n",
      "General options:\n",
      "\tpackage_name: NanopolishComp\n",
      "\tpackage_version: 0.6.1\n",
      "\ttimestamp: 2019-05-02 10:36:49.281772\n",
      "\tquiet: False\n",
      "\tverbose: False\n",
      "\tmotif: cpg\n",
      "\tsplit_group: True\n",
      "\tmin_meth_freq: 0.05\n",
      "\tmin_depth: 10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "\r",
      "0 lines [00:00, ? lines/s]\r",
      "3660 lines [00:00, 36594.36 lines/s]\r",
      "7305 lines [00:00, 36549.98 lines/s]\r",
      "11022 lines [00:00, 36732.82 lines/s]\r",
      "14590 lines [00:00, 36404.52 lines/s]\r",
      "18237 lines [00:00, 36423.08 lines/s]\r",
      "21489 lines [00:00, 35155.99 lines/s]\r",
      "25229 lines [00:00, 35798.62 lines/s]\r",
      "28738 lines [00:00, 35580.30 lines/s]\r",
      "32470 lines [00:00, 36084.52 lines/s]\r",
      "36035 lines [00:01, 35952.29 lines/s]\r",
      "39662 lines [00:01, 36044.87 lines/s]\r",
      "43309 lines [00:01, 36170.40 lines/s]\r",
      "46878 lines [00:01, 35059.47 lines/s]\r",
      "50474 lines [00:01, 35324.45 lines/s]\r",
      "53989 lines [00:01, 35260.27 lines/s]\r",
      "57580 lines [00:01, 35452.52 lines/s]\r",
      "61173 lines [00:01, 35592.97 lines/s]\r",
      "64912 lines [00:01, 36113.56 lines/s]\r",
      "68522 lines [00:01, 35292.55 lines/s]\r",
      "72054 lines [00:02, 33361.66 lines/s]\r",
      "75414 lines [00:02, 30677.86 lines/s]\r",
      "78612 lines [00:02, 31055.75 lines/s]\r",
      "82136 lines [00:02, 32201.57 lines/s]\r",
      "85397 lines [00:02, 31997.81 lines/s]\r",
      "89118 lines [00:02, 33398.45 lines/s]\r",
      "92866 lines [00:02, 34525.73 lines/s]\r",
      "96592 lines [00:02, 35300.95 lines/s]\r",
      "100303 lines [00:02, 35823.20 lines/s]\r",
      "103908 lines [00:02, 35688.30 lines/s]\r",
      "107493 lines [00:03, 34436.79 lines/s]\r",
      "110972 lines [00:03, 34538.99 lines/s]\r",
      "114441 lines [00:03, 34271.26 lines/s]\r",
      "117966 lines [00:03, 34558.07 lines/s]\r",
      "121635 lines [00:03, 35169.98 lines/s]\r",
      "125225 lines [00:03, 35383.59 lines/s]\r",
      "128770 lines [00:03, 35304.81 lines/s]\r",
      "132305 lines [00:03, 31838.46 lines/s]\r",
      "135628 lines [00:03, 32242.62 lines/s]\r",
      "138901 lines [00:04, 32062.87 lines/s]\r",
      "142334 lines [00:04, 32710.07 lines/s]\r",
      "145680 lines [00:04, 32930.49 lines/s]\r",
      "149075 lines [00:04, 33227.86 lines/s]\r",
      "152683 lines [00:04, 34033.20 lines/s]\r",
      "156124 lines [00:04, 34143.54 lines/s]\r",
      "159549 lines [00:04, 34134.39 lines/s]\r",
      "163173 lines [00:04, 34739.13 lines/s]\r",
      "166803 lines [00:04, 35191.56 lines/s]\r",
      "170452 lines [00:04, 35569.04 lines/s]\r",
      "174015 lines [00:05, 35381.86 lines/s]\r",
      "177692 lines [00:05, 35785.94 lines/s]\r",
      "181407 lines [00:05, 36184.00 lines/s]\r",
      "185030 lines [00:05, 31620.45 lines/s]\r",
      "188423 lines [00:05, 32277.76 lines/s]\r",
      "191770 lines [00:05, 32624.24 lines/s]\r",
      "195088 lines [00:05, 31701.81 lines/s]\r",
      "198303 lines [00:05, 31735.59 lines/s]\r",
      "201508 lines [00:05, 31789.64 lines/s]\r",
      "204959 lines [00:05, 32557.50 lines/s]\r",
      "208235 lines [00:06, 30753.33 lines/s]\r",
      "211657 lines [00:06, 31716.38 lines/s]\r",
      "214870 lines [00:06, 31839.28 lines/s]\r",
      "218317 lines [00:06, 32584.32 lines/s]\r",
      "221798 lines [00:06, 33221.07 lines/s]\r",
      "225283 lines [00:06, 33691.84 lines/s]\r",
      "229067 lines [00:06, 34836.27 lines/s]\r",
      "232570 lines [00:06, 33889.07 lines/s]\r",
      "236123 lines [00:06, 34360.87 lines/s]\r",
      "239575 lines [00:07, 31071.29 lines/s]\r",
      "242753 lines [00:07, 30260.52 lines/s]\r",
      "245834 lines [00:07, 30093.04 lines/s]\r",
      "249300 lines [00:07, 27995.33 lines/s]\r",
      "252156 lines [00:07, 27620.32 lines/s]\r",
      "255788 lines [00:07, 29757.23 lines/s]\r",
      "259139 lines [00:07, 30790.04 lines/s]\r",
      "262280 lines [00:07, 29578.48 lines/s]\r",
      "265338 lines [00:07, 29870.61 lines/s]\r",
      "268808 lines [00:08, 31171.04 lines/s]\r",
      "272411 lines [00:08, 32485.18 lines/s]\r",
      "275848 lines [00:08, 33025.07 lines/s]\r",
      "279362 lines [00:08, 33631.51 lines/s]\r",
      "282958 lines [00:08, 34296.32 lines/s]\r",
      "286513 lines [00:08, 34662.18 lines/s]\r",
      "289995 lines [00:08, 34403.68 lines/s]\r",
      "293447 lines [00:08, 34383.59 lines/s]\r",
      "296894 lines [00:08, 34124.02 lines/s]\r",
      "300313 lines [00:08, 33738.57 lines/s]\r",
      "303861 lines [00:09, 34242.39 lines/s]\r",
      "307291 lines [00:09, 33435.05 lines/s]\r",
      "310650 lines [00:09, 33478.89 lines/s]\r",
      "314004 lines [00:09, 33308.01 lines/s]\r",
      "317339 lines [00:09, 27854.65 lines/s]\r",
      "320540 lines [00:09, 28981.76 lines/s]\r",
      "323669 lines [00:09, 29636.99 lines/s]\r",
      "326874 lines [00:09, 30318.30 lines/s]\r",
      "329971 lines [00:09, 29616.83 lines/s]\r",
      "332982 lines [00:10, 28386.56 lines/s]\r",
      "336457 lines [00:10, 30035.29 lines/s]\r",
      "340047 lines [00:10, 31582.00 lines/s]\r",
      "343267 lines [00:10, 30833.07 lines/s]\r",
      "346398 lines [00:10, 30084.85 lines/s]\r",
      "349444 lines [00:10, 24273.24 lines/s]\r",
      "352120 lines [00:10, 24968.22 lines/s]\r",
      "354767 lines [00:10, 24715.84 lines/s]\r",
      "357792 lines [00:10, 26149.37 lines/s]\r",
      "361312 lines [00:11, 28333.77 lines/s]\r",
      "364983 lines [00:11, 30414.84 lines/s]\r",
      "368152 lines [00:11, 28075.67 lines/s]\r",
      "371085 lines [00:11, 27074.29 lines/s]\r",
      "373888 lines [00:11, 26998.36 lines/s]\r",
      "376941 lines [00:11, 27964.39 lines/s]\r",
      "379793 lines [00:11, 28087.97 lines/s]\r",
      "382641 lines [00:11, 28128.82 lines/s]\r",
      "385808 lines [00:11, 29104.40 lines/s]\r",
      "389139 lines [00:12, 30249.78 lines/s]\r",
      "393013 lines [00:12, 32377.42 lines/s]\r",
      "396886 lines [00:12, 34051.53 lines/s]\r",
      "400717 lines [00:12, 35225.31 lines/s]\r",
      "404297 lines [00:12, 35224.60 lines/s]\r",
      "407860 lines [00:12, 35110.34 lines/s]\r",
      "411400 lines [00:12, 34914.95 lines/s]\r",
      "414935 lines [00:12, 35042.96 lines/s]\r",
      "418454 lines [00:12, 34941.25 lines/s]\r",
      "421986 lines [00:12, 35052.46 lines/s]\r",
      "425499 lines [00:13, 34683.09 lines/s]\r",
      "429014 lines [00:13, 34821.44 lines/s]\r",
      "432501 lines [00:13, 34679.64 lines/s]\r",
      "435973 lines [00:13, 34555.08 lines/s]\r",
      "439431 lines [00:13, 34209.03 lines/s]\r",
      "442855 lines [00:13, 33675.42 lines/s]\r",
      "446227 lines [00:13, 33561.87 lines/s]\r",
      "449586 lines [00:13, 33156.07 lines/s]\r",
      "452905 lines [00:13, 32692.79 lines/s]\r",
      "456178 lines [00:13, 32700.10 lines/s]\r",
      "459451 lines [00:14, 25474.49 lines/s]\r",
      "462441 lines [00:14, 26657.37 lines/s]\r",
      "465538 lines [00:14, 27818.41 lines/s]\r",
      "468747 lines [00:14, 28969.41 lines/s]\r",
      "471758 lines [00:14, 23724.22 lines/s]\r",
      "474366 lines [00:14, 23195.75 lines/s]\r",
      "476853 lines [00:14, 21367.01 lines/s]\r",
      "480303 lines [00:14, 24121.06 lines/s]\r",
      "483746 lines [00:15, 26500.57 lines/s]\r",
      "486945 lines [00:15, 27933.65 lines/s]\r",
      "490096 lines [00:15, 28916.40 lines/s]\r",
      "493351 lines [00:15, 29916.61 lines/s]\r",
      "496912 lines [00:15, 31422.36 lines/s]\r",
      "500577 lines [00:15, 32825.30 lines/s]\r",
      "504187 lines [00:15, 33741.61 lines/s]\r",
      "507624 lines [00:15, 33744.02 lines/s]\r",
      "511043 lines [00:15, 30707.59 lines/s]\r",
      "514198 lines [00:16, 25399.90 lines/s]\r",
      "516949 lines [00:16, 23455.54 lines/s]\r",
      "519472 lines [00:16, 23478.95 lines/s]\r",
      "521944 lines [00:16, 23361.67 lines/s]\r",
      "525381 lines [00:16, 25844.99 lines/s]\r",
      "528863 lines [00:16, 28010.59 lines/s]\r",
      "532352 lines [00:16, 29770.84 lines/s]\r",
      "535884 lines [00:16, 31242.33 lines/s]\r",
      "539132 lines [00:16, 31601.62 lines/s]\r",
      "542405 lines [00:17, 31930.57 lines/s]\r",
      "545927 lines [00:17, 32849.11 lines/s]\r",
      "549553 lines [00:17, 33802.19 lines/s]\r",
      "552981 lines [00:17, 33942.53 lines/s]\r",
      "556559 lines [00:17, 34473.06 lines/s]\r",
      "560028 lines [00:17, 34339.25 lines/s]\r",
      "563698 lines [00:17, 35013.97 lines/s]\r",
      "567214 lines [00:17, 34478.92 lines/s]\r",
      "570674 lines [00:17, 26730.81 lines/s]\r",
      "573775 lines [00:18, 27883.53 lines/s]\r",
      "577224 lines [00:18, 29582.22 lines/s]\r",
      "580722 lines [00:18, 31016.60 lines/s]\r",
      "584152 lines [00:18, 31932.82 lines/s]\r",
      "587638 lines [00:18, 32757.70 lines/s]\r",
      "590992 lines [00:18, 32408.33 lines/s]\r",
      "594342 lines [00:18, 32726.35 lines/s]\r",
      "597887 lines [00:18, 33498.45 lines/s]\r",
      "601454 lines [00:18, 34121.18 lines/s]\r",
      "604929 lines [00:18, 34304.67 lines/s]\r",
      "605248 lines [00:18, 31903.81 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "\r",
      "  0%|          | 0/455542 [00:00<?, ? positions/s]\r",
      " 13%|█▎        | 60956/455542 [00:00<00:00, 609557.67 positions/s]\r",
      " 27%|██▋       | 122135/455542 [00:00<00:00, 610224.96 positions/s]\r",
      " 41%|████      | 185490/455542 [00:00<00:00, 617038.58 positions/s]\r",
      " 54%|█████▍    | 248112/455542 [00:00<00:00, 619762.59 positions/s]\r",
      " 68%|██████▊   | 308294/455542 [00:00<00:00, 614267.31 positions/s]\r",
      " 80%|████████  | 366070/455542 [00:00<00:00, 602837.63 positions/s]\r",
      " 93%|█████████▎| 425574/455542 [00:00<00:00, 600475.40 positions/s]\r",
      "100%|██████████| 455542/455542 [00:00<00:00, 606728.55 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Load local bashrc and activate virtual environment\n",
    "source ~/.bashrc\n",
    "workon Nanopolish_0.11.1\n",
    "\n",
    "cat data/freq_meth_calculate/methylation_calls.tsv | NanopolishComp Freq_meth_calculate -o ./output/freq_meth_calculate/ --split_group\n",
    "\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.bed\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.tsv\n",
    "head ./output/freq_meth_calculate/out_freq_meth_calculate.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python API usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import the package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:37:09.164849Z",
     "start_time": "2019-05-02T09:37:09.154730Z"
    }
   },
   "outputs": [],
   "source": [
    "# Import main program\n",
    "from NanopolishComp.Freq_meth_calculate import Freq_meth_calculate\n",
    "\n",
    "# Import helper functions\n",
    "from NanopolishComp.common import jhelp, head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### python API help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:37:09.193741Z",
     "start_time": "2019-05-02T09:37:09.167671Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "**NanopolishComp.Freq_meth_calculate.__init__**\n",
       "\n",
       "Calculate methylation frequency at genomic CpG sites from the output of nanopolish call-methylation\n",
       "\n",
       "---\n",
       "\n",
       "* **input_fn** *: str (required)*\n",
       "\n",
       "Path to a nanopolish call_methylation tsv output file\n",
       "\n",
       "* **outdir** *: str (default = ./)*\n",
       "\n",
       "Path to the output folder\n",
       "\n",
       "* **outprefix** *: str (default = out)*\n",
       "\n",
       "text outprefix for all the files generated\n",
       "\n",
       "* **min_llr** *: float (default = 2.5)*\n",
       "\n",
       "Log likelihood ratio threshold\n",
       "\n",
       "* **min_depth** *: int (default = 10)*\n",
       "\n",
       "Minimal number of reads covering a site to be reported\n",
       "\n",
       "* **min_meth_freq** *: float (default = 0.05)*\n",
       "\n",
       "Minimal methylation frequency of a site to be reported\n",
       "\n",
       "* **split_group** *: bool (default = False)*\n",
       "\n",
       "If True, multi motif groups (sequence with close motifs) are split in individual site\n",
       "\n",
       "* **motif** *: {cpg,gpc,dam,dcm} (default = cpg)*\n",
       "\n",
       "Methylation motif type\n",
       "\n",
       "* **verbose** *: bool (default = False)*\n",
       "\n",
       "Increase verbosity\n",
       "\n",
       "* **quiet** *: bool (default = False)*\n",
       "\n",
       "Reduce verbosity\n",
       "\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "jhelp(Freq_meth_calculate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-23T09:19:41.284517Z",
     "start_time": "2019-04-23T09:19:41.165804Z"
    },
    "scrolled": false
   },
   "source": [
    "#### basic setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:37:28.557304Z",
     "start_time": "2019-05-02T09:37:09.197191Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:18, 32549.90 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 340081/340081 [00:00<00:00, 517704.63 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "General options:                      \n",
      "package_name: NanopolishComp          \n",
      "package_version: 0.6.1                \n",
      "timestamp: 2019-05-02 10:37:09.200694 \n",
      "quiet: False                          \n",
      "verbose: False                        \n",
      "motif: cpg                            \n",
      "split_group: False                    \n",
      "min_meth_freq: 0.05                   \n",
      "min_depth: 10                         \n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence       num_motifs meth_freq \n",
      "chr-XII 465335 465339 -      217103  6                5                  29              CAGCACGACGGAGT 2          0.545455  \n",
      "chr-XII 456198 456202 -      216869  7                7                  32              CAGCACGACGGAGT 2          0.500000  \n",
      "chr-XII 460773 460774 +      216597  17               20                 77              CTCATCGTATA    1          0.459459  \n",
      "chr-XII 457104 457105 +      216519  6                8                  43              AACTACGAGCT    1          0.428571  \n",
      "chr-XII 461274 461275 +      216614  5                7                  85              TTATCCGAATG    1          0.416667  \n",
      "chr-XII 463406 463407 +      216682  6                12                 29              TCTTTCGCCCC    1          0.333333  \n",
      "chr-XII 457887 457888 -      216914  6                12                 53              GCTGCCGGAAA    1          0.333333  \n",
      "chr-XII 461077 461078 +      216605  8                17                 74              CTTTCCGCCAA    1          0.320000  \n",
      "chr-XII 462283 462284 -      217018  5                11                 57              AGGCACGCAAG    1          0.312500  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t216740\t000090\t-\n",
      "chr-XII\t451976\t451980\t216746\t000071\t-\n",
      "chr-XII\t452137\t452138\t216753\t000050\t-\n",
      "chr-XII\t452217\t452221\t216756\t000272\t-\n",
      "chr-XII\t452331\t452341\t216384\t000062\t+\n",
      "chr-XII\t452632\t452633\t216391\t000066\t+\n",
      "chr-XII\t452654\t452655\t216769\t000100\t-\n",
      "chr-XII\t453086\t453087\t216781\t000055\t-\n",
      "chr-XII\t453336\t453337\t216790\t000066\t-\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\")\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.log\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### split motif groups in individual genomic position "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:37:50.552609Z",
     "start_time": "2019-05-02T09:37:28.559341Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:20, 28854.33 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 455542/455542 [00:00<00:00, 510214.98 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "General options:                      \n",
      "package_name: NanopolishComp          \n",
      "package_version: 0.6.1                \n",
      "timestamp: 2019-05-02 10:37:28.561040 \n",
      "quiet: False                          \n",
      "verbose: False                        \n",
      "motif: cpg                            \n",
      "split_group: True                     \n",
      "min_meth_freq: 0.05                   \n",
      "min_depth: 10                         \n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence    num_motifs meth_freq \n",
      "chr-XII 465335 465336 -      630203  6                5                  29              CAGCACGACGG 2          0.545455  \n",
      "chr-XII 465338 465339 -      630204  6                5                  29              CACGACGGAGT 2          0.545455  \n",
      "chr-XII 456198 456199 -      629841  7                7                  32              CAGCACGACGG 2          0.500000  \n",
      "chr-XII 456201 456202 -      629842  7                7                  32              CACGACGGAGT 2          0.500000  \n",
      "chr-XII 460773 460774 +      629419  17               20                 77              CTCATCGTATA 1          0.459459  \n",
      "chr-XII 457104 457105 +      629297  6                8                  43              AACTACGAGCT 1          0.428571  \n",
      "chr-XII 461274 461275 +      629440  5                7                  85              TTATCCGAATG 1          0.416667  \n",
      "chr-XII 463406 463407 +      629541  6                12                 29              TCTTTCGCCCC 1          0.333333  \n",
      "chr-XII 457887 457888 -      629906  6                12                 53              GCTGCCGGAAA 1          0.333333  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t451856\t451857\t629638\t000090\t-\n",
      "chr-XII\t451976\t451977\t629644\t000071\t-\n",
      "chr-XII\t451979\t451980\t629645\t000071\t-\n",
      "chr-XII\t452137\t452138\t629655\t000050\t-\n",
      "chr-XII\t452217\t452218\t629658\t000272\t-\n",
      "chr-XII\t452220\t452221\t629659\t000272\t-\n",
      "chr-XII\t452331\t452332\t629080\t000062\t+\n",
      "chr-XII\t452340\t452341\t629081\t000062\t+\n",
      "chr-XII\t452632\t452633\t629089\t000066\t+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\",\n",
    "    split_group=True)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.log\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Changing filtering threshold (not recommended)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:38:10.483098Z",
     "start_time": "2019-05-02T09:37:50.555169Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "## Checking arguments ##\n",
      "Test input file readability\n",
      "Testing output dir writability\n",
      "## Parsing methylation_calls file ##\n",
      "Starting to parse file methylation_calls file\n",
      "605248 lines [00:19, 31538.44 lines/s]\n",
      "Filtering out positions with low coverage or methylation frequency\n",
      "100%|██████████| 340081/340081 [00:00<00:00, 539066.80 positions/s]\n",
      "## Write output files ##\n",
      "Writing bed file\n",
      "Writing tsv file\n",
      "Writing log file\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "General options:                      \n",
      "package_name: NanopolishComp          \n",
      "package_version: 0.6.1                \n",
      "timestamp: 2019-05-02 10:37:50.556724 \n",
      "quiet: False                          \n",
      "verbose: False                        \n",
      "motif: cpg                            \n",
      "split_group: False                    \n",
      "min_meth_freq: 0.3                    \n",
      "min_depth: 20                         \n",
      "\n",
      "chrom   start  end    strand site_id methylated_reads unmethylated_reads ambiguous_reads sequence       num_motifs meth_freq \n",
      "chr-XII 461541 461542 +      1012246 26               3                  52              AATTCCGAGGG    1          0.896552  \n",
      "chr-XII 455556 455557 +      1012101 16               6                  20              AGATCCGTTGT    1          0.727273  \n",
      "chr-XII 462606 462607 +      1012278 21               9                  28              AATTCCGGGGT    1          0.700000  \n",
      "chr-XII 458208 458209 -      1012544 25               15                 37              TTAAACGCAAA    1          0.625000  \n",
      "chr-XII 463513 463514 -      1012680 13               8                  29              TCTTTCGGGTC    1          0.619048  \n",
      "chr-XII 461032 461033 -      1012599 27               20                 60              CCCCACGCAAA    1          0.574468  \n",
      "chr-XII 456198 456202 -      1012492 17               13                 16              CAGCACGACGGAGT 2          0.566667  \n",
      "chr-XII 461274 461275 +      1012237 21               21                 55              TTATCCGAATG    1          0.500000  \n",
      "chr-XII 465464 465465 -      1012730 10               10                 19              TCCATCGGCTT    1          0.500000  \n",
      "\n",
      "track name='out' description='Methylation frequency track generated with NanopolishComp' useScore=1\n",
      "chr-XII\t455151\t455152\t1012091\t000409\t+\n",
      "chr-XII\t455556\t455557\t1012101\t000727\t+\n",
      "chr-XII\t456198\t456202\t1012492\t000566\t-\n",
      "chr-XII\t456451\t456452\t1012499\t000303\t-\n",
      "chr-XII\t456686\t456687\t1012128\t000321\t+\n",
      "chr-XII\t457045\t457046\t1012140\t000476\t+\n",
      "chr-XII\t457104\t457105\t1012142\t000447\t+\n",
      "chr-XII\t457887\t457888\t1012537\t000361\t-\n",
      "chr-XII\t458208\t458209\t1012544\t000625\t-\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f = Freq_meth_calculate(\n",
    "    input_fn=\"./data/freq_meth_calculate/methylation_calls.tsv\",\n",
    "    outdir=\"./output/freq_meth_calculate/\",\n",
    "    min_llr=1,\n",
    "    min_depth=20,\n",
    "    min_meth_freq=0.3)\n",
    "\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.log\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.tsv\")\n",
    "head(\"./output/freq_meth_calculate/out_freq_meth_calculate.bed\")"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python3.6",
   "language": "python",
   "name": "python3.6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
